<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Space Plasma Simulations with Vlasiator on LUMI Supercomputer documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css" />
      <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
      <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
      <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
      <link rel="stylesheet" type="text/css" href="_static/sphinx_lesson.css" />
      <link rel="stylesheet" type="text/css" href="_static/sphinx_rtd_theme_ext_color_contrast.css" />
      <link rel="stylesheet" type="text/css" href="_static/tabs.css" />
      <link rel="stylesheet" type="text/css" href="_static/overrides.css" />

  
    <link rel="shortcut icon" href="_static/favicon.ico"/>
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script src="_static/clipboard.min.js"></script>
        <script src="_static/copybutton.js"></script>
        <script src="_static/minipres.js"></script>
        <script>let toggleHintShow = 'Click to show';</script>
        <script>let toggleHintHide = 'Click to hide';</script>
        <script>let toggleOpenOnPrint = 'true';</script>
        <script src="_static/togglebutton.js"></script>
        <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
        <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="#" class="icon icon-home">
            Space Plasma Simulations with Vlasiator on LUMI Supercomputer
              <img src="_static/ENCCS.jpg" class="logo" alt="Logo"/>
          </a>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Prerequisites</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-introduction">Introduction to Vlasiator</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#rules-of-the-road">Rules of the Road</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-prepare_workspace">Preparations for the workshop</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">The lessons</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-installing-vlasiator">Installing Vlasiator</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-benchmarks_and_running">How to configure Vlasiator for performance</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#exercises">Exercises</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-vlasiator_outputs">Vlasiator output data</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-visit">VisIt and VLSV bootcamp</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-analysator">Introduction to Analysator</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-analysator_exercises">Analysator exercises</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-quick-reference">Quick Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-vlasiator-cfg-reference">Vlasiator config reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-analysator_supported">Analysator supported data reducers and vlasiator variables</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-new_project">New project from template</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">Space Plasma Simulations with Vlasiator on LUMI Supercomputer</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="#" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Space Plasma Simulations with Vlasiator on LUMI Supercomputer  documentation</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/ENCCS/content/blob/main/content/index" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="space-plasma-simulations-with-vlasiator-on-lumi-supercomputer">
<h1>Space Plasma Simulations with Vlasiator on LUMI Supercomputer<a class="headerlink" href="#space-plasma-simulations-with-vlasiator-on-lumi-supercomputer" title="Link to this heading"></a></h1>
<p>Plasma is one of four fundamental states of matter, characterized by the presence of
a significant portion of charged particles in any combination of ions or electrons.
Plasma is a powerful phase of matter as it is key to many applications and technologies,
from astrophysics to space physics, from clean energy production based on fusion to
compact and cheap particle accelerators. However, it is still a mysterious phase of matter
as its dynamics is inherently nonlinear, multidimensional and multiscale.</p>
<p>The theoretical description of plasma using analytic modeling tools is somewhat limited and
great progress in plasma physics has been achieved using computer simulations.
Indeed, advances in massively parallel simulations using high performance computing (HPC)
resources have have further boosted our understanding of plasma science at an unprecedented
resolutions and levels of physical fidelity. These simulations have been producing a new
wealth of knowledge and enabling key applications for science, industry, and society.</p>
<p><strong>Plasma-PEPSC</strong> (Plasma Exascale-Performance Simulation Centre) is a European Centre of Excellence leading
plasma science into the era of exascale computing to drive scientific breakthroughs in plasma science’s most
significant challenges (fusion energy, accelerator devices and space physics) through cutting-edge hardware and
software advancements.
The overarching goal of Plasma-PEPSC is to take this technological development to the next level, enabling unprecedented
simulations on current pre-exascale and future exascale platforms across Europe. Four flagship plasma codes with a
large user base – BIT, GENE, PIConGPU, and Vlasiator – serve as the focal points of the centre of excellence.
By maximising their parallel performance and efficiency, we aim to achieve breakthroughs in controlling plasma-material
interfaces, optimising magnetically confined fusion plasmas, designing next-generation plasma accelerators and predicting
space plasma dynamics within the Earth’s magnetosphere.</p>
<p><a class="reference external" href="https://www.helsinki.fi/en/researchgroups/vlasiator">**Vlasiator**</a> (<a class="reference external" href="https://github.com/fmihpc/vlasiator">GitHub</a>) is
the state-of-the-art hybrid-Vlasov simulation for ion-scale physics in a global
magnetospheric setting. It is the only 6D hybrid-Vlasov code capable of simulating
the Earth’s magnetosphere. In Vlasiator, ions are represented as velocity distribution functions,
while electrons are a massless charge-neutralizing fluid, enabling a self-consistent global plasma simulation
that can describe multi-temperature plasmas to resolve non-MHD processes that currently
cannot be self-consistently described by the existing global space weather simulations.
The novelty is that by modelling ions as velocity distribution functions the outcome will be
numerically noiseless. Due to the multi-dimensional approach at ion scales, Vlasiator’s
computational challenges are immense. Advanced HPC techniques will be adopted using tens of
thousands of cores to perform massively parallel computations.</p>
<div class="admonition-prerequisites prerequisites admonition" id="prerequisites-0">
<p class="admonition-title">Prerequisites</p>
<ul class="simple">
<li><p>PhD students, postdocs, industry engineers</p></li>
<li><p>Basic familiarity with general physics and plasma physics</p></li>
<li><p>Some previous practical experience running some CFD code</p></li>
<li><p>Basic familiarity with Unix shell and HPC environment</p></li>
<li><p>Some materials towards advanced users</p></li>
</ul>
</div>
<table class="docutils align-default">
<tbody>
<tr class="row-odd"><td><p>20 min</p></td>
<td><p><span class="xref std std-doc">filename</span></p></td>
</tr>
</tbody>
</table>
<div class="toctree-wrapper compound">
<span id="document-introduction"></span><section id="introduction-to-vlasiator">
<h2>Introduction to Vlasiator<a class="headerlink" href="#introduction-to-vlasiator" title="Link to this heading"></a></h2>
<section id="why-we-teach-this-lesson">
<h3>Why we teach this lesson<a class="headerlink" href="#why-we-teach-this-lesson" title="Link to this heading"></a></h3>
<p>Here we introduce what is Vlasiator.</p>
</section>
<section id="intended-learning-outcomes">
<h3>Intended learning outcomes<a class="headerlink" href="#intended-learning-outcomes" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>Know what Vlasiator is and does</p></li>
<li><p>Understand what you can do with Vlasiator</p></li>
<li><p>What resources do I need to run Vlasiator for some given task?</p></li>
<li><p>Rules of the Road for using Vlasiator</p></li>
</ul>
</section>
<section id="timing">
<h3>Timing<a class="headerlink" href="#timing" title="Link to this heading"></a></h3>
<p>Tuesday morning, 1h</p>
</section>
<section id="preparing-exercises">
<h3>Preparing exercises<a class="headerlink" href="#preparing-exercises" title="Link to this heading"></a></h3>
<p>Here you can find <a class="reference external" href="https://datacloud.helsinki.fi/index.php/s/wEZdF3szjBfapSs">Vlasiator introductory lectures</a>.</p>
<p>Reference papers for Vlasiator simulation:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://doi.org/10.1016/j.jastp.2014.08.012">Alfthan et al., 2014</a>, Vlasiator: First global hybrid-Vlasov simulations of Earth’s foreshock and magnetosheath. <em>JASTP</em>.</p></li>
<li><p><a class="reference external" href="https://doi.org/10.1007/s41115-018-0003-2">Palmroth et al., 2018</a>, Vlasov methods in space physics and astrophysics. <em>LRCA</em>.</p></li>
<li><p><a class="reference external" href="https://doi.org/10.1063/5.0134387">Ganse et al., 2023</a>,  Enabling technology for global 3D + 3V hybrid-Vlasov simulations of near-Earth space. <em>Phys. Plasmas</em>.</p></li>
</ul>
<p>Highlight results:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://doi.org/10.1038/s41567-022-01837-z">Turc et al., 2022</a>, Transmission of foreshock waves through Earth’s bow shock. <em>Nat. Phys</em>.</p></li>
<li><p><a class="reference external" href="https://doi.org/10.1038/s41561-023-01206-2">Palmroth et al. 2023</a>, Magnetotail plasma eruptions driven by magnetic reconnection and kinetic instabilities. <em>Nat. Geosci</em>.</p></li>
</ul>
<p>Find more at <a class="reference external" href="https://www.helsinki.fi/en/researchgroups/vlasiator/publications-and-presentations">the group website</a>.</p>
</section>
<section id="vlasiator">
<h3>Vlasiator<a class="headerlink" href="#vlasiator" title="Link to this heading"></a></h3>
<p><a class="reference external" href="https://www.helsinki.fi/en/researchgroups/vlasiator">**Vlasiator**</a> (<a class="reference external" href="https://github.com/fmihpc/vlasiator">GitHub</a>) is the state-of-the-art hybrid-Vlasov simulation for ion-scale physics in a global magnetospheric setting. It is the only 6D hybrid-Vlasov code capable of simulating the Earth’s magnetosphere. In Vlasiator, ions are represented as velocity distribution functions, while electrons are massless charge-neutralizing fluid, enabling a self-consistent global plasma simulation that can describe multi-temperature plasmas to resolve non-MHD processes that currently cannot be self-consistently described by the existing global space weather simulations. The novelty is that by modelling ions as velocity distribution functions the outcome will be numerically noiseless. Due to the multi-dimensional approach at ion scales, Vlasiator’s computational challenges are immense.</p>
<figure class="align-default" id="id4">
<a class="reference internal image-reference" href="_images/BCH_LRCA.webp"><img alt="Example of a Vlasiator 5D run (LRCA)" src="_images/BCH_LRCA.webp" style="width: 500px;" /></a>
<figcaption>
<p><span class="caption-text">A Vlasiator view of the Earth’s magnetosphere (5D, Palmroth+2018)</span><a class="headerlink" href="#id4" title="Link to this image"></a></p>
</figcaption>
</figure>
<p>Vlasiator simulates the dynamics of plasma using a hybrid-Vlasov model, where protons are described by their distribution function f(r,v,t) in ordinary (r) and velocity (v) space, and electrons are a charge-neutralising fluid. This approach neglects electron kinetic effects but retains ion kinetics. The time-evolution of f(r,v,t) is given by Vlasov’s equation,</p>
<a class="reference internal image-reference" href="_images/vlasov-eq.webp"><img alt="Vlasov's equation" src="_images/vlasov-eq.webp" style="width: 600px;" /></a>
<p>which is coupled self-consistently to Maxwell’s equations giving the evolution of the electric and magnetic fields E and B. Maxwell’s equations neglect the displacement current.</p>
<a class="reference internal image-reference" href="_images/faraday.png"><img alt="Faraday's law" src="_images/faraday.png" style="width: 200px;" /></a>
<p>and</p>
<a class="reference internal image-reference" href="_images/ampere.png"><img alt="Ampere's law" src="_images/ampere.png" style="width: 250px;" /></a>
<p>The equations are closed by a generalised Ohm’s law including the Hall term (and the electron pressure term, not shown below).</p>
<a class="reference internal image-reference" href="_images/ohm.webp"><img alt="Vlasov's equation" src="_images/ohm.webp" style="width: 300px;" /></a>
</section>
<section id="state-of-the-art">
<h3>State-of-the-art<a class="headerlink" href="#state-of-the-art" title="Link to this heading"></a></h3>
<p>Vlasiator is parallelized at multiple scales via MPI, threading support and vectorization, and we are continuously improving the performance and the feature set. GPU porting is in progress: solvers have been ported, but they need to be optimized for GPUs to be useful in production.</p>
<p>The data structures in Vlasiator have been optimized to propagate the 6D solution efficiently with explicit solvers. The velocity distribution functions are associated with a spatial, hierarchically-adapted Cartesian grid (the Vlasov grid, SpatialGrid or the AMR grid), with each such spatial cell containing a VDF for each ion species in the simulation. The VDF themselves are stored on a Cartesian velocity-space grid as a sparse data structure - otherwise the production simulations could not fit any available computers. This implies slight loss of mass at the fringes (under a set sparsity threshold, the VDF is not stored), but the VDF can optionally be re-scaled to conserve mass. Suitably chosen sparsity thresholds enable global runs with negligible adverse effects. Some 98% of the phase-space volume is discarded.</p>
<figure class="align-default" id="id5">
<a class="reference internal image-reference" href="_images/sparse-YPK2016.webp"><img alt="Example of sparse velocity grid." src="_images/sparse-YPK2016.webp" style="width: 500px;" /></a>
<figcaption>
<p><span class="caption-text">Example of sparse velocity grid, Yann Pfau-Kempf, 2016.</span><a class="headerlink" href="#id5" title="Link to this image"></a></p>
</figcaption>
</figure>
<figure class="align-default" id="id6">
<a class="reference internal image-reference" href="_images/aBlockyVDF.png"><img alt="Example of sparse velocity grid." src="_images/aBlockyVDF.png" style="width: 500px;" /></a>
<figcaption>
<p><span class="caption-text">Example of sparse velocity grid. The 4x4x4 structure stems from the block data structure.</span><a class="headerlink" href="#id6" title="Link to this image"></a></p>
</figcaption>
</figure>
<p>The numerical algorithm for the propagation of the Vlasov equation is based on operator splitting: spatial translation and velocity-space acceleration are leapfrogged. Spatial translation remaps the VDF as 1D shears. The acceleration re-mapping of the VDF is handled by the semi-Lagrangian SLICE-3D method by Zerroukat and Allen (2012), by decomposing v-space rotation and translation to a series of shear operations. These mappings are conservative by themselves.</p>
<figure class="align-default" id="id7">
<a class="reference internal image-reference" href="_images/Strang-Splitting.png"><img alt="_images/Strang-Splitting.png" src="_images/Strang-Splitting.png" style="width: 400px;" /></a>
<figcaption>
<p><span class="caption-text">Schematic of Strang-splitting the Vlasov propagation, left (yellow) the translation part, right (green) the acceleration part.</span><a class="headerlink" href="#id7" title="Link to this image"></a></p>
</figcaption>
</figure>
<p>Spatial AMR has lately enabled truly 6D simulations of the Earth’s magnetosphere (see Ganse+2023). The initial 6D production runs have used statically-refined grid, with a dynamic AMR currently being tested in production.</p>
<p>Spatial AMR, however, is <em>not</em> extended to the field solver (<em>heterologous</em> grids): Field solver has simple load balancing, while the Vlasov solver does not. The solution is to have <em>two</em> grids, with the field solver grid at the highest resolution of the Vlasov grid, throughout the entire domain. There are some details in cross-coupling e.g. plasma moments across these grids. With four levels of refinement, the field solver is contributing about 10% of runtime as it is - Vlasov solver is still the bottleneck.</p>
<figure class="align-default" id="id8">
<a class="reference internal image-reference" href="_images/FsGrid_LB.png"><img alt="_images/FsGrid_LB.png" src="_images/FsGrid_LB.png" style="width: 400px;" /></a>
<figcaption>
<p><span class="caption-text">Example of fieldsolver load balancing</span><a class="headerlink" href="#id8" title="Link to this image"></a></p>
</figcaption>
</figure>
<figure class="align-default" id="id9">
<a class="reference internal image-reference" href="_images/SpatialGrid_LB_RCB.png"><img alt="_images/SpatialGrid_LB_RCB.png" src="_images/SpatialGrid_LB_RCB.png" style="width: 400px;" /></a>
<figcaption>
<p><span class="caption-text">Example of SpatialGrid load balancing, legacy RCB</span><a class="headerlink" href="#id9" title="Link to this image"></a></p>
</figcaption>
</figure>
<p>Improved load balancing algorithms matter. In the above Recursive Coordinate Bisection (RCB) example, one can see small-scale structure in the domain boundaries, which is not optimal for MPI communication buffers. Forcing rectilinear domains for RCB helps, but so do using other methods, like Recursive Inertial Bisection (RIB, below), that can balance MPI communications over the three principal directions.</p>
<figure class="align-default" id="id10">
<a class="reference internal image-reference" href="_images/SpatialGrid_LB_RIB.png"><img alt="Example of load balancing of SpatialGrid, MPI ranks" src="_images/SpatialGrid_LB_RIB.png" style="width: 400px;" /></a>
<figcaption>
<p><span class="caption-text">Example of load balancing of SpatialGrid, showing the decomposition to MPI ranks, currently used in production.</span><a class="headerlink" href="#id10" title="Link to this image"></a></p>
</figcaption>
</figure>
<figure class="align-default" id="id11">
<a class="reference internal image-reference" href="_images/SpatialGrid_LB_blocks.png"><img alt="Example of load balancing of SpatialGrid" src="_images/SpatialGrid_LB_blocks.png" style="width: 400px;" /></a>
<figcaption>
<p><span class="caption-text">Example of load balancing of SpatialGrid, showing the velocity-space block-counts.</span><a class="headerlink" href="#id11" title="Link to this image"></a></p>
</figcaption>
</figure>
<p>Recently, we have also included a proper ionospheric inner boundary condition, with an ionosphere solver along the lines of MHD solvers. Further developments are ongoing, incl. better ionization and conductivity profiles. Outer boundary conditions now include the possibility of time-varying solar wind parameters.</p>
<p>Lately, we have included an electron module, eVlasiator (<a class="reference external" href="https://doi.org/10.5194/angeo-39-85-2021">Battarbee+2021</a>, <a class="reference external" href="https://doi.org/10.1029/2022GL098329">Alho+2022</a>), that takes an existing Vlasiator solution and runs electron VDFs against that background.</p>
<p>Deploying Vlasiator to various environments has lately (with Plasma-PEPSC!) been made much easier, including compilation on ARM. We’ll look at the build process later!</p>
</section>
<section id="practical-aspects-of-running-vlasiator">
<h3>Practical aspects of running Vlasiator<a class="headerlink" href="#practical-aspects-of-running-vlasiator" title="Link to this heading"></a></h3>
<p>What sorts on constraints does Vlasiator have?</p>
<p>Six-dimensional simulations are expensive! Further, there are some constraints from physics on resolution - if you want to be accurate, you will need to resolve some scales.</p>
<ol class="arabic">
<li><p>Timestep: Ion cyclotron frequency.</p>
<p>Especially restrictive with global magnetospheric simulations with high magnetic field values close to the poles, with an r^-3 dependence.</p>
<p>We constrain the timestep to provide a maximum of 22 degrees of cyclotron rotation per step. Further, the Vlasov solver timestep is <em>so far</em> a global timestep. This is a bottleneck for the inner boundary.</p>
</li>
<li><p>Velocity-space resolution: numerical heating</p>
<p>Low velocity-space resolution is constrained by thermal velocity: dv ~&lt; v_th/3. For exampe, solar wind Maxwellians exhibit numerical heating for dv &gt;~ v_th/3.</p>
</li>
<li><p>Velocity-space sparsity threshold</p>
<p>Around 10^-15 m^6 s^-3 has been found a decent compromise for the sparsity threshold, retaining the parts of VDFs that are relevant for the dynamics of the system. If you wish to model high-energy tails, the threshold should be pushed down… at the cost of significantly increased memory footprint.</p>
</li>
<li><p>Velocity-space extents</p>
<p>+- 4000…8000 km/s are decent extents for terrestrial magnetospheric simulations with the current sparsity thresholds. Usually VDFs ‘hit the walls’ only when there is anomalous acceleration, or otherwise very fast and/or hot flows. This results in 200..400 velocity cells per dimension in our usual runs. Sparsity really saves on the memory footprint!</p>
</li>
<li><p>Spatial resolution</p>
<p>Preferably, ion kinetic scale should be resolved (~200 km in Earth’s magnetosphere). In 6D global production runs, we have not yet reached this (max resolution 1000 km). 5D runs can go to sufficient resolution - plenty of foreshock studies!</p>
<p>Anomalous effects can be seen in the magnetosheath, for example. 1000 km resolution does not resolve all relevant EMIC waves, leading to anomalous loss-cone distributions. See e.g. <a class="reference external" href="https://doi.org/10.1063/5.0176376">Dubart+2023</a> for ongoing work towards a subgrid diffusion model to mimic the effects of these waves.</p>
</li>
<li><p>Spatial extents</p>
<p>Outflow boundaries are simple copy-condition boundaries. For example, the bow shock hitting the outer walls should be avoided, leading to approx. +- 60 RE simulation extents for the Earth.</p>
</li>
</ol>
</section>
<section id="what-can-i-do">
<h3>What can I do?<a class="headerlink" href="#what-can-i-do" title="Link to this heading"></a></h3>
<p>Let’s see what <em>has</em> been done.</p>
<ul>
<li><p>ULF wave field studies, see <a class="reference external" href="https://doi.org/10.1038/s41567-022-01837-z">Turc et al., 2022</a>, Transmission of foreshock waves through Earth’s bow shock. <em>Nat. Phys</em>.</p>
<p>5D/2D runs - can resolve ion kinetic scales. Topics include generation of counterstreaming foreshock populations, resulting ULF wave fields, and their effects!</p>
<p>Similar runs are now being run at O(10M) CPUh on Mahti (CSC/Finland). Smaller 2D runs have been nowaways performed also on our local cluster of 10 nodes of 2x AMD Epyc 7302/32 cores per node.</p>
</li>
<li><p>Relatedly: Jets and SLAMS, see the works by Joonas Suni, e.g., <a class="reference external" href="https://doi.org/10.1029/2021GL095655">Suni et al.,</a></p></li>
<li><p>Reconnection in a global context <a class="reference external" href="https://doi.org/10.1038/s41561-023-01206-2">Palmroth et al. 2023</a>, Magnetotail plasma eruptions driven by magnetic reconnection and kinetic instabilities. <em>Nat. Geosci</em>.</p>
<p>These are huge runs at several tens of MCPU-hours.</p>
<p>Ion kinetics make the reconnection phenomena very dynamic, presenting need for more tools for finding them. See <a class="reference external" href="https://doi.org/10.5194/egusphere-2023-2300">Alho et al., 2023 (revised)</a></p>
</li>
<li><p>Directly utilize the VDF data of the global simulations, e.g. in precipitation of particles from magnetospheric processes, both from 2D (<a class="reference external" href="https://doi.org/10.5194/angeo-37-791-2019">Grandin et al. 2019</a>) and 3D simulations (<cite>Grandin et al. 2023 &lt;https://doi.org/10.1051/swsc/2023017&gt;</cite>)</p></li>
<li><p>Local ion-kinetic simulations, e.g. Kelvin-Helmholz instabilities (Tarvus et al, in prep.), EMIC wave studies (<a class="reference external" href="https://doi.org/10.1063/5.0176376">Dubart et al. 2023</a>, with <em>many</em> small simulations), shock tube simulations (YPK)… More discussion on Vlasiator projects later.</p></li>
</ul>
<p>See the <a class="reference external" href="https://www.helsinki.fi/en/researchgroups/vlasiator/publications-and-presentations">Vlasiator website/publications</a> for more examples!</p>
<p>We note that since these runs are expensive, we often mine multiple papers out of a single run.</p>
</section>
</section>
<section id="rules-of-the-road">
<h2>Rules of the Road<a class="headerlink" href="#rules-of-the-road" title="Link to this heading"></a></h2>
<p>As noted at the <a class="reference external" href="https://www.helsinki.fi/en/researchgroups/vlasiator/rules-of-the-road">Vlasiator website</a>.</p>
<p><em>Vlasiator is a significant investment, and its development has taken many years (since 2008).</em></p>
<p>Vlasiator is funded by the European Research Council and the Research Council of Finland.</p>
<p>Vlasiator requires considerable amount of human resources. The code is being developed by external funding. Runs are carried out with supercomputer resources that the PI team applies for from competitive sources (such as PRACE and CSC Grand Challenge). Each run needs to be babysat.</p>
<p>While Vlasiator is licensed under GPL-2, these rules of the road have a similar philosophy as are generally in use for instruments. Vlasiator is the only one of its kind in the world, having no similar benchmark. Hence we need to make sure that all Vlasiator results have been verified by by the PI-team.</p>
<p>The user is strongly advised to utilise Vlasiator as described below.</p>
<section id="pi-and-the-pi-team">
<h3>PI and the PI-team<a class="headerlink" href="#pi-and-the-pi-team" title="Link to this heading"></a></h3>
<p>The PI-team makes all decisions pertaining to the Vlasiator master version. All data requests and other support questions should be addressed to the PI. The PI-team decides about the time and place in which the peer-reviewed data becomes public.</p>
</section>
<section id="vlasiator-enthusiasts">
<h3>Vlasiator enthusiasts<a class="headerlink" href="#vlasiator-enthusiasts" title="Link to this heading"></a></h3>
<p>The PI-team welcomes collaborations! Do reach out to us and make a data request, which we handle on the best effort basis. Any publications or presentations need to follow the publication rules below, and the further distribution of the accessed data is not allowed without the consent of the PI-team. The movies made public through the Vlasiator web pages can be freely used in scientific work, presentations and publications, bearing in mind the publications rules of the road below.</p>
</section>
<section id="publications-and-presentations">
<h3>Publications and presentations<a class="headerlink" href="#publications-and-presentations" title="Link to this heading"></a></h3>
<p>All publications or presentations showing Vlasiator data need to be inspected by someone from the Vlasiator PI-team. The PI and the relevant PI-team members shall be added as co-authors in the publication. The Vlasiator PI-team may publish the Vlasiator data shown in the publication through the Vlasiator web page.</p>
<section id="acknowledgement-for-publications-and-presentations">
<h4>Acknowledgement for publications and presentations<a class="headerlink" href="#acknowledgement-for-publications-and-presentations" title="Link to this heading"></a></h4>
<p>Vlasiator is developed by the European Research Council Starting grant 200141-QuESpace, and Consolidator grant GA682068-PRESTISSIMO received by the Vlasiator PI. Vlasiator has also received funding from the Academy of Finland. See www.helsinki.fi/vlasiator</p>
</section>
<section id="presentations">
<h4>Presentations<a class="headerlink" href="#presentations" title="Link to this heading"></a></h4>
<p>All presentations showing Vlasiator data shall include a Vlasiator slide given by the PI-team. The presentation shall also include the Vlasiator logo. The slide and acknowledgement are available from the PI time upon request.</p>
</section>
<section id="publications">
<h4>Publications<a class="headerlink" href="#publications" title="Link to this heading"></a></h4>
<p>All publications showing Vlasiator data shall cite these two Vlasiator architectural papers:</p>
<p>Enabling technology for global 3D+3V hybrid-Vlasov simulations of near Earth space, U. Ganse, T. Koskela, M. Battarbee, Y. Pfau-Kempf, K. Papadakis, M. Alho, M. Bussov, G. Cozzani, M. Dubart, H. George, E. Gordeev, M. Grandin, K. Horaites, J. Suni,
V. Tarvus, F. Tesema Kebede, L. Turc, H Zhou, and M. Palmroth. Physics of Plasmas 30, 042902 (2023)
<a class="reference external" href="https://doi.org/10.1063/5.0134387">https://doi.org/10.1063/5.0134387</a></p>
<p>Palmroth, M., Ganse, U., Pfau-Kempf, Y., Battarbee, M., Turc, L., Brito, T., Grandin, M., Hoilijoki, S., Sandroos, A., and von Alfthan, S., “Vlasov methods in space physics and astrophysics”, Living Rev Comput Astrophys., 4:1, <a class="reference external" href="doi:10.1007/s41115-018-0003-2">doi:10.1007/s41115-018-0003-2</a>, 2018.</p>
</section>
<section id="additional-informative-technological-publications">
<h4>Additional informative technological publications:<a class="headerlink" href="#additional-informative-technological-publications" title="Link to this heading"></a></h4>
<p>Parametrization of coefficients for sub-grid modeling of pitch-angle diffusion in global magnetospheric hybrid-Vlasov simulations, M. Dubart, M. Battarbee, U. Ganse, A. Osmane, F. Spanier, J. Suni, G. Cozzani, K. Horaites, K. Papadakis, Y. Pfau-Kempf, V. Tarvus, and M. Palmroth, Physics of Plasmas 30, 123903 (2023)
<a class="reference external" href="https://doi.org/10.1063/5.0176376">https://doi.org/10.1063/5.0176376</a></p>
<p>Spatial filtering in a 6D hybrid-Vlasov scheme for alleviating AMR artifacts: a case study with Vlasiator, versions 5.0, 5.1, 5.2.1, K. Papadakis, Y. Pfau-Kempf, U. Ganse, M. Battarbee, M. Alho, M. Grandin, M. Dubart, L. Turc, H. Zhou, K. Horaites, I. Zaitsev, G. Cozzani, M. Bussov, E. Gordeev, F. Tesema, H. George, J. Suni, V. Tarvus, and M. Palmroth. Geosci. Model Dev., 15, 7903–7912 (2022)
<a class="reference external" href="https://doi.org/10.5194/gmd-15-7903-2022">https://doi.org/10.5194/gmd-15-7903-2022</a></p>
<p>Vlasov simulation of electrons in the context of hybrid global models: an eVlasiator approach, M. Battarbee, T. Brito, M. Alho, Y. Pfau-Kempf, M. Grandin, U. ganse, K. Papadakis, A. Johlander, L. Turc, M. Dubart, and M. Palmroth. Ann. Geophys. 39, 85–103 (2021)
<a class="reference external" href="https://doi.org/10.5194/angeo-39-85-2021">https://doi.org/10.5194/angeo-39-85-2021</a></p>
<p>Vlasiator: First global hybrid-Vlasov simulations of Earth’s foreshock and magnetosheath, S. von Alfthan, D. Pokhotelov, Y. Kempf, S. Hoilijoki, I. Honkonen, A. Sandroos, M. Palmroth. Journal of Atmospheric and Solar-Terrestrial Physics, Volume 120, December 2014, Pages 24-35, <a class="reference external" href="https://doi.org/10.1016/j.jastp.2014.08.012">https://doi.org/10.1016/j.jastp.2014.08.012</a></p>
</section>
</section>
<section id="interesting-questions-you-might-get">
<h3>Interesting questions you might get<a class="headerlink" href="#interesting-questions-you-might-get" title="Link to this heading"></a></h3>
</section>
<section id="typical-pitfalls">
<h3>Typical pitfalls<a class="headerlink" href="#typical-pitfalls" title="Link to this heading"></a></h3>
</section>
</section>
<span id="document-prepare_workspace"></span><section id="preparations-for-the-workshop">
<h2>Preparations for the workshop<a class="headerlink" href="#preparations-for-the-workshop" title="Link to this heading"></a></h2>
<ol class="arabic simple">
<li><p>Watch the <a href="#id1"><span class="problematic" id="id2">`introductory lectures&lt;https://datacloud.helsinki.fi/index.php/s/wEZdF3szjBfapSs&gt;`_</span></a> about Vlasiator.</p></li>
<li><p>Install locally <a class="reference external" href="https://visit-dav.github.io/visit-website/releases-as-tables/">VisIt 3.3.3</a></p></li>
<li><p>Get LUMI access (information provided via ENCCS)</p></li>
<li><p>Verify ssh key access to LUMI (makes everything easier)</p></li>
<li><p>Verify GitHub ssh access from LUMI</p></li>
<li><p>Clone <a class="reference external" href="https://github.com/fmihpc/analysator">Analysator</a> onto your LUMI workspace and set it up – see <em>Introduction to Analysator</em></p></li>
</ol>
<p>Optional:</p>
<ol class="arabic simple">
<li><p>Clone <a class="reference external" href="https://github.com/fmihpc/vlasiator">Vlasiator</a> and its prerequisite libraries on LUMI (see <em>Installing Vlasiator</em>)</p></li>
</ol>
</section>
</div>
<div class="toctree-wrapper compound">
<span id="document-installing-vlasiator"></span><section id="installing-vlasiator">
<span id="id1"></span><h2>Installing Vlasiator<a class="headerlink" href="#installing-vlasiator" title="Link to this heading"></a></h2>
<section id="why-we-teach-this-lesson">
<h3>Why we teach this lesson<a class="headerlink" href="#why-we-teach-this-lesson" title="Link to this heading"></a></h3>
<p>Vlasiator is hosted on GitHub and is open source, but it is, still, a specialist code under development. Here we show how to obtain the up-to-date stable Vlasiator version with the requisite libraries, and go through what the libraries are and what are they used for.</p>
</section>
<section id="intended-learning-outcomes">
<h3>Intended learning outcomes<a class="headerlink" href="#intended-learning-outcomes" title="Link to this heading"></a></h3>
<p>You can install a correct version of Vlasiator, required libraries and build everything.</p>
</section>
<section id="timing">
<h3>Timing<a class="headerlink" href="#timing" title="Link to this heading"></a></h3>
<p>Tuesday pre-lunch.</p>
</section>
<section id="how-to-install-vlasiator">
<h3>How to install Vlasiator<a class="headerlink" href="#how-to-install-vlasiator" title="Link to this heading"></a></h3>
<p>Installing Vlasiator is easy and straightforward!</p>
<p>Tasks:</p>
<ol class="arabic simple">
<li><p>Clone Vlasiator <em>with submodule support</em></p></li>
<li><p>Build libraries</p></li>
<li><p>Make new makefile for your machine in MAKE folder</p></li>
<li><p>Compile!</p></li>
</ol>
<p>Here are some general steps. More machine-specific details may be detailed on one of the following pages:</p>
<section id="cloning-vlasiator">
<h4>Cloning Vlasiator<a class="headerlink" href="#cloning-vlasiator" title="Link to this heading"></a></h4>
<p>We are transferring to use <code class="docutils literal notranslate"><span class="pre">git</span> <span class="pre">submodules</span></code> for the dependent libraries. So far, some of the header libraries have been moved to this framework, and some need to be installed manually (see above).</p>
<p>Use the <code class="docutils literal notranslate"><span class="pre">--recurse-submodules</span></code> when cloning, pulling, or checking out branches:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>git<span class="w"> </span>clone<span class="w"> </span>--recurse-submodules<span class="w"> </span>https://github.com/fmihpc/Vlasiator
git<span class="w"> </span>checkout<span class="w"> </span>master<span class="w"> </span>--recurse-submodules
git<span class="w"> </span>submodule<span class="w"> </span>update<span class="w"> </span>--init<span class="w"> </span>--recursive
</pre></div>
</div>
<p>Task: create a folder with your username under <code class="docutils literal notranslate"><span class="pre">/projappl/project_465000693</span></code>, <code class="docutils literal notranslate"><span class="pre">cd</span></code> into it and clone Vlasiator master branch into it.</p>
<section id="vlasiator-main-branches">
<h5>Vlasiator main branches<a class="headerlink" href="#vlasiator-main-branches" title="Link to this heading"></a></h5>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">master</span></code>: Main branch, stable, lagging behind from development branch. Main releases.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">dev</span></code>: Latest features, but potentially not stable.</p></li>
</ul>
<p>… plus a plethora of topic branches.</p>
</section>
</section>
</section>
<section id="building-libraries">
<h3>Building libraries<a class="headerlink" href="#building-libraries" title="Link to this heading"></a></h3>
<p>Vlasiator needs a number of libraries, a part of which need to be built. Some header libraries have been transferred to submodules, and those are automatically fetched with git (… when <code class="docutils literal notranslate"><span class="pre">--recurse-submodules</span></code> is used correctly!).</p>
<p>When building libraries and the code, we want to stick to a particular toolchain of compilers and MPI libraries, etc. On LUMI, we use the following modules:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>module<span class="w"> </span>load<span class="w"> </span>LUMI/22.08
module<span class="w"> </span>load<span class="w"> </span>cpeGNU
module<span class="w"> </span>load<span class="w"> </span>papi
module<span class="w"> </span>load<span class="w"> </span>Eigen
module<span class="w"> </span>load<span class="w"> </span>Boost/1.79.0-cpeGNU-22.08
</pre></div>
</div>
<p>Each cluster and supercomputer will have different modules available. If the prerequisite libraries are not available as modules, they need to be downloaded and built by the user. On debian-based system (such as ubuntu and cubbli), some of the dependencies are provided as packages, installable via <code class="docutils literal notranslate"><span class="pre">apt-get</span> <span class="pre">install</span> <span class="pre">libeigen3-dev</span> <span class="pre">libboost-dev</span> <span class="pre">libboost-program-options-dev</span> <span class="pre">libopenmpi-dev</span></code>. Use of the <a class="reference external" href="https://launchpad.net/~boost-latest/+archive/ppa">boost-latest ppa</a> is recommended on Ubuntu.</p>
<p>Building the prerequisite libraries of Vlasiator can be done with the following script, included in the Vlasiator repository: <a class="reference external" href="https://github.com/fmihpc/vlasiator/blob/master/build_libraries.sh">build_libraries.sh</a>. Our usual practice is to use a centralized library folder, but we’ll set up one for each user as an exercise.</p>
<p>Tasks:</p>
<ol class="arabic simple">
<li><p>copy <code class="docutils literal notranslate"><span class="pre">build_libraries.sh</span></code> from Vlasiator root to <code class="docutils literal notranslate"><span class="pre">projappl/project_465000693/&lt;user&gt;</span></code>.</p></li>
<li><p>load the above toolchain with the module load commands.</p></li>
<li><p>Build the libraries with a descriptive name for the toolchain: <code class="docutils literal notranslate"><span class="pre">./build_libraries.sh</span> <span class="pre">LUMI-22.08-GNU-PEPSC</span></code></p></li>
<li><p>Find the built libraries then under <code class="docutils literal notranslate"><span class="pre">libraries-LUMI-22.08-GNU-PEPSC/</span></code>. We’ll use this path for our Makefile.</p></li>
</ol>
<p>That’s done! Below, the libraries used for reference.</p>
<section id="library-reference">
<h4>Library reference<a class="headerlink" href="#library-reference" title="Link to this heading"></a></h4>
<section id="require-building">
<h5>Require building<a class="headerlink" href="#require-building" title="Link to this heading"></a></h5>
<ul class="simple">
<li><p><a class="reference external" href="http://www.cs.sandia.gov/zoltan/">Zoltan</a> (<a class="reference external" href="https://github.com/fmihpc/Vlasiator/wiki/Installing-Vlasiator#zoltan">install instructions</a>)</p>
<ul>
<li><p>Load balancing library.</p></li>
</ul>
</li>
<li><p><a class="reference external" href="http://www.boost.org/">Boost</a> (<a class="reference external" href="https://github.com/fmihpc/Vlasiator/wiki/Installing-Vlasiator#boost">install instructions</a>)</p>
<ul>
<li><p>Configuration parser.</p></li>
</ul>
</li>
<li><p><a class="reference external" href="http://eigen.tuxfamily.org/index.php?title=Main_Page">Eigen</a> (<a class="reference external" href="https://github.com/fmihpc/Vlasiator/wiki/Installing-Vlasiator#eigen">install instructions</a>)</p>
<ul>
<li><p>Linear algebra</p></li>
</ul>
</li>
<li><p><a class="reference external" href="https://github.com/fmihpc/phiprof">Phiprof</a> (<a class="reference external" href="https://github.com/fmihpc/Vlasiator/wiki/Installing-Vlasiator#phiprof">install instructions</a>)</p>
<ul>
<li><p>Lightweight profiling.</p></li>
</ul>
</li>
<li><p><a class="reference external" href="https://github.com/fmihpc/vlsv">VLSV</a> (<a class="reference external" href="https://github.com/fmihpc/Vlasiator/wiki/Installing-Vlasiator#vlsv">install instructions</a>)</p>
<ul>
<li><p>Custom file format library, with parallel MPI I/O support.</p></li>
<li><p>NB: Contains the buildable VLSV plugin for VisIt.</p></li>
</ul>
</li>
<li><p>MPI</p></li>
<li><p>C++17 compiler with OpenMP &gt;=3 support</p></li>
</ul>
</section>
<section id="header-libraries-fetched-via-submodules">
<h5>Header libraries fetched via submodules<a class="headerlink" href="#header-libraries-fetched-via-submodules" title="Link to this heading"></a></h5>
<p>These libraries are handled via <code class="docutils literal notranslate"><span class="pre">git</span> <span class="pre">submodules</span></code> (nb. clone/pull instructions for submodules below), you do not need to install these separately.</p>
<ul class="simple">
<li><p><a class="reference external" href="https://github.com/fmihpc/dccrg">DCCRG</a> (<a class="reference external" href="https://github.com/fmihpc/Vlasiator/wiki/Installing-Vlasiator#dccrg">install instructions</a>)</p>
<ul>
<li><p>Generic MPI grid library used for the Vlasov solver grid with AMR.</p></li>
<li><p>DCCRG has its own prerequisites (MPI 2, Zoltan, and Boost). See the linked install instructions for required libraries!</p></li>
</ul>
</li>
<li><p><a class="reference external" href="https://github.com/fmihpc/fsgrid">FsGrid</a> (<a class="reference external" href="https://github.com/fmihpc/Vlasiator/wiki/Installing-Vlasiator#fsgrid">install instructions</a>)</p>
<ul>
<li><p>Lightweight parallel grid library used for the uniform field solver grid.</p></li>
</ul>
</li>
<li><p><a class="reference external" href="http://www.agner.org/optimize/#vectorclass">Vectorclass</a> (<a class="reference external" href="https://github.com/fmihpc/Vlasiator/wiki/Installing-Vlasiator#vectorclass">install instructions</a>)</p>
<ul>
<li><p>SIMD support</p></li>
<li><p>See instructions for the required addon library if installing manually.</p></li>
</ul>
</li>
</ul>
</section>
<section id="optional-libraries">
<h5>Optional libraries<a class="headerlink" href="#optional-libraries" title="Link to this heading"></a></h5>
<p>And also a number of optional but useful libraries</p>
<ul class="simple">
<li><p><a class="reference external" href="www.canonware.com/jemalloc/download.html">Jemalloc</a> (<a class="reference external" href="https://github.com/fmihpc/Vlasiator/wiki/Installing-Vlasiator#jemalloc">install instructions</a>)</p>
<ul>
<li><p>Memory allocator with reduced memory fragmentation (recommended for performance)</p></li>
</ul>
</li>
<li><p><a class="reference external" href="http://icl.cs.utk.edu/papi/">Papi</a> (<a class="reference external" href="https://github.com/fmihpc/Vlasiator/wiki/Installing-Vlasiator#papi">install instructions</a>)</p>
<ul>
<li><p>Memory measurement, module often available on-site</p></li>
</ul>
</li>
</ul>
</section>
</section>
<section id="make-a-new-makefile">
<h4>Make a new makefile<a class="headerlink" href="#make-a-new-makefile" title="Link to this heading"></a></h4>
<p>The main makefile is in the vlasiator main folder. There should be no need to modify that. All settings are in a separate machine specific file that is in the MAKE folder, where compiler names, compiler flags and library locations are set. In the MAKE folder there are several examples from various machines. The file name is <code class="docutils literal notranslate"><span class="pre">Makefile.machine_name</span></code>, where machine_name is whatever you want to call your machine. It is best to start from a makefile that is similar to the machine you are compiling on. The Makefile.home corresponds to a Linux computer with all libraries in <code class="docutils literal notranslate"><span class="pre">${HOME}/lib</span></code> and <code class="docutils literal notranslate"><span class="pre">${HOME}/include</span></code>.</p>
<p>We’ll do a new Makefile based on <a class="reference external" href="shared_files/Makefile.lumi-template">a template</a>.</p>
<p>Firstly, note that mark, as comments, the module toolchain that we use with this Makefile:</p>
<div class="highlight-makefile notranslate"><div class="highlight"><pre><span></span><span class="c"># Modules loaded</span>
<span class="c"># module load LUMI/22.08 ; module load cpeGNU ; module load papi; module load Eigen; module load Boost/1.79.0-cpeGNU-22.08</span>
</pre></div>
</div>
<p>These will need to be loaded while compiling and running, and need to match your library toolchain.</p>
<p>Find the LIBRARY_PREFIX variables and modify them to match your library paths:</p>
<div class="highlight-makefile notranslate"><div class="highlight"><pre><span></span><span class="nv">LIBRARY_PREFIX</span><span class="w"> </span><span class="o">=</span><span class="w"> </span>&lt;library-dir/lib&gt;
<span class="nv">LIBRARY_PREFIX_HEADERS</span><span class="w"> </span><span class="o">=</span><span class="w"> </span>&lt;library-dir/include&gt;
</pre></div>
</div>
<p>This is enough! But note how these are used later, for example:</p>
<div class="highlight-make notranslate"><div class="highlight"><pre><span></span><span class="nv">INC_ZOLTAN</span><span class="w"> </span><span class="o">=</span><span class="w"> </span>-isystem<span class="k">$(</span>LIBRARY_PREFIX_HEADERS<span class="k">)</span>
<span class="nv">LIB_ZOLTAN</span><span class="w"> </span><span class="o">=</span><span class="w"> </span>-L<span class="k">$(</span>LIBRARY_PREFIX<span class="k">)</span><span class="w"> </span>-lzoltan<span class="w"> </span>-Wl,-rpath<span class="o">=</span><span class="k">$(</span>LIBRARY_PREFIX<span class="k">)</span>
</pre></div>
</div>
<p>If you wish, you can choose to point to different libraries via modifying these paths.</p>
</section>
<section id="compile">
<h4>Compile!<a class="headerlink" href="#compile" title="Link to this heading"></a></h4>
<p>After one has created the makefile, one should set an environment variable with the name of your machine, matching the name used for the MAKE/Makefile.machine_name file. For example, to use the home makefile one can set it like this:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">export</span><span class="w"> </span><span class="nv">VLASIATOR_ARCH</span><span class="o">=</span>home
</pre></div>
</div>
<p>To make the environment variable one can put it into the initialization files for your shell, e.g. .profile. or .bashrc.</p>
<p>After ensuring all libraries and compile options are made available for Vlasiator, and the correct machine-specific makefile has been set, one can simply</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>make<span class="w"> </span>clean
make<span class="w"> </span>-j<span class="w"> </span><span class="m">12</span>
</pre></div>
</div>
<p>to make Vlasiator, or</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>make<span class="w"> </span>clean
make<span class="w"> </span>-j<span class="w"> </span><span class="m">12</span><span class="w"> </span>tools
</pre></div>
</div>
<p>to make the Vlasiator tools.</p>
<p>Note: The -j flag tells GNU Make to build the program in parallel on several threads. If you are building on a smaller computer, it is not recommended to have a -j count greater than the number of available cores on the frontend where you are compiling. This will not impact how many threads the actual simulation will run on.</p>
</section>
</section>
<section id="detailed-installation-instructions-for-libraries">
<h3>Detailed installation instructions for Libraries<a class="headerlink" href="#detailed-installation-instructions-for-libraries" title="Link to this heading"></a></h3>
<p>If the install script or fetching submodules fails, you can review the more in-depth guidelines available at <a class="reference external" href="https://github.com/fmihpc/vlasiator/wiki/Installing-Vlasiator">https://github.com/fmihpc/vlasiator/wiki/Installing-Vlasiator</a> though it should not be necessary for the purposes of this tutorial.</p>
</section>
<section id="other-practical-aspects">
<h3>Other practical aspects<a class="headerlink" href="#other-practical-aspects" title="Link to this heading"></a></h3>
</section>
<section id="interesting-questions-you-might-get">
<h3>Interesting questions you might get<a class="headerlink" href="#interesting-questions-you-might-get" title="Link to this heading"></a></h3>
</section>
<section id="typical-pitfalls">
<h3>Typical pitfalls<a class="headerlink" href="#typical-pitfalls" title="Link to this heading"></a></h3>
<p>Some wise words of the pitfalls of submodules and git commands:
So trying with a fresh clone with <strong>no</strong> <code class="docutils literal notranslate"><span class="pre">--recurse-submodules</span></code>, this gets the correct vlasiator-version target for dccrg:</p>
<p><code class="docutils literal notranslate"><span class="pre">git</span> <span class="pre">checkout</span> <span class="pre">dev</span></code>
<code class="docutils literal notranslate"><span class="pre">git</span> <span class="pre">pull</span> <span class="pre">origin</span> <span class="pre">dev</span> <span class="pre">--recurse-submodules</span></code></p>
<p>This works as well</p>
<p><code class="docutils literal notranslate"><span class="pre">git</span> <span class="pre">checkout</span> <span class="pre">dev</span> <span class="pre">--recurse-submodules</span></code>
<code class="docutils literal notranslate"><span class="pre">git</span> <span class="pre">submodule</span> <span class="pre">update</span> <span class="pre">--init</span> <span class="pre">--recursive</span></code></p>
<p>This however does not fetch the correct submodule commits:</p>
<p><code class="docutils literal notranslate"><span class="pre">git</span> <span class="pre">checkout</span> <span class="pre">dev</span></code></p>
<p>This does not fetch submodules by itself:</p>
<p><code class="docutils literal notranslate"><span class="pre">git</span> <span class="pre">checkout</span> <span class="pre">dev</span> <span class="pre">--recurse-submodules</span></code></p>
<p>but it needs then</p>
<p><code class="docutils literal notranslate"><span class="pre">git</span> <span class="pre">submodule</span> <span class="pre">update</span> <span class="pre">--init</span> <span class="pre">--recursive</span></code></p>
<p>But,</p>
<p><code class="docutils literal notranslate"><span class="pre">git</span> <span class="pre">checkout</span> <span class="pre">dev</span></code>
<code class="docutils literal notranslate"><span class="pre">git</span> <span class="pre">submodule</span> <span class="pre">update</span> <span class="pre">--init</span> <span class="pre">--recursive</span></code></p>
<p>is bad, since that will get the default master branch tip as the submodule commits and then updates the submodules to those ones. But then, if you start with</p>
<p><code class="docutils literal notranslate"><span class="pre">git</span> <span class="pre">clone</span> <span class="pre">--recurse-submodules</span> <span class="pre">https://github.com/fmihpc/vlasiator</span></code></p>
<p>you can do</p>
<p><code class="docutils literal notranslate"><span class="pre">git</span> <span class="pre">checkout</span> <span class="pre">dev</span></code>
<code class="docutils literal notranslate"><span class="pre">git</span> <span class="pre">submodule</span> <span class="pre">update</span> <span class="pre">--init</span> <span class="pre">--recursive</span></code></p>
</section>
</section>
<span id="document-benchmarks_and_running"></span><section id="how-to-configure-vlasiator-for-performance">
<h2>How to configure Vlasiator for performance<a class="headerlink" href="#how-to-configure-vlasiator-for-performance" title="Link to this heading"></a></h2>
<section id="why-we-teach-this-lesson">
<h3>Why we teach this lesson<a class="headerlink" href="#why-we-teach-this-lesson" title="Link to this heading"></a></h3>
<p>Vlasiator requires a lot of resources to run, but thankfully it scales pretty well (super-linearly, even!). Good scaling requires, however, understanding of the system in use, and knowing how to profile and optimize the HPC environment from balancing threads vs tasks to parallel I/O. Here we introduce built-in Vlasiator profiling tools, rules-of-thumb for setting up the Slurm batch jobs, LUSTRE striping and ROMIO flags relevant to LUMI.</p>
</section>
<section id="intended-learning-outcomes">
<h3>Intended learning outcomes<a class="headerlink" href="#intended-learning-outcomes" title="Link to this heading"></a></h3>
<p>The user understands how to run Vlasiator, how to scale it and the options to affect scaling properties and can benchmark scaling.</p>
</section>
<section id="timing">
<h3>Timing<a class="headerlink" href="#timing" title="Link to this heading"></a></h3>
<p>Tue afternoon.</p>
</section>
<section id="preparing-exercises">
<h3>Preparing exercises<a class="headerlink" href="#preparing-exercises" title="Link to this heading"></a></h3>
<p>Scaling exercise: few configurations to set up. Template file is decent, template jobscripts exist.</p>
</section>
<section id="running-vlasiator">
<h3>Running Vlasiator<a class="headerlink" href="#running-vlasiator" title="Link to this heading"></a></h3>
<p>The recipe for a Vlasiator run. We use a benchmarking run for this exercise.</p>
<ol class="arabic simple">
<li><p>Folder to run the simulation. Use scratch for Lustre, our training project scratch is in <code class="docutils literal notranslate"><span class="pre">/scratch/project_465000693/</span></code> - make yourself a folder there with your username for your use.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">vlasiator</span></code> executable.</p></li>
</ol>
<blockquote>
<div><ul class="simple">
<li><p>These store versioning information, including git hashes and possible diffs. Probably a good idea to copy the executable to your run folder!</p></li>
</ul>
</div></blockquote>
<ol class="arabic simple">
<li><p>A Vlasiator configuration file. We will inspect our scaling test config soon, copy it from <code class="docutils literal notranslate"><span class="pre">/scratch/project_465000693/example_runs/scaling/baseline/Flowthrough_amr.cfg</span></code></p></li>
<li><p>Auxiliary input files for Vlasiator.</p></li>
</ol>
<blockquote>
<div><ul class="simple">
<li><p>In this case, grab also the <code class="docutils literal notranslate"><span class="pre">sw1.dat</span></code> file from the above folder. This one defines inflow plasma parameters.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">NRLMSIS.dat</span></code> is another possibility, for an ionospheric profile for magnetospheric runs with an ionosphere.</p></li>
</ul>
</div></blockquote>
<ol class="arabic simple">
<li><p>A Slurm job script, for defining and requesting the HPC environment. Grab <code class="docutils literal notranslate"><span class="pre">job-debug.sh</span></code> from the above folder.</p></li>
</ol>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="ch">#!/bin/bash -l</span>
<span class="c1">#SBATCH --job-name=PEPSC-demo-debug</span>
<span class="c1">#SBATCH --partition=debug</span>
<span class="c1">#SBATCH --nodes=2</span>
<span class="c1">#SBATCH --mem=0</span>
<span class="c1">#SBATCH --ntasks-per-node=16</span>
<span class="c1">#SBATCH --cpus-per-task=8</span>
<span class="c1">#SBATCH --time=0:30:00</span>
<span class="c1">#SBATCH --account=project_465000693</span>
<span class="c1">##SBATCH --hint=multithread</span>
<span class="c1">#SBATCH --exclusive # enforced on &gt;=standard partitions, not on small</span>
<span class="c1">##SBATCH --dependency=singleton # useful for restarting</span>

date

<span class="nb">export</span><span class="w"> </span><span class="nv">OMP_NUM_THREADS</span><span class="o">=</span><span class="nv">$SLURM_CPUS_PER_TASK</span>
<span class="c1">#export OMP_PLACES=cores</span>
<span class="c1">#export OMP_PROC_BIND=spread</span>
<span class="nb">export</span><span class="w"> </span><span class="nv">SRUN_CPUS_PER_TASK</span><span class="o">=</span><span class="nv">$SLURM_CPUS_PER_TASK</span>

<span class="nb">export</span><span class="w"> </span><span class="nv">TASKS_PER_NODE</span><span class="o">=</span><span class="k">$((</span><span class="w"> </span><span class="m">128</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="nv">$SLURM_CPUS_PER_TASK</span><span class="w"> </span><span class="k">))</span>
<span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;We use &quot;</span><span class="si">${</span><span class="nv">TASKS_PER_NODE</span><span class="si">}</span><span class="s2">&quot; tasks per node&quot;</span>

<span class="c1"># https://docs.olcf.ornl.gov/systems/crusher_quick_start_guide.html</span>
<span class="nb">export</span><span class="w"> </span><span class="nv">MPICH_SMP_SINGLE_COPY_MODE</span><span class="o">=</span>NONE

<span class="nb">ulimit</span><span class="w"> </span>-c<span class="w"> </span>unlimited
<span class="nb">export</span><span class="w"> </span><span class="nv">PHIPROF_PRINTS</span><span class="o">=</span>detailed
<span class="nb">umask</span><span class="w"> </span><span class="m">007</span>

module<span class="w"> </span>--force<span class="w"> </span>purge
module<span class="w"> </span>load<span class="w"> </span>LUMI/22.08
module<span class="w"> </span>load<span class="w"> </span>cpeGNU
module<span class="w"> </span>load<span class="w"> </span>craype-x86-milan
module<span class="w"> </span>load<span class="w"> </span>papi
module<span class="w"> </span>load<span class="w"> </span>Boost
module<span class="w"> </span>load<span class="w"> </span>Eigen

module<span class="w"> </span>list

<span class="nb">cd</span><span class="w"> </span><span class="nv">$SLURM_SUBMIT_DIR</span>

squeue<span class="w"> </span>--job<span class="w"> </span><span class="nv">$SLURM_JOB_ID</span><span class="w"> </span>-l

sleep<span class="w"> </span><span class="m">5</span>

srun<span class="w"> </span>./vlasiator<span class="w"> </span>--version

sleep<span class="w"> </span><span class="m">5</span>

srun<span class="w"> </span>./vlasiator<span class="w"> </span>--run_config<span class="w"> </span>Flowthrough_amr.cfg<span class="w"> </span><span class="se">\</span>
<span class="w">        </span><span class="c1">#--restart.filename $( ls restart/restart*vlsv | tail -n 1 )</span>
</pre></div>
</div>
<section id="the-vlasiator-configuration-file">
<h4>The Vlasiator Configuration file<a class="headerlink" href="#the-vlasiator-configuration-file" title="Link to this heading"></a></h4>
<p>The configuration file uses the Boost program_options library. You can extract all available options by running <code class="docutils literal notranslate"><span class="pre">./vlasiator</span> <span class="pre">--help</span></code>. There’s plenty, so you may want to pipe that into a text file… but we have done that for you, see the Vlasiator cfg reference!</p>
<p>Let’s inspect the benchmark case config, here:</p>
<div class="highlight-cfg notranslate"><div class="highlight"><pre><span></span><span class="c1"># The root options describe toplevel solver properties and populations</span>
<span class="na">ParticlePopulations</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">proton</span>

<span class="na">project</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">Flowthrough</span>
<span class="na">propagate_field</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">1</span>
<span class="na">propagate_vlasov_acceleration</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">1</span>
<span class="na">propagate_vlasov_translation</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">1</span>
<span class="na">dynamic_timestep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">1</span>

<span class="c1"># &lt;population&gt;_properties, multiple populations are supported!</span>
<span class="k">[proton_properties]</span>
<span class="na">mass</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">1</span>
<span class="na">mass_units</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">PROTON</span>
<span class="na">charge</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">1</span>

<span class="c1"># Adaptive Mesh Refinement - new development!</span>
<span class="k">[AMR]</span>
<span class="na">max_spatial_level</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">2</span><span class="w">             </span><span class="c1"># Maximum number of refinements</span>
<span class="na">max_allowed_spatial_level</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">2</span><span class="w">     </span><span class="c1"># *Currently* allowed number of refinements</span>
<span class="w">                                  </span><span class="c1"># - can be increased after restarts, gradually!</span>
<span class="na">adapt_refinement</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">1</span>

<span class="c1"># tuning parameters</span>
<span class="na">use_alpha1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">1</span>
<span class="na">alpha1_coarsen_threshold</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">0.1</span>
<span class="na">alpha1_refine_threshold</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">0.4</span>
<span class="na">use_alpha2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">1</span>
<span class="na">alpha2_refine_threshold</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">1</span>

<span class="c1"># Cadence and safeties</span>
<span class="na">refine_on_restart</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">0</span>
<span class="na">refine_cadence</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">4</span><span class="w">            </span><span class="c1"># refinement cadence is in units of load balances</span>
<span class="na">refine_after</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">50.0</span><span class="w">           </span><span class="c1"># First adapt after 50s from start of run</span>
<span class="na">refine_radius</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">7320e6</span><span class="w">        </span><span class="c1"># We do not want to refine outer boundary cells</span>

<span class="c1"># Spatial grid parameters - coarsest level grid.</span>
<span class="k">[gridbuilder]</span>
<span class="na">x_length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">48</span>
<span class="na">y_length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">12</span>
<span class="na">z_length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">12</span>
<span class="na">x_min</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">-16e7</span>
<span class="na">x_max</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">16e7</span>
<span class="na">y_min</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">-4e7</span>
<span class="na">y_max</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">4e7</span>
<span class="na">z_min</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">-4e7</span>
<span class="na">z_max</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">4e7</span>
<span class="na">t_max</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">400.0</span>
<span class="na">dt</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">2.0</span>

<span class="c1"># &lt;population&gt;_vspace - velocity grid definitions per population</span>
<span class="k">[proton_vspace]</span>
<span class="na">vx_min</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">-1.92e6</span>
<span class="na">vx_max</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">+1.92e6</span>
<span class="na">vy_min</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">-1.92e6</span>
<span class="na">vy_max</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">+1.92e6</span>
<span class="na">vz_min</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">-1.92e6</span>
<span class="na">vz_max</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">+1.92e6</span>
<span class="c1"># vi_length are in units of block width (4, so far)</span>
<span class="c1"># - these define 64x64x64 velocity-space cells over +-1.92e6 m/s.</span>
<span class="c1"># -&gt; dv = 61.250 km/s for this case</span>
<span class="na">vx_length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">16</span>
<span class="na">vy_length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">16</span>
<span class="na">vz_length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">16</span>

<span class="k">[io]</span>
<span class="na">diagnostic_write_interval</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">1</span>
<span class="na">write_initial_state</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">0</span>

<span class="c1"># Reduced data writeouts</span>
<span class="na">system_write_t_interval</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">10.0</span>
<span class="na">system_write_file_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">bulk</span>

<span class="c1"># These set up strides for saving VDF data in the reduced data readouts</span>
<span class="na">system_write_distribution_stride</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">0</span>
<span class="na">system_write_distribution_xline_stride</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">0</span>
<span class="na">system_write_distribution_yline_stride</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">0</span>
<span class="na">system_write_distribution_zline_stride</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">0</span>

<span class="c1"># Reduced data outputs</span>
<span class="k">[variables]</span>
<span class="na">output</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">populations_vg_rho</span>
<span class="na">output</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">populations_vg_v</span>
<span class="na">output</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">populations_vg_ptensor</span>
<span class="na">output</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">fg_e</span>
<span class="na">output</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">fg_b</span>
<span class="na">output</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">vg_boundarytype</span>
<span class="na">output</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">vg_boundarylayer</span>
<span class="na">output</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">vg_rank</span>
<span class="na">output</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">vg_b_vol_derivatives</span>
<span class="na">output</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">vg_amr_alpha</span>
<span class="na">output</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">vg_amr_jperb</span>
<span class="na">output</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">vg_loadbalance_weight</span>
<span class="na">output</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">populations_vg_blocks</span>
<span class="na">diagnostic</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">populations_vg_blocks</span>

<span class="c1"># Declare boundary conditions</span>
<span class="k">[boundaries]</span>
<span class="na">periodic_x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">no</span>
<span class="na">periodic_y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">yes</span>
<span class="na">periodic_z</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">yes</span>
<span class="na">boundary</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">Outflow</span>
<span class="na">boundary</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">Maxwellian</span>

<span class="k">[outflow]</span>
<span class="na">precedence</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">3</span>

<span class="c1"># NB population-specific boundary conditions!</span>
<span class="k">[proton_outflow]</span>
<span class="na">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">x+</span>
<span class="c1">#face = y-</span>
<span class="c1">#face = y+</span>
<span class="c1">#face = z-</span>
<span class="c1">#face = z+</span>

<span class="k">[maxwellian]</span>
<span class="na">precedence</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">4</span>
<span class="na">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">x-</span>

<span class="k">[proton_maxwellian]</span>
<span class="na">dynamic</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">0</span>
<span class="c1"># select the sw1.dat file for inflow Maxwellian parameters</span>
<span class="na">file_x-</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">sw1.dat</span>

<span class="k">[proton_sparse]</span>
<span class="na">minValue</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">1.0e-15</span>


<span class="c1"># Project settings</span>
<span class="k">[Flowthrough]</span>
<span class="na">Bx</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">1.0e-9</span>
<span class="na">By</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">1.0e-9</span>
<span class="na">Bz</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">1.0e-9</span>

<span class="c1"># Population-specific project settings - initial condition</span>
<span class="k">[proton_Flowthrough]</span>
<span class="na">T</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">1.0e5</span>
<span class="na">rho</span><span class="w">  </span><span class="o">=</span><span class="w"> </span><span class="s">1.0e6</span>
<span class="na">VX0</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">1e5</span>
<span class="na">VY0</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">0</span>
<span class="na">VZ0</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">0</span>

<span class="k">[loadBalance]</span>
<span class="c1"># algorithm = RIB</span>
<span class="na">algorithm</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">RCB</span>
<span class="na">optionKey</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">RCB_RECTILINEAR_BLOCKS</span><span class="w"> </span><span class="c1"># Recommended to use with RCB</span>
<span class="na">optionValue</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">1</span>
<span class="na">rebalanceInterval</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">5</span><span class="w"> </span><span class="c1"># in timesteps</span>

<span class="c1"># Safety bailouts (stores restart for potential recovery)</span>
<span class="k">[bailout]</span>
<span class="na">velocity_space_wall_block_margin</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">0</span>
</pre></div>
</div>
</section>
</section>
<section id="performance-monitoring">
<h3>Performance monitoring<a class="headerlink" href="#performance-monitoring" title="Link to this heading"></a></h3>
<p>Logfile.txt can be used for basic monitoring.</p>
<p><code class="docutils literal notranslate"><span class="pre">phiprof</span></code> is the default, lightweight performance tool used in Vlasiator. These timers track time spent in pre-defined code sections, with nested levels.</p>
<div class="highlight-cfg notranslate"><div class="highlight"><pre><span></span><span class="w">                                                        </span><span class="na">All timers. Set of identical timers has 461 processes with up to 16 threads each.</span>
<span class="na">------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------</span>
<span class="w"> </span><span class="na">|                                                                             | Threads |                         Time (s)                          | Calls  |               Workunit-rate               |</span>
<span class="na">------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------</span>
<span class="w"> </span><span class="na">| Id  | Lvl | Grp | Label                                                     | Avg     | Avg       | %         |  Max time,rank  |  Min time,rank  | Avg    | Total     | Per process | Unit            |</span>
<span class="na">------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------</span>
<span class="w"> </span><span class="na">| 1   | 1   |     | main                                                      | 1       | 8447      | 100       | 8447      | 29  | 8446      | 352 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 2   | 2   |     |   Initialization                                          | 1       | 56.34     | 0.667     | 56.35     | 240 | 56.31     | 486 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 3   | 3   |     |     Read parameters                                       | 1       | 0.09863   | 0.1751    | 0.1109    | 272 | 0.06298   | 486 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 4   | 3   |     |     open logFile &amp; diagnostic                             | 1       | 8.322e-05 | 0.0001477 | 0.005583  | 0   | 3.35e-05  | 299 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 5   | 3   |     |     Init project                                          | 1       | 1.566e-05 | 2.78e-05  | 0.0001138 | 101 | 1.223e-06 | 491 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 6   | 3   |     |     Init fieldsolver grids                                | 1       | 0.02803   | 0.04976   | 0.03005   | 187 | 0.02409   | 0   | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 7   | 3   |     |     Init grids                                            | 1       | 56.05     | 99.49     | 56.07     | 419 | 56.03     | 63  | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 8   | 4   |     |       Refine spatial cells                                | 1       | 13.26     | 23.66     | 13.8      | 509 | 13.04     | 269 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 9   | 5   |     |         Override refines                                  | 1       | 3.119     | 23.52     | 3.513     | 0   | 0.01176   | 225 | 2      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 10  | 5   |     |         Induce refines                                    | 1       | 1.417     | 10.69     | 1.419     | 80  | 1.416     | 326 | 2      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 11  | 5   |     |         Override unrefines                                | 1       | 0.002455  | 0.01851   | 0.003982  | 259 | 0.0004373 | 80  | 2      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 12  | 5   |     |         Map Refinement Level to FsGrid                    | 1       | 0.002119  | 0.01598   | 0.005737  | 283 | 0.00164   | 436 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 5   |     |         Other                                             | 1       | 8.722     | 65.76     | 11.61     | 225 | 8.221     | 409 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 13  | 4   |     |       Initial load-balancing                              | 1       | 3.444     | 6.145     | 4.54      | 265 | 2.344     | 368 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 14  | 4   |     |       Set initial state                                   | 1       | 38.26     | 68.25     | 39.11     | 8   | 37.5      | 61  | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 15  | 5   |     |         Set spatial cell coordinates                      | 1       | 0.00045   | 0.001176  | 0.0008721 | 91  | 0.0001717 | 480 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 16  | 5   |     |         Initialize system boundary conditions             | 1       | 0.08176   | 0.2137    | 0.1041    | 8   | 0.07811   | 133 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 17  | 6   |     |           ionosphere-sphericalFibonacci                   | 1       | 0.01461   | 17.87     | 0.01762   | 431 | 0.0142    | 11  | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 18  | 6   |     |           ionosphere-subdivideElement                     | 1       | 0.01286   | 15.73     | 0.01678   | 239 | 0.01116   | 316 | 12758  |           |             |                 |</span>
<span class="w"> </span><span class="na">| 19  | 6   |     |           ionosphere-readAtmosphericModelFile             | 1       | 0.002412  | 2.95      | 0.02269   | 8   | 0.002031  | 288 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 6   |     |           Other                                           | 1       | 0.05188   | 63.45     | 0.05909   | 442 | 0.0505    | 67  | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 20  | 5   |     |         Classify cells (sys boundary conditions)          | 1       | 0.4128    | 1.079     | 1.44      | 17  | 0.142     | 2   | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 21  | 5   |     |         Read restart                                      | 1       | 14.5      | 37.91     | 14.51     | 48  | 14.5      | 339 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 22  | 6   |     |           readGrid                                        | 1       | 14.5      | 99.99     | 14.51     | 56  | 14.5      | 339 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 23  | 7   |     |             readScalars                                   | 1       | 0.02776   | 0.1914    | 0.03241   | 56  | 0.02425   | 339 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 24  | 7   |     |             readDatalayout                                | 1       | 6.164     | 42.5      | 6.165     | 268 | 6.164     | 320 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 25  | 7   |     |             readCellParameters                            | 1       | 1.334     | 9.197     | 1.334     | 320 | 1.333     | 268 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 26  | 7   |     |             readBlockData                                 | 1       | 5.964     | 41.12     | 6.419     | 384 | 5.861     | 9   | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 27  | 7   |     |             updateMpiGridNeighbors                        | 1       | 0.4356    | 3.004     | 0.5568    | 17  | 0.01425   | 416 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 28  | 7   |     |             readFsGrid                                    | 1       | 0.4992    | 3.442     | 0.9608    | 188 | 0.4594    | 385 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 29  | 8   |     |               updateGhostCells                            | 1       | 0.01472   | 2.949     | 0.04934   | 469 | 0.00222   | 7   | 2      |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 8   |     |               Other                                       | 1       | 0.4845    | 97.05     | 0.9564    | 188 | 0.4167    | 470 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 30  | 7   |     |             readIonosphere                                | 1       | 0.07711   | 0.5317    | 0.07732   | 48  | 0.0764    | 398 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 7   |     |             Other                                         | 1       | 0.001704  | 0.01175   | 0.01768   | 489 | 0.001147  | 48  | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 6   |     |           Other                                           | 1       | 0.001097  | 0.007561  | 0.004076  | 48  | 0.0005522 | 365 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 31  | 5   |     |         Check boundary refinement                         | 1       | 0.0007237 | 0.001892  | 0.007383  | 422 | 2.734e-05 | 274 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 32  | 5   |     |         Apply system boundary conditions state            | 1       | 0.0007087 | 0.001853  | 0.02225   | 388 | 1.001e-06 | 440 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 33  | 5   | D   |         Balancing load                                    | 1       | 6.279     | 16.41     | 6.762     | 451 | 5.144     | 467 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 34  | 6   |     |           deallocate boundary data                        | 1       | 0.001852  | 0.02949   | 0.00632   | 385 | 0.0006113 | 225 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 35  | 6   |     |           dccrg.initialize_balance_load                   | 1       | 0.544     | 8.664     | 0.5548    | 114 | 0.519     | 452 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 36  | 6   |     |           Data transfers                                  | 1       | 0.4398    | 7.005     | 0.82      | 8   | 0.04192   | 54  | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 37  | 7   |     |             Preparing receives                            | 1       | 0.05701   | 12.96     | 0.169     | 448 | 6.501e-07 | 511 | 1643   | 1.329e+07 | 2.883e+04   | Spatial cells/s |</span>
<span class="w"> </span><span class="na">| 38  | 7   |     |             transfer_all_data                             | 1       | 0.2809    | 63.87     | 0.6845    | 8   | 0.01069   | 454 | 5      |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 7   |     |             Other                                         | 1       | 0.1019    | 23.17     | 0.5186    | 242 | 0.005889  | 220 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 39  | 6   |     |           dccrg.finish_balance_load                       | 1       | 1.458     | 23.22     | 4.899     | 19  | 0.4524    | 10  | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 40  | 6   |     |           compute_amr_transfer_flags                      | 1       | 0.006867  | 0.1094    | 0.07725   | 419 | 0.001871  | 38  | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 41  | 6   |     |           update block lists                              | 1       | 2.087     | 33.25     | 4.207     | 346 | 0.09587   | 407 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 42  | 7   | E   |             Velocity block list update                    | 1       | 1.967     | 94.25     | 4.054     | 271 | 0.02872   | 231 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 43  | 7   |     |             Preparing receives                            | 1       | 0.1198    | 5.742     | 0.2648    | 29  | 0.03104   | 441 | 1      | 1.547e+07 | 3.355e+04   | SpatialCells/s  |</span>
<span class="w"> </span><span class="na">|     | 7   |     |             Other                                         | 1       | 0.0001002 | 0.0048    | 0.004895  | 448 | 9.859e-06 | 253 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 44  | 6   |     |           update sysboundaries                            | 1       | 0.002269  | 0.03614   | 0.03475   | 187 | 7.975e-05 | 258 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 45  | 7   |     |             updateSysBoundariesAfterLoadBalance           | 1       | 0.002259  | 99.57     | 0.03474   | 187 | 7.706e-05 | 258 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 7   |     |             Other                                         | 1       | 9.825e-06 | 0.433     | 0.0007654 | 424 | 1.844e-06 | 173 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 46  | 6   |     |           Init solvers                                    | 1       | 5.478e-06 | 8.725e-05 | 0.000501  | 424 | 1.503e-06 | 186 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 47  | 6   |     |           set face neighbor ranks                         | 1       | 0.0251    | 0.3997    | 0.08548   | 69  | 0.008767  | 372 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 48  | 6   |     |           GetSeedIdsAndBuildPencils                       | 1       | 0.04302   | 0.6851    | 0.4       | 451 | 0.0137    | 372 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 49  | 7   |     |             getSeedIds                                    | 1       | 0.007916  | 18.4      | 0.05184   | 451 | 0.002307  | 292 | 3      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 50  | 7   |     |             buildPencils                                  | 1       | 0.03479   | 80.88     | 0.3432    | 451 | 0.01127   | 372 | 3      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 51  | 8   |     |               check_ghost_cells                           | 1       | 0.001349  | 3.878     | 0.2473    | 451 | 8.518e-05 | 51  | 3      |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 8   |     |               Other                                       | 1       | 0.03344   | 96.12     | 0.1216    | 425 | 0.01118   | 372 | 3      |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 7   |     |             Other                                         | 1       | 0.0003073 | 0.7143    | 0.004975  | 451 | 5.954e-05 | 323 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 6   |     |           Other                                           | 1       | 1.67      | 26.6      | 4.312     | 335 | 0.01161   | 235 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 52  | 5   | E   |         Fetch Neighbour data                              | 1       | 0.1985    | 0.5189    | 1.346     | 467 | 0.008552  | 419 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 53  | 5   |     |         setProjectBField                                  | 1       | 15.92     | 41.63     | 16.22     | 390 | 15.56     | 452 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 54  | 5   |     |         Init moments                                      | 1       | 0.1236    | 0.3231    | 0.1522    | 422 | 0.1015    | 110 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 55  | 5   |     |         Finish fsgrid setup                               | 1       | 0.2395    | 0.6261    | 0.275     | 289 | 0.2013    | 69  | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 56  | 6   |     |           AMR Filtering-Triangle-3D                       | 1       | 0.1186    | 49.51     | 0.1751    | 353 | 0.06054   | 0   | 2      |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 6   |     |           Other                                           | 1       | 0.1209    | 50.49     | 0.1792    | 456 | 0.06876   | 89  | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 5   |     |         Other                                             | 1       | 0.4901    | 1.281     | 1.414     | 435 | 0.006689  | 291 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 4   |     |       Other                                               | 1       | 1.089     | 1.943     | 2.09      | 207 | 0.7609    | 506 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 57  | 3   |     |     Init DROs                                             | 1       | 0.0004287 | 0.0007609 | 0.007283  | 448 | 0.000176  | 274 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 58  | 3   |     |     getFieldsFromFsGrid                                   | 1       | 0.06079   | 0.1079    | 0.0892    | 63  | 0.03098   | 224 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 59  | 3   |     |     ionosphere-updateIonosphereCommunicator               | 1       | 0.0208    | 0.03692   | 0.04898   | 85  | 0.001027  | 284 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 60  | 3   |     |     fieldtracing-ionosphere-fsgridCoupling                | 1       | 0.07022   | 0.1246    | 0.07346   | 64  | 0.04959   | 476 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 61  | 3   |     |     ionosphere-initSolver                                 | 1       | 0.004485  | 0.007961  | 0.01555   | 425 | 0.00365   | 294 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 3   |     |     Other                                                 | 1       | 0.003553  | 0.006306  | 0.02718   | 495 | 5.835e-05 | 128 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 62  | 2   |     |   report-memory-consumption                               | 1       | 0.05788   | 0.0006852 | 0.08273   | 0   | 0.04845   | 407 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 63  | 2   |     |   Simulation                                              | 1       | 8390      | 99.33     | 8391      | 29  | 8390      | 352 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 64  | 3   |     |     IO                                                    | 1       | 152.3     | 1.815     | 200.5     | 95  | 126.1     | 29  | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 65  | 4   |     |       checkExternalCommands                               | 1       | 0.003633  | 0.002385  | 1.571     | 0   | 0.0001467 | 334 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 66  | 4   |     |       logfile-io                                          | 1       | 0.03492   | 0.02293   | 0.2181    | 0   | 0.02167   | 68  | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 67  | 4   |     |       Bailout-allreduce                                   | 1       | 33.55     | 22.03     | 81.67     | 95  | 8.082     | 29  | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 68  | 4   |     |       compute-is-restart-written-and-extra-LB             | 1       | 0.01469   | 0.009645  | 0.01983   | 496 | 0.01111   | 78  | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 169 | 4   |     |       diagnostic-io                                       | 1       | 0.5115    | 0.3358    | 6.001     | 352 | 0.01028   | 329 | 76     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 196 | 4   |     |       Calculate volume gradients                          | 1       | 0.3295    | 0.2163    | 0.8119    | 95  | 0.09899   | 491 | 10     | 4.973e+07 | 1.079e+05   | Spatial Cells/s |</span>
<span class="w"> </span><span class="na">| 197 | 5   | E   |         Start comm                                        | 1       | 0.2989    | 90.71     | 0.7672    | 95  | 0.07832   | 491 | 10     | 5.482e+07 | 1.189e+05   | Spatial Cells/s |</span>
<span class="w"> </span><span class="na">| 198 | 5   |     |         Compute cells                                     | 1       | 0.03055   | 9.271     | 0.09107   | 371 | 0.01067   | 192 | 10     | 5.364e+08 | 1.163e+06   | Spatial Cells/s |</span>
<span class="w"> </span><span class="na">|     | 5   |     |         Other                                             | 1       | 4.922e-05 | 0.01494   | 0.0006382 | 254 | 2.93e-05  | 334 | 10     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 199 | 4   |     |       fieldtracing-ionosphere-openclosedTracing           | 1       | 1.076     | 0.7062    | 1.416     | 235 | 0.9152    | 19  | 10     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 200 | 4   |     |       fieldtracing-fullAndFluxTracing                     | 1       | 72.3      | 47.47     | 72.33     | 247 | 72.29     | 323 | 10     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 201 | 5   |     |         initialization-loop                               | 1       | 0.9673    | 1.338     | 1.02      | 247 | 0.9424    | 445 | 10     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 202 | 5   |     |         loop                                              | 1       | 67.91     | 93.93     | 67.97     | 0   | 67.86     | 107 | 10     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 203 | 6   |     |           MPI-loop                                        | 16      | 183.1     | 269.6     | 231.2     | 429 | 108.1     | 285 | 500    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 204 | 5   |     |         final-loop                                        | 1       | 0.3835    | 0.5304    | 0.4107    | 247 | 0.3736    | 323 | 10     |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 5   |     |         Other                                             | 1       | 3.041     | 4.206     | 3.077     | 107 | 2.968     | 0   | 10     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 205 | 4   |     |       write-system                                        | 1       | 18.68     | 12.26     | 18.89     | 360 | 18.32     | 9   | 10     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 206 | 5   | EA  |         Barrier-entering-writegrid                        | 1       | 0.4146    | 2.22      | 0.6223    | 360 | 0.0607    | 9   | 10     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 207 | 5   |     |         writeGrid-reduced                                 | 1       | 18.26     | 97.77     | 18.26     | 304 | 18.24     | 239 | 10     | 0.5316    | 0.001153    | GB/s            |</span>
<span class="w"> </span><span class="na">| 208 | 6   |     |           open                                            | 1       | 0.7832    | 4.289     | 0.8339    | 464 | 0.715     | 421 | 10     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 209 | 6   |     |           metadataIO                                      | 1       | 2.243     | 12.28     | 2.244     | 128 | 2.201     | 473 | 10     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 210 | 6   |     |           velocityspaceIO                                 | 1       | 0.1373    | 0.7521    | 0.1374    | 8   | 0.1373    | 494 | 10     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 211 | 6   |     |           reduceddataIO                                   | 1       | 15.06     | 82.45     | 15.06     | 505 | 15.05     | 64  | 10     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 212 | 7   |     |             writeDataReducer                              | 1       | 15.05     | 99.99     | 15.05     | 509 | 15.05     | 3   | 10     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 358 | 7   | EA  |             Barrier                                       | 1       | 0.000446  | 0.002963  | 0.0005074 | 1   | 0.0002238 | 476 | 10     |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 7   |     |             Other                                         | 1       | 0.0009875 | 0.006559  | 0.00157   | 473 | 0.0002494 | 400 | 10     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 359 | 6   |     |           close                                           | 1       | 0.03292   | 0.1803    | 0.03422   | 358 | 0.02962   | 456 | 10     |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 6   |     |           Other                                           | 1       | 0.00834   | 0.04567   | 0.0748    | 416 | 0.0003039 | 167 | 10     |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 5   |     |         Other                                             | 1       | 0.00232   | 0.01242   | 0.03255   | 376 | 0.0003313 | 321 | 10     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 360 | 4   |     |       write-restart                                       | 1       | 25.43     | 16.69     | 25.52     | 29  | 25.31     | 95  | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 361 | 5   | EA  |         BarrierEnteringWriteRestart                       | 1       | 0.0008407 | 0.003306  | 0.001098  | 6   | 1.548e-05 | 471 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 362 | 5   |     |         writeRestart                                      | 1       | 25.42     | 99.99     | 25.52     | 29  | 25.31     | 95  | 1      | 10.42     | 0.0226      | GB/s            |</span>
<span class="w"> </span><span class="na">| 363 | 6   |     |           DeallocateRemoteBlocks                          | 1       | 0.0256    | 0.1007    | 0.05758   | 29  | 0.0102    | 489 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 364 | 6   |     |           open                                            | 1       | 0.04866   | 0.1914    | 0.04887   | 0   | 0.0477    | 424 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 365 | 6   |     |           metadataIO                                      | 1       | 2.969     | 11.68     | 2.969     | 262 | 2.969     | 7   | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 366 | 6   |     |           reduceddataIO                                   | 1       | 10.58     | 41.63     | 10.58     | 0   | 10.57     | 105 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 433 | 6   |     |           velocityspaceIO                                 | 1       | 11.29     | 44.39     | 11.29     | 143 | 11.27     | 175 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 434 | 6   |     |           close                                           | 1       | 0.182     | 0.7159    | 0.1822    | 0   | 0.1632    | 7   | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 435 | 6   |     |           updateRemoteBlocks                              | 1       | 0.1936    | 0.7617    | 0.2922    | 29  | 0.08067   | 95  | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 436 | 7   | E   |             Velocity block list update                    | 1       | 0.07758   | 40.06     | 0.1156    | 313 | 0.04427   | 97  | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 437 | 7   |     |             Preparing receives                            | 1       | 0.116     | 59.92     | 0.2124    | 29  | 0.03089   | 95  | 1      | 1.601e+07 | 3.473e+04   | SpatialCells/s  |</span>
<span class="w"> </span><span class="na">|     | 7   |     |             Other                                         | 1       | 3.954e-05 | 0.02042   | 0.008907  | 94  | 1.009e-05 | 153 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 6   |     |           Other                                           | 1       | 0.1344    | 0.5287    | 0.1609    | 97  | 0.1022    | 29  | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 5   |     |         Other                                             | 1       | 0.002589  | 0.01018   | 0.03064   | 195 | 6.623e-05 | 355 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 4   |     |       Other                                               | 1       | 0.3924    | 0.2576    | 0.7548    | 9   | 0.1819    | 360 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 69  | 3   |     |     Propagate                                             | 1       | 8068      | 96.15     | 8094      | 29  | 8019      | 95  | 759    | 5.439e+09 | 1.18e+07    | Cells/s         |</span>
<span class="w"> </span><span class="na">| 70  | 4   |     |       Spatial-space                                       | 1       | 4172      | 51.71     | 4180      | 307 | 4164      | 110 | 759    | 1.052e+10 | 2.282e+07   | Cells/s         |</span>
<span class="w"> </span><span class="na">| 71  | 5   |     |         semilag-trans                                     | 1       | 4172      | 100       | 4180      | 307 | 4164      | 110 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 72  | 6   |     |           compute_cell_lists                              | 1       | 0.1408    | 0.003375  | 0.6254    | 371 | 0.0462    | 258 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 73  | 6   |     |           translate proton                                | 1       | 4127      | 98.93     | 4127      | 309 | 4127      | 371 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 74  | 7   | BE  |             barrier-trans-pre-z                           | 1       | 1.021     | 0.02475   | 1.122     | 309 | 0.5356    | 371 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 75  | 7   | E   |             transfer-stencil-data-z                       | 1       | 286.7     | 6.948     | 408.4     | 27  | 97.35     | 69  | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 76  | 7   |     |             compute-mapping-z                             | 1       | 357.6     | 8.664     | 673.1     | 140 | 207.3     | 186 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 77  | 8   |     |               setup                                       | 1       | 9.725     | 2.72      | 27.76     | 494 | 4.79      | 372 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 78  | 9   |     |                 buildBlockList                            | 1       | 4.477     | 46.03     | 14.49     | 113 | 1.243     | 148 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 79  | 9   |     |                 computeSpatialTargetCellsForPencils       | 1       | 5.018     | 51.59     | 18.2      | 494 | 1.425     | 162 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 9   |     |                 Other                                     | 1       | 0.2311    | 2.376     | 0.7481    | 441 | 0.09204   | 433 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 80  | 8   |     |               mapping                                     | 16      | 217.2     | 60.74     | 412.5     | 201 | 106.5     | 187 | 91523  |           |             |                 |</span>
<span class="w"> </span><span class="na">| 81  | 8   |     |               store                                       | 16      | 109.3     | 30.57     | 207.2     | 140 | 47.21     | 217 | 91523  |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 8   |     |               Other                                       | 1       | 21.37     | 5.975     | 75.72     | 69  | 4.764     | 332 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 82  | 7   | BE  |             barrier-trans-pre-update_remote-z             | 1       | 398       | 9.644     | 668.9     | 187 | 115.7     | 201 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 83  | 7   | E   |             update_remote-z                               | 1       | 388.4     | 9.41      | 397.6     | 27  | 380.2     | 69  | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 84  | 7   | BE  |             barrier-trans-pre-x                           | 1       | 9.746     | 0.2361    | 17.89     | 69  | 0.5317    | 27  | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 85  | 7   | E   |             transfer-stencil-data-x                       | 1       | 277.6     | 6.726     | 368       | 368 | 69.34     | 187 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 86  | 7   |     |             compute-mapping-x                             | 1       | 356       | 8.627     | 747.7     | 337 | 202.1     | 186 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 87  | 8   |     |               setup                                       | 1       | 12.76     | 3.585     | 46.76     | 113 | 5.537     | 190 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 88  | 9   |     |                 buildBlockList                            | 1       | 7.574     | 59.34     | 25.34     | 113 | 2.459     | 148 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 89  | 9   |     |                 computeSpatialTargetCellsForPencils       | 1       | 4.935     | 38.66     | 22.7      | 252 | 1.359     | 355 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 9   |     |                 Other                                     | 1       | 0.2551    | 1.999     | 0.7575    | 441 | 0.1093    | 437 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 90  | 8   |     |               mapping                                     | 16      | 219.3     | 61.61     | 487.5     | 337 | 130.3     | 187 | 91798  |           |             |                 |</span>
<span class="w"> </span><span class="na">| 91  | 8   |     |               store                                       | 16      | 103.9     | 29.18     | 201.5     | 337 | 47.16     | 186 | 91798  |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 8   |     |               Other                                       | 1       | 20.03     | 5.626     | 73.57     | 422 | 4.528     | 221 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 92  | 7   | BE  |             barrier-trans-pre-update_remote-x             | 1       | 443       | 10.73     | 735.9     | 187 | 1.142     | 337 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 93  | 7   | E   |             update_remote-x                               | 1       | 340.7     | 8.255     | 346.4     | 126 | 334.5     | 187 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 94  | 7   | BE  |             barrier-trans-pre-y                           | 1       | 6.601     | 0.16      | 12.8      | 187 | 0.8621    | 126 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 95  | 7   | E   |             transfer-stencil-data-y                       | 1       | 275.6     | 6.678     | 363.1     | 290 | 54.37     | 187 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 96  | 7   |     |             compute-mapping-y                             | 1       | 347.6     | 8.423     | 569.8     | 281 | 165.9     | 329 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 97  | 8   |     |               setup                                       | 1       | 14.53     | 4.18      | 36.49     | 113 | 7.491     | 319 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 98  | 9   |     |                 buildBlockList                            | 1       | 9.368     | 64.48     | 20.45     | 113 | 2.522     | 148 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 99  | 9   |     |                 computeSpatialTargetCellsForPencils       | 1       | 4.917     | 33.84     | 20.54     | 252 | 1.438     | 184 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 9   |     |                 Other                                     | 1       | 0.244     | 1.679     | 0.7736    | 398 | 0.1029    | 155 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 100 | 8   |     |               mapping                                     | 16      | 209.3     | 60.22     | 373       | 281 | 105       | 187 | 91156  |           |             |                 |</span>
<span class="w"> </span><span class="na">| 101 | 8   |     |               store                                       | 16      | 103       | 29.63     | 179.7     | 113 | 38.33     | 329 | 91156  |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 8   |     |               Other                                       | 1       | 20.76     | 5.972     | 86.9      | 69  | 4.44      | 326 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 102 | 7   | BE  |             barrier-trans-pre-update_remote-y             | 1       | 294.6     | 7.138     | 619.4     | 187 | 13.89     | 281 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 103 | 7   | E   |             update_remote-y                               | 1       | 335.3     | 8.124     | 343.7     | 11  | 327.9     | 324 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 104 | 7   | BE  |             barrier-trans-post-trans                      | 1       | 8.532     | 0.2067    | 15.89     | 324 | 0.11      | 11  | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 7   |     |             Other                                         | 1       | 0.06994   | 0.001695  | 0.1841    | 489 | 0.04843   | 385 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 105 | 6   |     |           compute-moments-n                               | 1       | 44.68     | 1.071     | 52.86     | 307 | 36.79     | 110 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 6   |     |           Other                                           | 1       | 0.02026   | 0.0004856 | 0.07562   | 371 | 0.009199  | 323 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 5   |     |         Other                                             | 1       | 0.008274  | 0.0001983 | 0.0533    | 492 | 0.002867  | 162 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 106 | 4   |     |       Update system boundaries (Vlasov post-translation)  | 1       | 205.7     | 2.55      | 239.2     | 29  | 150.9     | 95  | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 107 | 5   | E   |         Velocity block list update                        | 1       | 109.3     | 53.14     | 145.2     | 173 | 49.15     | 253 | 1518   |           |             |                 |</span>
<span class="w"> </span><span class="na">| 108 | 5   |     |         Preparing receives                                | 1       | 49.77     | 24.2      | 138.5     | 252 | 8.072     | 187 | 1518   | 6.556e+07 | 1.422e+05   | SpatialCells/s  |</span>
<span class="w"> </span><span class="na">| 109 | 5   | E   |         Start comm of cell and block data                 | 1       | 4.594     | 2.233     | 38.41     | 316 | 0.4894    | 298 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 110 | 5   |     |         Compute process inner cells                       | 1       | 2.093     | 1.018     | 28.05     | 253 | 0.03307   | 220 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 111 | 6   |     |           compute-moments-n                               | 1       | 0.5362    | 25.62     | 8.324     | 253 | 0.003975  | 204 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 6   |     |           Other                                           | 1       | 1.557     | 74.38     | 20.44     | 405 | 0.02823   | 220 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 112 | 5   | EG  |         Wait for receives                                 | 1       | 32.31     | 15.71     | 103.7     | 369 | 0.5165    | 29  | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 113 | 5   |     |         Compute process boundary cells                    | 1       | 1.428     | 0.6943    | 14.68     | 253 | 0.04366   | 29  | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 114 | 6   |     |           compute-moments-n                               | 1       | 0.3696    | 25.88     | 4.468     | 253 | 0.003528  | 7   | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 6   |     |           Other                                           | 1       | 1.058     | 74.12     | 10.21     | 253 | 0.0394    | 29  | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 5   |     |         Other                                             | 1       | 6.195     | 3.012     | 16.34     | 38  | 1.249     | 373 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 115 | 4   |     |       Compute interp moments                              | 1       | 0.3852    | 0.004775  | 1.547     | 328 | 0.1292    | 84  | 1518   |           |             |                 |</span>
<span class="w"> </span><span class="na">| 116 | 4   |     |       Propagate Fields                                    | 1       | 2338      | 28.98     | 2387      | 235 | 2300      | 29  | 759    | 5.319e+05 | 1154        | SpatialCells/s  |</span>
<span class="w"> </span><span class="na">| 117 | 5   |     |         fsgrid-coupling-in                                | 1       | 149.1     | 6.376     | 199.3     | 95  | 117.1     | 29  | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 118 | 6   |     |           AMR Filtering-Triangle-3D                       | 1       | 79.18     | 53.11     | 106.6     | 95  | 57.38     | 284 | 1518   |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 6   |     |           Other                                           | 1       | 69.9      | 46.89     | 122.7     | 352 | 33.39     | 243 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 119 | 5   |     |         Propagate magnetic field                          | 1       | 333.8     | 14.28     | 579       | 416 | 194.1     | 219 | 21252  | 2.857e+09 | 6.198e+06   | Spatial Cells/s |</span>
<span class="w"> </span><span class="na">| 120 | 6   |     |           Compute cells                                   | 1       | 43.73     | 13.1      | 68.27     | 0   | 41.85     | 400 | 21252  | 2.181e+10 | 4.731e+07   | Spatial Cells/s |</span>
<span class="w"> </span><span class="na">| 121 | 6   | E   |           MPI                                             | 1       | 180.9     | 54.2      | 316.3     | 457 | 118.7     | 31  | 42504  |           |             |                 |</span>
<span class="w"> </span><span class="na">| 122 | 6   |     |           Compute system boundary cells                   | 1       | 109       | 32.67     | 394.2     | 416 | 4.367     | 99  | 63756  | 1.749e+10 | 3.794e+07   | /s              |</span>
<span class="w"> </span><span class="na">|     | 6   |     |           Other                                           | 1       | 0.1105    | 0.0331    | 0.1611    | 150 | 0.08503   | 28  | 21252  |           |             |                 |</span>
<span class="w"> </span><span class="na">| 123 | 5   |     |         Calculate face derivatives                        | 1       | 240.1     | 10.27     | 374       | 345 | 134.6     | 63  | 21252  | 3.971e+09 | 8.615e+06   | Spatial Cells/s |</span>
<span class="w"> </span><span class="na">| 124 | 6   | E   |           MPI                                             | 1       | 162.8     | 67.81     | 281.3     | 353 | 72.67     | 319 | 21252  |           |             |                 |</span>
<span class="w"> </span><span class="na">| 125 | 6   |     |           Compute cells                                   | 1       | 77.25     | 32.17     | 94.67     | 337 | 51.95     | 59  | 21252  | 1.234e+10 | 2.678e+07   | Spatial Cells/s |</span>
<span class="w"> </span><span class="na">|     | 6   |     |           Other                                           | 1       | 0.05519   | 0.02299   | 0.09542   | 209 | 0.03441   | 451 | 21252  |           |             |                 |</span>
<span class="w"> </span><span class="na">| 126 | 5   |     |         Calculate GradPe term                             | 1       | 16.23     | 0.6942    | 27.52     | 283 | 10.33     | 327 | 1518   | 4.197e+09 | 9.104e+06   | Spatial Cells/s |</span>
<span class="w"> </span><span class="na">| 127 | 6   | E   |           MPI                                             | 1       | 15.05     | 92.76     | 26.17     | 283 | 9.326     | 327 | 1518   |           |             |                 |</span>
<span class="w"> </span><span class="na">| 128 | 6   |     |           Compute cells                                   | 1       | 1.171     | 7.214     | 1.465     | 300 | 0.8719    | 5   | 1518   | 5.818e+10 | 1.262e+08   | Spatial Cells/s |</span>
<span class="w"> </span><span class="na">|     | 6   |     |           Other                                           | 1       | 0.004606  | 0.02838   | 0.006637  | 244 | 0.003104  | 384 | 1518   |           |             |                 |</span>
<span class="w"> </span><span class="na">| 129 | 5   |     |         Calculate Hall term                               | 1       | 663       | 28.36     | 845.5     | 274 | 486.1     | 63  | 21252  | 1.438e+09 | 3.12e+06    | Spatial Cells/s |</span>
<span class="w"> </span><span class="na">| 130 | 6   | E   |           MPI                                             | 1       | 173.8     | 26.21     | 316.8     | 274 | 95        | 127 | 21252  |           |             |                 |</span>
<span class="w"> </span><span class="na">| 131 | 6   |     |           Compute cells                                   | 1       | 489.1     | 73.78     | 529.6     | 418 | 383.2     | 455 | 21252  |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 6   |     |           Other                                           | 1       | 0.0765    | 0.01154   | 0.1332    | 213 | 0.04423   | 384 | 21252  |           |             |                 |</span>
<span class="w"> </span><span class="na">| 132 | 5   |     |         Calculate upwinded electric field                 | 1       | 784.3     | 33.55     | 1072      | 508 | 648.5     | 72  | 21252  | 1.216e+09 | 2.637e+06   | Spatial Cells/s |</span>
<span class="w"> </span><span class="na">| 133 | 6   | E   |           MPI                                             | 1       | 439.7     | 56.06     | 806.6     | 508 | 271.5     | 72  | 42504  |           |             |                 |</span>
<span class="w"> </span><span class="na">| 134 | 6   |     |           Compute cells                                   | 1       | 344.5     | 43.92     | 396.3     | 339 | 239.1     | 7   | 21252  | 2.768e+09 | 6.005e+06   | Spatial Cells/s |</span>
<span class="w"> </span><span class="na">|     | 6   |     |           Other                                           | 1       | 0.1828    | 0.02331   | 0.2597    | 174 | 0.1262    | 384 | 21252  |           |             |                 |</span>
<span class="w"> </span><span class="na">| 135 | 5   |     |         FS subcycle stuff                                 | 1       | 82.19     | 3.516     | 178.6     | 63  | 32.71     | 347 | 10626  |           |             |                 |</span>
<span class="w"> </span><span class="na">| 136 | 6   |     |           MPI_Allreduce                                   | 1       | 80.68     | 98.16     | 177.1     | 63  | 31.16     | 347 | 9867   |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 6   |     |           Other                                           | 1       | 1.515     | 1.844     | 1.661     | 377 | 1.287     | 44  | 10626  |           |             |                 |</span>
<span class="w"> </span><span class="na">| 137 | 5   |     |         Calculate volume averaged fields                  | 1       | 1.224     | 0.05237   | 1.469     | 436 | 0.9462    | 5   | 759    | 2.782e+10 | 6.034e+07   | Spatial Cells/s |</span>
<span class="w"> </span><span class="na">| 138 | 5   |     |         Calculate volume derivatives                      | 1       | 7.981     | 0.3414    | 10.1      | 438 | 5.108     | 7   | 759    | 4.267e+09 | 9.256e+06   | Spatial Cells/s |</span>
<span class="w"> </span><span class="na">| 139 | 6   | E   |           Start comm                                      | 1       | 7.188     | 90.06     | 9.251     | 438 | 4.523     | 7   | 759    | 4.738e+09 | 1.028e+07   | Spatial Cells/s |</span>
<span class="w"> </span><span class="na">| 140 | 6   |     |           Compute cells                                   | 1       | 0.7908    | 9.909     | 0.9636    | 154 | 0.5727    | 63  | 759    | 4.307e+10 | 9.342e+07   | Spatial Cells/s |</span>
<span class="w"> </span><span class="na">|     | 6   |     |           Other                                           | 1       | 0.002362  | 0.02959   | 0.01186   | 110 | 0.001497  | 505 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 141 | 5   |     |         Calculate curvature                               | 1       | 4.991     | 0.2135    | 7.732     | 127 | 4.105     | 27  | 759    | 6.824e+09 | 1.48e+07    | Spatial Cells/s |</span>
<span class="w"> </span><span class="na">| 142 | 6   | E   |           Start comm                                      | 1       | 3.773     | 75.6      | 6.749     | 127 | 2.927     | 27  | 759    | 9.027e+09 | 1.958e+07   | Spatial Cells/s |</span>
<span class="w"> </span><span class="na">|     | 6   |     |           Other                                           | 1       | 1.218     | 24.4      | 1.414     | 366 | 0.8458    | 0   | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 143 | 5   |     |         getFieldsFromFsGrid                               | 1       | 49.23     | 2.106     | 63.17     | 284 | 31.06     | 88  | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 5   |     |         Other                                             | 1       | 5.849     | 0.2502    | 11.48     | 119 | 3.534     | 210 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 144 | 4   |     |       Velocity-space                                      | 1       | 893.8     | 11.08     | 928.5     | 7   | 830.9     | 397 | 759    | 4.91e+10  | 1.065e+08   | Cells/s         |</span>
<span class="w"> </span><span class="na">| 145 | 5   |     |         semilag-acc                                       | 1       | 852.6     | 95.4      | 892.9     | 7   | 795.9     | 397 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 146 | 6   |     |           Compute _V moments                              | 1       | 22.43     | 2.63      | 24.92     | 140 | 19.45     | 38  | 1518   |           |             |                 |</span>
<span class="w"> </span><span class="na">| 147 | 6   |     |           cell-semilag-acc                                | 16      | 470.6     | 55.19     | 558.6     | 242 | 395.8     | 10  | 158608 |           |             |                 |</span>
<span class="w"> </span><span class="na">| 148 | 7   |     |             compute-transform                             | 16      | 0.2146    | 0.04561   | 1.066     | 499 | 0.06968   | 372 | 158608 |           |             |                 |</span>
<span class="w"> </span><span class="na">| 149 | 7   |     |             compute-intersections                         | 16      | 0.07004   | 0.01488   | 0.3508    | 400 | 0.02048   | 258 | 158608 |           |             |                 |</span>
<span class="w"> </span><span class="na">| 150 | 7   |     |             compute-mapping                               | 16      | 470.1     | 99.9      | 558.1     | 242 | 395.5     | 10  | 158608 |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 7   |     |             Other                                         | 16      | 0.1851    | 0.03933   | 1.155     | 308 | 0.051     | 192 | 158608 |           |             |                 |</span>
<span class="w"> </span><span class="na">| 151 | 6   | C   |           re-adjust blocks                                | 1       | 321.6     | 37.71     | 408.2     | 10  | 215.3     | 484 | 1518   |           |             |                 |</span>
<span class="w"> </span><span class="na">| 152 | 7   |     |             Compute with_content_list                     | 1       | 39.93     | 12.42     | 65.07     | 307 | 27.81     | 333 | 1518   |           |             |                 |</span>
<span class="w"> </span><span class="na">| 153 | 7   | E   |             Transfer with_content_list                    | 1       | 134.5     | 41.83     | 233.7     | 10  | 31.66     | 97  | 1518   |           |             |                 |</span>
<span class="w"> </span><span class="na">| 154 | 7   |     |             Adjusting blocks                              | 1       | 85.3      | 26.53     | 106.9     | 292 | 65.34     | 398 | 1518   |           |             |                 |</span>
<span class="w"> </span><span class="na">| 155 | 7   | E   |             Velocity block list update                    | 1       | 41.97     | 13.05     | 70.57     | 302 | 16.17     | 332 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 156 | 7   |     |             Preparing receives                            | 1       | 19.81     | 6.161     | 48.75     | 243 | 3.158     | 187 | 759    | 7.112e+07 | 1.543e+05   | SpatialCells/s  |</span>
<span class="w"> </span><span class="na">|     | 7   |     |             Other                                         | 1       | 0.01958   | 0.006089  | 0.03275   | 135 | 0.01447   | 253 | 1518   |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 6   |     |           Other                                           | 1       | 38.09     | 4.467     | 56.59     | 88  | 22.79     | 284 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 157 | 5   |     |         Compute _V moments                                | 1       | 41.07     | 4.595     | 52.77     | 47  | 23.02     | 68  | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 5   |     |         Other                                             | 1       | 0.063     | 0.007049  | 0.1671    | 328 | 0.02175   | 124 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 158 | 4   |     |       Update system boundaries (Vlasov post-acceleration) | 1       | 214.2     | 2.655     | 261.9     | 397 | 182.1     | 179 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 159 | 5   | E   |         Velocity block list update                        | 1       | 109.9     | 51.31     | 158.7     | 69  | 50.82     | 240 | 1518   |           |             |                 |</span>
<span class="w"> </span><span class="na">| 160 | 5   |     |         Preparing receives                                | 1       | 47.05     | 21.97     | 132.6     | 252 | 7.282     | 187 | 1518   | 6.935e+07 | 1.504e+05   | SpatialCells/s  |</span>
<span class="w"> </span><span class="na">| 161 | 5   | E   |         Start comm of cell and block data                 | 1       | 4.782     | 2.233     | 36.23     | 316 | 0.4917    | 298 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 162 | 5   |     |         Compute process inner cells                       | 1       | 2.086     | 0.9741    | 27.95     | 253 | 0.03692   | 124 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 163 | 6   |     |           Compute _V moments                              | 1       | 0.5466    | 26.2      | 8.406     | 253 | 0.004     | 204 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 6   |     |           Other                                           | 1       | 1.54      | 73.8      | 20.01     | 307 | 0.03212   | 124 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 164 | 5   | EG  |         Wait for receives                                 | 1       | 32.29     | 15.08     | 100.9     | 369 | 0.8707    | 29  | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 165 | 5   |     |         Compute process boundary cells                    | 1       | 1.429     | 0.667     | 14.67     | 253 | 0.04702   | 29  | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 166 | 6   |     |           Compute _V moments                              | 1       | 0.3795    | 26.57     | 4.489     | 253 | 0.003492  | 7   | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 6   |     |           Other                                           | 1       | 1.049     | 73.43     | 10.19     | 253 | 0.04324   | 29  | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 5   |     |         Other                                             | 1       | 16.63     | 7.766     | 54.55     | 398 | 0.9602    | 7   | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 170 | 4   |     |       fieldtracing-ionosphere-fsgridCoupling              | 1       | 2.399     | 0.02973   | 3.617     | 88  | 1.359     | 284 | 49     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 171 | 4   |     |       ionosphere-mapDownMagnetosphere                     | 1       | 0.6545    | 0.008113  | 0.6979    | 421 | 0.6445    | 398 | 49     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 172 | 4   |     |       ionosphere-calculateConductivityTensor              | 1       | 0.1307    | 0.00162   | 0.1341    | 234 | 0.13      | 96  | 49     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 173 | 4   |     |       ionosphere-solve                                    | 1       | 240.4     | 2.98      | 250.5     | 416 | 236.1     | 58  | 49     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 174 | 5   |     |         ionosphere-initSolver                             | 1       | 0.2439    | 0.1014    | 0.2703    | 427 | 0.2394    | 228 | 49     |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 5   |     |         Other                                             | 1       | 240.2     | 99.9      | 250.3     | 416 | 235.9     | 58  | 49     |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 4   |     |       Other                                               | 1       | 0.02644   | 0.0003277 | 0.105     | 374 | 0.01267   | 309 | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 167 | 3   |     |     Project endTimeStep                                   | 1       | 0.0009495 | 1.132e-05 | 0.002589  | 371 | 0.0006758 | 17  | 759    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 168 | 3   |     |     compute-timestep                                      | 1       | 29.74     | 0.3545    | 29.84     | 94  | 29.39     | 371 | 758    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 175 | 3   | D   |     Balancing load                                        | 1       | 124.4     | 1.483     | 137.3     | 498 | 104.1     | 51  | 38     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 176 | 4   |     |       deallocate boundary data                            | 1       | 0.9602    | 0.7719    | 1.997     | 28  | 0.4402    | 187 | 38     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 177 | 4   |     |       dccrg.initialize_balance_load                       | 1       | 3.131     | 2.517     | 3.744     | 395 | 2.197     | 28  | 38     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 178 | 4   |     |       Data transfers                                      | 1       | 13.69     | 11.01     | 16.1      | 141 | 11.93     | 474 | 38     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 179 | 5   |     |         Preparing receives                                | 1       | 0.2315    | 1.691     | 2.429     | 195 | 0.02703   | 68  | 5440   | 1.082e+07 | 2.348e+04   | Spatial cells/s |</span>
<span class="w"> </span><span class="na">| 180 | 5   |     |         transfer_all_data                                 | 1       | 5.931     | 43.32     | 10.13     | 440 | 2.845     | 411 | 190    |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 5   |     |         Other                                             | 1       | 7.527     | 54.98     | 10.53     | 248 | 3.963     | 202 | 38     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 181 | 4   |     |       dccrg.finish_balance_load                           | 1       | 30.99     | 24.91     | 77.35     | 371 | 14.49     | 271 | 38     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 182 | 4   |     |       compute_amr_transfer_flags                          | 1       | 0.08713   | 0.07004   | 0.3571    | 371 | 0.01466   | 124 | 38     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 183 | 4   |     |       update block lists                                  | 1       | 44.49     | 35.76     | 80.37     | 30  | 14.63     | 371 | 38     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 184 | 5   | E   |         Velocity block list update                        | 1       | 41.15     | 92.5      | 76.77     | 30  | 10.89     | 278 | 38     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 185 | 5   |     |         Preparing receives                                | 1       | 3.334     | 7.495     | 6.418     | 29  | 1.116     | 187 | 38     | 2.116e+07 | 4.589e+04   | SpatialCells/s  |</span>
<span class="w"> </span><span class="na">|     | 5   |     |         Other                                             | 1       | 0.0003724 | 0.0008372 | 0.005403  | 364 | 0.0002562 | 162 | 38     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 186 | 4   |     |       update sysboundaries                                | 1       | 0.07129   | 0.05731   | 0.8916    | 371 | 0.0038    | 309 | 38     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 187 | 5   |     |         updateSysBoundariesAfterLoadBalance               | 1       | 0.07104   | 99.64     | 0.8914    | 371 | 0.003696  | 309 | 38     |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 5   |     |         Other                                             | 1       | 0.0002579 | 0.3617    | 0.02458   | 247 | 9.585e-05 | 338 | 38     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 188 | 4   |     |       Init solvers                                        | 1       | 2.258e-05 | 1.815e-05 | 0.0002023 | 285 | 1.138e-05 | 187 | 38     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 189 | 4   |     |       set face neighbor ranks                             | 1       | 0.9661    | 0.7766    | 3.083     | 69  | 0.3416    | 274 | 38     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 190 | 4   |     |       GetSeedIdsAndBuildPencils                           | 1       | 1.562     | 1.256     | 4.632     | 75  | 0.5381    | 192 | 38     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 191 | 5   |     |         getSeedIds                                        | 1       | 0.2891    | 18.5      | 0.9359    | 247 | 0.09034   | 274 | 114    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 192 | 5   |     |         buildPencils                                      | 1       | 1.262     | 80.8      | 3.842     | 75  | 0.4409    | 192 | 114    |           |             |                 |</span>
<span class="w"> </span><span class="na">| 193 | 6   |     |           check_ghost_cells                               | 1       | 0.008167  | 0.647     | 0.1068    | 14  | 0.001828  | 452 | 114    |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 6   |     |           Other                                           | 1       | 1.254     | 99.35     | 3.832     | 75  | 0.4389    | 192 | 114    |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 5   |     |         Other                                             | 1       | 0.01092   | 0.6987    | 0.05609   | 371 | 0.002525  | 128 | 38     |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 4   |     |       Other                                               | 1       | 28.45     | 22.87     | 54.78     | 177 | 1.772     | 352 | 38     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 194 | 3   |     |     Shrink_to_fit                                         | 1       | 0.006649  | 7.925e-05 | 0.01907   | 395 | 0.003076  | 298 | 38     |           |             |                 |</span>
<span class="w"> </span><span class="na">| 195 | 3   |     |     ionosphere-updateIonosphereCommunicator               | 1       | 16.12     | 0.1922    | 36.39     | 51  | 3.218     | 498 | 38     |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 3   |     |     Other                                                 | 1       | 0.1116    | 0.00133   | 0.4158    | 371 | 0.04017   | 309 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">| 438 | 2   |     |   Finalization                                            | 1       | 0.0001897 | 2.246e-06 | 0.01933   | 9   | 1.181e-05 | 208 | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 2   |     |   Other                                                   | 1       | 1.658e-05 | 1.963e-07 | 0.0006017 | 181 | 8.618e-06 | 50  | 1      |           |             |                 |</span>
<span class="w"> </span><span class="na">|     | 1   |     | Other                                                     | 1       | 0.08239   | 0.0009754 | 0.1725    | 352 | 0.002773  | 466 | 0      |           |             |                 |</span>
<span class="na">------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">tau</span></code> profiler can also be used to hook into <code class="docutils literal notranslate"><span class="pre">phiprof</span></code> timers.</p>
<p>PAPI can be used for memory use monitoring, and it is recommended to be used - monitoring the high water marks of memory use is well worth the trouble.</p>
</section>
<section id="lustre-striping">
<h3>Lustre striping<a class="headerlink" href="#lustre-striping" title="Link to this heading"></a></h3>
<p>Please refer to <a class="reference external" href="https://docs.lumi-supercomputer.eu/storage/parallel-filesystems/lustre/#file-striping">LUMI docs</a> for details.</p>
<p>Striping refers to spreading a file across several storage targets, and it is used to have better performance for parallel writes for large files.</p>
<p>Rules of thumb:</p>
<ul class="simple">
<li><p>Number tasks should be divisible by the number of stripes.</p></li>
<li><p>Do not use more stripes than there are OSTs (for example, we use 20/32).</p></li>
<li><p>Do not stripe small files: have one folder for restart files, and another for (each class of) bulk files.</p></li>
<li><p>One stripe per ~5 GB of file size is what we have used for bulk files</p></li>
</ul>
<p><code class="docutils literal notranslate"><span class="pre">lfs</span> <span class="pre">getstripe</span> <span class="pre">&lt;path&gt;</span></code> and <code class="docutils literal notranslate"><span class="pre">lfs</span> <span class="pre">setstripe</span> <span class="pre">--count</span> <span class="pre">&lt;n&gt;</span> <span class="pre">&lt;path&gt;</span></code> are the relevant commands.</p>
<section id="exercise">
<h4>Exercise:<a class="headerlink" href="#exercise" title="Link to this heading"></a></h4>
<ol class="arabic simple">
<li><p>Create a prototype run folder</p></li>
<li><p>Create a folder <code class="docutils literal notranslate"><span class="pre">restart/</span></code></p></li>
<li><p>Given an estimate of 1TB per restart file, set the striping of the <code class="docutils literal notranslate"><span class="pre">restart/</span></code> folder to a suitable values.
#. <code class="docutils literal notranslate"><span class="pre">lfs</span> <span class="pre">setstripe</span> <span class="pre">--count</span> <span class="pre">&lt;n&gt;</span> <span class="pre">&lt;path&gt;</span></code>
#. <code class="docutils literal notranslate"><span class="pre">lfs</span> <span class="pre">setstripe</span> <span class="pre">--stripe-size</span> <span class="pre">16777216</span> <span class="pre">&lt;path&gt;</span></code></p></li>
<li><p>Create a folder <code class="docutils literal notranslate"><span class="pre">bulks/</span></code></p></li>
<li><p>Given an estimate of 20GB per bulk file, set the striping of <code class="docutils literal notranslate"><span class="pre">bulks/</span></code> to a suitable value.</p></li>
<li><p>Check the stripe counts for both folders with <code class="docutils literal notranslate"><span class="pre">lfs</span> <span class="pre">getstripe</span></code>.</p></li>
</ol>
<p>Next: how to communicate these to Vlasiator!</p>
</section>
<section id="i-o-config-flags">
<h4>I/O config flags<a class="headerlink" href="#i-o-config-flags" title="Link to this heading"></a></h4>
<p>Example from current large production run (5.5 TB restart files currently):</p>
<div class="highlight-cfg notranslate"><div class="highlight"><pre><span></span><span class="k">[io]</span>
<span class="na">diagnostic_write_interval</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">10</span>
<span class="na">write_initial_state</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">0</span>
<span class="na">restart_walltime_interval</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">28400</span>
<span class="na">restart_write_path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">restart</span>
<span class="na">number_of_restarts</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">6</span><span class="w"> </span><span class="c1"># = 8h / 28800s, change if modifying time limit or restart interval</span>
<span class="na">vlsv_buffer_size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">0</span>
<span class="na">restart_write_mpiio_hint_key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">cb_buffer_size</span>
<span class="na">restart_write_mpiio_hint_value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">16777216</span>
<span class="na">restart_write_mpiio_hint_key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">striping_unit</span>
<span class="na">restart_write_mpiio_hint_value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">16777216</span>
<span class="na">restart_write_mpiio_hint_key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">romio_cb_write</span>
<span class="na">restart_write_mpiio_hint_value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">disable</span>
<span class="na">restart_read_mpiio_hint_key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">romio_ds_read</span>
<span class="na">restart_read_mpiio_hint_value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">disable</span>
<span class="na">restart_read_mpiio_hint_key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">romio_cb_read</span>
<span class="na">restart_read_mpiio_hint_value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">disable</span>
<span class="na">write_restart_stripe_factor</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">20</span>
</pre></div>
</div>
<p>Let’s check these in a bit more detail.</p>
<p>These set up restart file storing intervals and the path where to write, see next section:</p>
<div class="highlight-cfg notranslate"><div class="highlight"><pre><span></span><span class="na">restart_walltime_interval</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">28400</span>
<span class="na">restart_write_path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">restart</span>
<span class="na">number_of_restarts</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">6</span><span class="w"> </span><span class="c1"># = 8h / 28800s, change if modifying time limit or restart interval</span>
</pre></div>
</div>
<p>These are hints for collective MPI I/O, in key-value pairs.</p>
<div class="highlight-cfg notranslate"><div class="highlight"><pre><span></span><span class="na">restart_write_mpiio_hint_key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">cb_buffer_size</span>
<span class="na">restart_write_mpiio_hint_value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">16777216</span>
<span class="na">restart_write_mpiio_hint_key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">striping_unit</span>
<span class="na">restart_write_mpiio_hint_value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">16777216</span>
<span class="na">restart_write_mpiio_hint_key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">romio_cb_write</span>
<span class="na">restart_write_mpiio_hint_value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">disable</span>
<span class="na">restart_read_mpiio_hint_key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">romio_ds_read</span>
<span class="na">restart_read_mpiio_hint_value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">disable</span>
<span class="na">restart_read_mpiio_hint_key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">romio_cb_read</span>
<span class="na">restart_read_mpiio_hint_value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">disable</span>
</pre></div>
</div>
<p>Notably, we are using here 16 MB buffers and a matching stripe unit. These give decent 90s writes for 5.5 TB restart files from 500 nodes on LUMI.</p>
<p>The following informs Vlasiator of the restart file striping on Lustre (see below):</p>
<div class="highlight-cfg notranslate"><div class="highlight"><pre><span></span><span class="na">write_restart_stripe_factor</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">20</span>
</pre></div>
</div>
</section>
</section>
<section id="babysitting">
<h3>Babysitting<a class="headerlink" href="#babysitting" title="Link to this heading"></a></h3>
<p>Vlasiator runs usually take a while to complete, and everything might not go as planned - node or interconnect failures come to mind. It is also easy to encounter edge cases where the plasma VDFs “hit the walls” of their velocity space, so prototyping and iteration of run configurations will come up. It is also good to keep track of the performance, memory consumption and accrued costs over the simulation run.</p>
<section id="writing-restarts">
<h4>Writing restarts<a class="headerlink" href="#writing-restarts" title="Link to this heading"></a></h4>
<p>We write restart files at given wall-clock time intervals, given in the config file, as already seen above:</p>
<div class="highlight-cfg notranslate"><div class="highlight"><pre><span></span><span class="na">restart_walltime_interval</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">28400</span>
<span class="na">restart_write_path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">restart</span>
<span class="na">number_of_restarts</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">6</span><span class="w"> </span><span class="c1"># = 8h / 28800s, change if modifying time limit or restart interval</span>
</pre></div>
</div>
<p>Here we write a restart file each (a bit less than) 8 hours, to have a good coverage over a 48h slot. The last restart will be written at 47 hours 20 minutes, after which the run will finalize. This allows for a bit of a margin wrt. file system hiccups when writing - in case the file system is clogged at the time of the last write and it takes 30min to go through, we prefer to sacrifice a bit of the slot time for safety.</p>
</section>
<section id="restarting">
<h4>Restarting<a class="headerlink" href="#restarting" title="Link to this heading"></a></h4>
<p>This is rather simple! To continue running from the last restart, one issues the <code class="docutils literal notranslate"><span class="pre">--restart.filename</span></code> program option at launch, either via the command line or by editing the config. Here, a snippet that finds and uses the last written restart file in the directory <code class="docutils literal notranslate"><span class="pre">./restart/</span></code>:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>srun<span class="w"> </span>./vlasiator<span class="w"> </span>--run_config<span class="w"> </span>Flowthrough_amr.cfg<span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--restart.filename<span class="w"> </span><span class="k">$(</span><span class="w"> </span>ls<span class="w"> </span>restart/restart*vlsv<span class="w"> </span><span class="p">|</span><span class="w"> </span>tail<span class="w"> </span>-n<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="k">)</span>
</pre></div>
</div>
<p>When restarting, you can change config file and job script parameters, to e.g. introduce new/forgotten output variables, updated binary, or additional nodes - as the run progresses, the kinetic VDFs will usually take up much more resources than the initial ones.</p>
<p>Restarted runs will append to an existing logfile. Even if you do multiple restarts from the same savestate.</p>
</section>
<section id="external-commands">
<h4>External commands<a class="headerlink" href="#external-commands" title="Link to this heading"></a></h4>
<p>One can signal Vlasiator during run-time via files in the run directory. For example, creating a STOP file by <code class="docutils literal notranslate"><span class="pre">touch</span> <span class="pre">STOP</span></code> will signal the run to dump a restart and quit gracefully. The filename is appended with a timestamp. Other commands are available:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">SAVE</span></code></p>
<p>Dump a new restart file.</p>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">STOP</span></code></p>
<p>Stop the run with a restart write.</p>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">KILL</span></code></p>
<p>Stop the run <em>without</em> a restart write.</p>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">DOLB</span></code></p>
<p>Force a load balance refresh - if walltime per timestep has grown unexpectedly, this might help.</p>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">DOMR</span></code></p>
<p>Force a mesh re-refinement.</p>
</li>
</ul>
</section>
</section>
</section>
<section id="exercises">
<h2>Exercises<a class="headerlink" href="#exercises" title="Link to this heading"></a></h2>
<p>Today we will look at running small-ish simulations and scaling tests. We’ll start by performing few runs to fill in a scaling test spreadsheet and draw some conclusions from those runs. There are few heavier prototypes for overnight runs or heavier testing, and we will get to know tools for inspecting the Vlasiator outputs tomorrow.</p>
<section id="scaling-test">
<h3>Scaling test<a class="headerlink" href="#scaling-test" title="Link to this heading"></a></h3>
<p><em>Getting to know the basic run setup.</em></p>
<p>We are going to be running a Flowthrough test to look a bit at weak scaling. Find the configuration file and the job script from <code class="docutils literal notranslate"><span class="pre">/scratch/project_465000693/example_runs/scaling/baseline/Flowthrough_amr.cfg</span></code>.</p>
<p>The test is a tube, with initial solar wind plasma flowing along the X direction. The inflow boundary injects faster, more dense solar wind into the domain, and the Y and Z directions are periodic. Dynamic AMR is applied to the simulation, tracking the interface between fast and slow flows.</p>
<p>To calculate weak scaling, we will expand the Y and Z dimensions of the domain with some factors, with a matching increase in cores. To inspect task-thread balance, we move from having many tasks with few cores per task to few tasks with many cores per task - but keep the number of cores constant!</p>
<p>Pick a line or two for yourselves and modify your config and job script accordingly. The shared sheet has some predetermined values, but feel free to experiment further (and add lines with notes).</p>
<p>To begin with, we inspect the <code class="docutils literal notranslate"><span class="pre">logfile.txt</span></code> produced by the simulation run - at the end of a successful run, the file will show you total wall-clock runtime and some other counters - mark up these for the exercise results! The logfile is handy for surface-level checks, but for in-depth analysis of performance one should consult the phiprof outputs.</p>
</section>
<section id="prototype-magnetosphere3d-ionosphere3d">
<h3>Prototype: Magnetosphere3D/Ionosphere3D<a class="headerlink" href="#prototype-magnetosphere3d-ionosphere3d" title="Link to this heading"></a></h3>
<p><em>Advanced playthings</em></p>
<p>These prototypes can be played with - Magnetosphere3D configuration is included in the Vlasiator master under <code class="docutils literal notranslate"><span class="pre">samples/</span></code>. The sample magnetosphere is a good example of inner boundary instabilities at low resolutions - inner boundary VDFs hit the walls after ~60s of simulation time.</p>
<p>Ionosphere3D is freshly adapted version of the above to use an ionospheric inner boundary, and is somewhat more stable (with a highest-resolution region used for the inner boundary).</p>
<p>Note that these are “cheap” to run, expected O(100k) CPUh cost for a ~fully-developed system at around 1000s. If you wish to run these or use altered parameters, please feel free, but it may be a good idea to team up! Expect to use 16 nodes, so it would be hard to guarantee slots for everyone. You may also pick up the Ionosphere3D example run restart and keep running from there.</p>
</section>
<section id="prototype-mercury5d">
<h3>Prototype: Mercury5D<a class="headerlink" href="#prototype-mercury5d" title="Link to this heading"></a></h3>
<p><em>Audience request</em></p>
<p>This is a prototype 2D/5D equatorial Mercury run, with a foreshock. Example run can be picked up for restarting. These runs have proved to be somewhat tricky, and proper treatment of Mercury will require some code extensions (and the coders to do that coding!). What would be required can be discussed in breakout session.</p>
<p>Find this run from:
<code class="docutils literal notranslate"><span class="pre">/pfs/lustrep2/scratch/project_465000693/example_runs/Mercury5D</span></code></p>
</section>
<section id="other-practical-aspects">
<h3>Other practical aspects<a class="headerlink" href="#other-practical-aspects" title="Link to this heading"></a></h3>
<p>The rest of the day we get to play with running Vlasiator! Feel free to scatter, discuss, and join the online workshopping rooms, see HackMD for specifics.</p>
</section>
<section id="interesting-questions-you-might-get">
<h3>Interesting questions you might get<a class="headerlink" href="#interesting-questions-you-might-get" title="Link to this heading"></a></h3>
</section>
<section id="typical-pitfalls">
<h3>Typical pitfalls<a class="headerlink" href="#typical-pitfalls" title="Link to this heading"></a></h3>
</section>
</section>
<span id="document-vlasiator_outputs"></span><section id="vlasiator-output-data">
<h2>Vlasiator output data<a class="headerlink" href="#vlasiator-output-data" title="Link to this heading"></a></h2>
<section id="why-we-teach-this-lesson">
<h3>Why we teach this lesson<a class="headerlink" href="#why-we-teach-this-lesson" title="Link to this heading"></a></h3>
<p>Vlasiator has several grids and types of variables. Here we introduce them and go through which ones you might want to use for diagnostics and which one for sciencing.</p>
</section>
<section id="intended-learning-outcomes">
<h3>Intended learning outcomes<a class="headerlink" href="#intended-learning-outcomes" title="Link to this heading"></a></h3>
<p>You will be familiar with the outputs generated by Vlasiator and will understand which files and structures contain which information.</p>
</section>
<section id="three-kinds-of-data">
<h3>Three kinds of data<a class="headerlink" href="#three-kinds-of-data" title="Link to this heading"></a></h3>
<p>Vlasiator produces three kinds of output files during a simulation run,
the contents of which vary based on simulation parameters:</p>
<p>1. <code class="docutils literal notranslate"><span class="pre">logfile.txt</span></code>, the simulation run log. This is a timestamped ascii
file providing basic diagnostic output of the run, including memory
usage, time steps, counts of spatial cells at refinement levels etc.</p>
<p>2. <code class="docutils literal notranslate"><span class="pre">diagnostic.txt</span></code>. The contents of this file
can be configured by the <code class="docutils literal notranslate"><span class="pre">diagnostic</span> <span class="pre">=</span></code> options in the run config
file. In general, this ascii file will contain one line per (1, 10, or
so) simulation timesteps, with the columns determined by the selected
data reducers. These include, for example, simple scalar values like
overall plasma mass, number of velocity space blocks in the simulation,
maximum time step allowed by each solver, mass loss due to sparsity etc.</p>
<p>3. <strong>VLSV files</strong> are
the main output data products. These files come in multiple varieties:</p>
<p>* Restart files. These checkpoint contain the whole simulation state, including
the full phase space density, all relevant electromagnetic fields and
metadata. Simulations can be restarted from them (hence the name), but
they tend to be very heavy, easily multiple terabytes in size for
production runs. They do not contain the output of data reducer
operators (detailed below).</p>
<p>* Bulk files. In these, reduced spatial
simulation data is written for further scientific analysis. Usually,
this includes moments of the distribution functions and electromagnetic
fields, but can also contain much more complex data reducer operators,
as listed below. It is also possible (and common) to configure a subset
(e.g. every 25th cell) of the velocity distribution functions to be
written for further analysis.</p>
<p>N.b. saving FSgrid variables for large 3D runs can lead to significant
disk space usage, due to the FSgrid being uniform at the highest resolution
setting. For this reason, storing FSgrid variables in bulk files should
be carefully considered. It is also possible to declare several different
bulk file settings, one of which can be defined to exclude FSgrid variables
and be output more often, with the FSgrid-variable including version output
only e.g. every 10 seconds.</p>
</section>
<section id="the-vlsv-file-format">
<h3>The VLSV file format<a class="headerlink" href="#the-vlsv-file-format" title="Link to this heading"></a></h3>
<p>The <a class="reference external" href="https://github.com/fmihpc/vlsv">VLSV library</a> is used to write
this versatile container format.
<a class="reference external" href="https://github.com/fmihpc/analysator">Analysator</a> can be used to
load and handle these files in python.</p>
<p>The file format is optimized for parallel write performance: Data is
dumped to disk in the same memory structure as it is in the Vlasiator
simulation, as binary blobs. Once all data is written, an XML footer
that describes the data gets added to the end.</p>
<p>An example XML footer might look like this:</p>
<div class="highlight-xml notranslate"><div class="highlight"><pre><span></span><span class="nt">&lt;VLSV&gt;</span>
<span class="w">   </span><span class="nt">&lt;MESH</span><span class="w"> </span><span class="na">arraysize=</span><span class="s">&quot;208101&quot;</span><span class="w"> </span><span class="na">datasize=</span><span class="s">&quot;8&quot;</span><span class="w"> </span><span class="na">datatype=</span><span class="s">&quot;uint&quot;</span><span class="w"> </span><span class="na">max_refinement_level=</span><span class="s">&quot;1&quot;</span><span class="w"> </span><span class="na">name=</span><span class="s">&quot;SpatialGrid&quot;</span><span class="w"> </span><span class="na">type=</span><span class="s">&quot;amr_ucd&quot;</span><span class="w"> </span><span class="na">vectorsize=</span><span class="s">&quot;1&quot;</span><span class="w"> </span><span class="na">xperiodic=</span><span class="s">&quot;no&quot;</span><span class="w"> </span><span class="na">yperiodic=</span><span class="s">&quot;no&quot;</span><span class="w"> </span><span class="na">zperiodic=</span><span class="s">&quot;no&quot;</span><span class="nt">&gt;</span>989580<span class="nt">&lt;/MESH&gt;</span>
<span class="w">   </span><span class="nt">&lt;MESH</span><span class="w"> </span><span class="na">arraysize=</span><span class="s">&quot;652800&quot;</span><span class="w"> </span><span class="na">datasize=</span><span class="s">&quot;8&quot;</span><span class="w"> </span><span class="na">datatype=</span><span class="s">&quot;uint&quot;</span><span class="w"> </span><span class="na">name=</span><span class="s">&quot;fsgrid&quot;</span><span class="w"> </span><span class="na">type=</span><span class="s">&quot;multi_ucd&quot;</span><span class="w"> </span><span class="na">vectorsize=</span><span class="s">&quot;1&quot;</span><span class="w"> </span><span class="na">xperiodic=</span><span class="s">&quot;no&quot;</span><span class="w"> </span><span class="na">yperiodic=</span><span class="s">&quot;no&quot;</span><span class="w"> </span><span class="na">zperiodic=</span><span class="s">&quot;no&quot;</span><span class="nt">&gt;</span>4011008<span class="nt">&lt;/MESH&gt;</span>
<span class="w">   </span><span class="nt">&lt;PARAMETER</span><span class="w"> </span><span class="na">arraysize=</span><span class="s">&quot;1&quot;</span><span class="w"> </span><span class="na">datasize=</span><span class="s">&quot;8&quot;</span><span class="w"> </span><span class="na">datatype=</span><span class="s">&quot;float&quot;</span><span class="w"> </span><span class="na">name=</span><span class="s">&quot;time&quot;</span><span class="w"> </span><span class="na">vectorsize=</span><span class="s">&quot;1&quot;</span><span class="nt">&gt;</span>989488<span class="nt">&lt;/PARAMETER&gt;</span>
<span class="w">   </span><span class="nt">&lt;PARAMETER</span><span class="w"> </span><span class="na">arraysize=</span><span class="s">&quot;1&quot;</span><span class="w"> </span><span class="na">datasize=</span><span class="s">&quot;8&quot;</span><span class="w"> </span><span class="na">datatype=</span><span class="s">&quot;float&quot;</span><span class="w"> </span><span class="na">name=</span><span class="s">&quot;dt&quot;</span><span class="w"> </span><span class="na">vectorsize=</span><span class="s">&quot;1&quot;</span><span class="nt">&gt;</span>989496<span class="nt">&lt;/PARAMETER&gt;</span>
<span class="w">   </span><span class="nt">&lt;VARIABLE</span><span class="w"> </span><span class="na">arraysize=</span><span class="s">&quot;123544&quot;</span><span class="w"> </span><span class="na">datasize=</span><span class="s">&quot;8&quot;</span><span class="w"> </span><span class="na">datatype=</span><span class="s">&quot;uint&quot;</span><span class="w"> </span><span class="na">mesh=</span><span class="s">&quot;SpatialGrid&quot;</span><span class="w"> </span><span class="na">name=</span><span class="s">&quot;CellID&quot;</span><span class="w"> </span><span class="na">vectorsize=</span><span class="s">&quot;1&quot;</span><span class="nt">&gt;</span>1136<span class="nt">&lt;/VARIABLE&gt;</span>
<span class="w">   </span><span class="nt">&lt;VARIABLE</span><span class="w"> </span><span class="na">arraysize=</span><span class="s">&quot;652800&quot;</span><span class="w"> </span><span class="na">datasize=</span><span class="s">&quot;8&quot;</span><span class="w"> </span><span class="na">datatype=</span><span class="s">&quot;float&quot;</span><span class="w"> </span><span class="na">mesh=</span><span class="s">&quot;fsgrid&quot;</span><span class="w"> </span><span class="na">name=</span><span class="s">&quot;fg_b&quot;</span><span class="w"> </span><span class="na">unit=</span><span class="s">&quot;T&quot;</span><span class="w"> </span><span class="na">unitConversion=</span><span class="s">&quot;1.0&quot;</span><span class="w"> </span><span class="na">unitLaTeX=</span><span class="s">&quot;$\mathrm{T}$&quot;</span><span class="w"> </span><span class="na">variableLaTeX=</span><span class="s">&quot;$B$&quot;</span><span class="w"> </span><span class="na">vectorsize=</span><span class="s">&quot;3&quot;</span><span class="nt">&gt;</span>9558184<span class="nt">&lt;/VARIABLE&gt;</span>
<span class="nt">&lt;/VLSV&gt;</span>
</pre></div>
</div>
<p>Each XML tag describes one dataset in the file, with <code class="docutils literal notranslate"><span class="pre">arraysize</span></code>,
<code class="docutils literal notranslate"><span class="pre">datatype</span></code>, <code class="docutils literal notranslate"><span class="pre">datasize</span></code> and <code class="docutils literal notranslate"><span class="pre">vectorsize</span></code> describing the array. The
XML tag’s content contains the byte offset in the file, where this
dataset’s raw binary data lies.</p>
<p>The two most important tag types are <code class="docutils literal notranslate"><span class="pre">PARAMETER</span></code>, for single numbers
describing the file as a whole, such as resolutions, timesteps etc., and
<code class="docutils literal notranslate"><span class="pre">VARIABLE</span></code>, for spatially varying data reducer data maps.</p>
<p>Additional metadata is often added to the datasets, such as their
physical units, LaTeX formatted plotting hints, etc.</p>
<p>It can sometimes be useful to use the command line to look directly at
the XML footer information which contains information on all variables
included in the file, e.g. <code class="docutils literal notranslate"><span class="pre">tail</span> <span class="pre">-n</span> <span class="pre">60</span> <span class="pre">bulk.0001234.vlsv</span> <span class="pre">|less</span></code>. You
can adjust the line count until you have the information you need.
Adding too many lines will result in human-unreadable binary output.</p>
</section>
<section id="spatial-ordering-vlasov-vs-fsgrid-vs-velocity-space-variables">
<h3>Spatial ordering: Vlasov- vs. FSGrid vs. Velocity space variables<a class="headerlink" href="#spatial-ordering-vlasov-vs-fsgrid-vs-velocity-space-variables" title="Link to this heading"></a></h3>
<p>Note that the XML tags in the file do not yet give sufficient
information to describe the spatial structure of the variable arrays.
The construction differs depending on the grid they are linked to
(denoted by the <code class="docutils literal notranslate"><span class="pre">mesh=</span></code> attribute):</p>
<ul>
<li><p><strong>Vlasov grid variables</strong>,
typically marked with a <code class="docutils literal notranslate"><span class="pre">vg_</span></code> in their name, are stored as cell
parameters in the <a class="reference external" href="https://github.com/fmihpc/dccrg">DCCRG</a> grid
underlying the vlasov solver. As the simulation is dynamically load
balanced, their memory order changes unpredictably, so the data must be
presumed completely unordered in the file.</p>
<p>Fortunately, the <code class="docutils literal notranslate"><span class="pre">CellID</span></code> variable gets written into the file first,
which contains the flattened spatial index of the given simulation cells
in the same order as all further Vlasov grid variables. In the simplest,
non mesh-refined version, the CellID is defined as</p>
<p><code class="docutils literal notranslate"><span class="pre">CellID</span> <span class="pre">=</span> <span class="pre">x_index</span> <span class="pre">+</span> <span class="pre">x_size</span> <span class="pre">*</span> <span class="pre">y_index</span> <span class="pre">+</span> <span class="pre">x_size</span> <span class="pre">*</span> <span class="pre">y_size</span> <span class="pre">*</span> <span class="pre">z_index</span> <span class="pre">+</span> <span class="pre">1</span></code></p>
<p>By reading both the intended target variable and the CellID, the data
can thus be brought into flattened spatial order by simply sorting both
arrays in the same order. In analysator, this is typically achieved by
running</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">c</span> <span class="o">=</span> <span class="n">file</span><span class="o">.</span><span class="n">read_variable</span><span class="p">(</span><span class="s2">&quot;CellID&quot;</span><span class="p">)</span>
<span class="n">b</span> <span class="o">=</span> <span class="n">file</span><span class="o">.</span><span class="n">read_variable</span><span class="p">(</span><span class="s2">&quot;rho&quot;</span><span class="p">)</span>
<span class="n">b</span> <span class="o">=</span> <span class="n">b</span><span class="p">[</span><span class="n">numpy</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">c</span><span class="p">)]</span>
<span class="n">b</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">get_spatial_mesh_size</span><span class="p">())</span>
</pre></div>
</div>
</li>
<li><p><strong>FSGrid variables</strong> are stored on the simulations <a class="reference external" href="https://github.com/fmihpc/fsgrid">fieldsolver
grid</a>, which is partitioned
quite differently for performance reasons. The spatial domain is
subdivided into equally sized rectangular domains, which are written
for each compute rank in parallel. If written from a simulation with
a single MPI rank, the resulting array is directly ordered in spatial
order, as by the cellID definition above. For simulations on multiple
ranks, every rank writes its data in this structure, end-to-end. The
<code class="docutils literal notranslate"><span class="pre">num_writing_ranks</span></code> and the <code class="docutils literal notranslate"><span class="pre">MESH_DECOMPOSITION</span></code> arguments in the
XML tag allow the spatial partition to be reconstructed on load time.</p></li>
<li><p><strong>Ionospheric grid variables</strong> are stored on the simulations <cite>ionosphere
grid</cite>, which is a statically refined triangular mesh designed for solving
ionospheric potentials.</p></li>
<li><p><strong>Velocity space variables</strong> (at the moment, this is only the phase
space density f for every species), follow yet another structure due
to the sparse velocity grid structure on which they are stored.</p></li>
</ul>
</section>
<section id="simulation-data-reducers">
<h3>Simulation data reducers<a class="headerlink" href="#simulation-data-reducers" title="Link to this heading"></a></h3>
<p>This is a (mostly) up-to date list of simulation output options
that can be enabled in the config file. Note that older simulation
possibly use slightly different names, as the code is in constant
development.</p>
<table class="docutils align-default" id="id1">
<caption><span class="caption-text">Vlasiator outputs</span><a class="headerlink" href="#id1" title="Link to this table"></a></caption>
<colgroup>
<col style="width: 19.0%" />
<col style="width: 19.0%" />
<col style="width: 4.8%" />
<col style="width: 38.1%" />
<col style="width: 19.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Variable name</p></th>
<th class="head"><p>config option</p></th>
<th class="head"><p>unit</p></th>
<th class="head"><p>meaning</p></th>
<th class="head"><p>literature ref</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>CellID</p></td>
<td><p>always written</p></td>
<td><p>cells</p></td>
<td><p>Spatial ordering of vlasov grid cells</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>fg_b</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">fg_b</span></code></p></td>
<td><p>T</p></td>
<td><p>Overall magnetic field (vector)</p></td>
<td><p>[Palmroth et al. 2018](<a class="reference external" href="https://link.springer.com/article/10.1007%2Fs41115-018-0003-2">https://link.springer.com/article/10.1007%2Fs41115-018-0003-2</a>)</p></td>
</tr>
<tr class="row-even"><td><p>fg_b_background</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">fg_backgroundb</span></code></p></td>
<td><p>T</p></td>
<td><p>Static background magnetic field (i.e. dipole field in a magnetosphere simulation. Vector.)</p></td>
<td><p>[Palmroth et al. 2018](<a class="reference external" href="https://link.springer.com/article/10.1007%2Fs41115-018-0003-2">https://link.springer.com/article/10.1007%2Fs41115-018-0003-2</a>)</p></td>
</tr>
<tr class="row-odd"><td><p>fg_b_perturbed</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">fg_perturbedb</span></code></p></td>
<td><p>T</p></td>
<td><p>Fluctuating component of the magnetic field (vector)</p></td>
<td><p>[Palmroth et al. 2018](<a class="reference external" href="https://link.springer.com/article/10.1007%2Fs41115-018-0003-2">https://link.springer.com/article/10.1007%2Fs41115-018-0003-2</a>)</p></td>
</tr>
<tr class="row-even"><td><p>fg_e</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">fg_e</span></code></p></td>
<td><p>V/m</p></td>
<td><p>Electric field calculated as ∇ × B (Vector)</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>vg_rhom</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">vg_rhom</span></code></p></td>
<td><p>kg/m³</p></td>
<td><p>combined mass density of all simulation species</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>fg_rhom</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">fg_rhom</span></code></p></td>
<td><p>kg/m³</p></td>
<td><p>-‘’-</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>vg_rhoq</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">vg_rhoq</span></code></p></td>
<td><p>C/m³</p></td>
<td><p>combined charge density of all simulation species</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>fg_rhoq</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">fg_rhoq</span></code></p></td>
<td><p>C/m³</p></td>
<td><p>-‘’-</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>proton_vg_rho</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">populations_rho</span></code></p></td>
<td><p>1/m³</p></td>
<td><p>Number density for each simulated particle population</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>vg_v</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">vg_v</span></code></p></td>
<td><p>m/s</p></td>
<td><p>Bulk plasma velocity (velocity of the centre-of-mass frame vector)</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>fg_v</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">fg_v</span></code></p></td>
<td><p>m/s</p></td>
<td><p>-‘’-</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>proton_vg_v</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">populations_v</span></code></p></td>
<td><p>m/s</p></td>
<td><p>Per-population bulk velocity</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>proton_vg_rho_thermal</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">populations_moments_thermal</span></code></p></td>
<td><p>1/m³</p></td>
<td><p>Number density for the <strong>thermal</strong> component of every population</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>proton_vg_v_thermal</p></td>
<td><p>-‘’-</p></td>
<td><p>m/s</p></td>
<td><p>Velocity (vector) for the <strong>thermal</strong> component of every population</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>proton_vg_ptensor_diagonal_thermal</p></td>
<td><p>-‘’-</p></td>
<td><p>Pa</p></td>
<td><p>Diagonal components of the pressure tensor for the <strong>thermal</strong> component of every population</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>proton_vg_ptensor_offdiagonal_thermal</p></td>
<td><p>-‘’-</p></td>
<td><p>Pa</p></td>
<td><p>Off-Diagonal components of the pressure tensor for the <strong>thermal</strong> component of every population</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>proton_vg_rho_nonthermal</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">populations_moments_nonthermal</span></code></p></td>
<td><p>1/m³</p></td>
<td><p>Number density for the <strong>nonthermal</strong> component of every population</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>proton_vg_v_nonthermal</p></td>
<td><p>-‘’-</p></td>
<td><p>m/s</p></td>
<td><p>Velocity (vector) for the <strong>nonthermal</strong> component of every population</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>proton_vg_ptensor_diagonal_nonthermal</p></td>
<td><p>-‘’-</p></td>
<td><p>Pa</p></td>
<td><p>Diagonal components of the pressure tensor for the <strong>nonthermal</strong> component of every population</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>proton_vg_ptensor_offdiagonal_nonthermal</p></td>
<td><p>-‘’-</p></td>
<td><p>Pa</p></td>
<td><p>Off-Diagonal components of the pressure tensor for the <strong>nonthermal</strong> component of every population</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>proton_minvalue</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">populations_vg_effectivesparsitythreshold</span></code></p></td>
<td><p>m^6/s³</p></td>
<td><p>Effective sparsity threshold for every cell.</p></td>
<td><p>[Yann’s PhD Thesis](<a class="reference external" href="http://urn.fi/URN:ISBN:978-952-336-001-3">http://urn.fi/URN:ISBN:978-952-336-001-3</a>) page 91</p></td>
</tr>
<tr class="row-even"><td><p>proton_rholossadjust</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">populations_vg_rho_loss_adjust</span></code></p></td>
<td><p>1/m³</p></td>
<td><p>Tracks how much mass was lost in the sparse velocity space block removal</p></td>
<td><p>[Yann’s PhD Thesis](<a class="reference external" href="http://urn.fi/URN:ISBN:978-952-336-001-3">http://urn.fi/URN:ISBN:978-952-336-001-3</a>) page 90</p></td>
</tr>
<tr class="row-odd"><td><p>vg_lbweight</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">vg_lbweight</span></code></p></td>
<td><p>arb. unit</p></td>
<td><p>Load balance metric</p></td>
<td><p>used for dynamic rebalancing of computational load between mpi tasks</p></td>
</tr>
<tr class="row-even"><td><p>vg_maxdt_acceleration</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">vg_maxdt_acceleration</span></code></p></td>
<td><p>s</p></td>
<td><p>Maximum timestep limit of the acceleration solver</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>proton_vg_maxdt_acceleration</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">populations_vg_maxdt_acceleration</span></code></p></td>
<td><p>s</p></td>
<td><p>-‘’-</p></td>
<td><p>per-population</p></td>
</tr>
<tr class="row-even"><td><p>vg_maxdt_translation</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">vg_maxdt_translation</span></code></p></td>
<td><p>s</p></td>
<td><p>Maximum timestep limit of the translation solver</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>proton_vg_maxdt_translation</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">populations_vg_maxdt_translation</span></code></p></td>
<td><p>s</p></td>
<td><p>-‘’-</p></td>
<td><p>per-population</p></td>
</tr>
</tbody>
</table>
</section>
<section id="vlsv-data-tools">
<h3>VLSV data tools<a class="headerlink" href="#vlsv-data-tools" title="Link to this heading"></a></h3>
<p>A short note on the included tools, compiled by:</p>
<p><code class="docutils literal notranslate"><span class="pre">make</span> <span class="pre">vlsvextract</span> <span class="pre">vlsvdiff</span></code></p>
<p>Some older tools included in <code class="docutils literal notranslate"><span class="pre">make</span> <span class="pre">tools</span></code> are not currently supported.</p>
<section id="vlsvextract">
<h4>vlsvextract<a class="headerlink" href="#vlsvextract" title="Link to this heading"></a></h4>
<p><code class="docutils literal notranslate"><span class="pre">vlsvextract</span></code> can be used to extract VDF data from vlsv files and store it as a separate VLSV file for visualization.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span> <span class="n">USAGE</span><span class="p">:</span> <span class="o">./</span><span class="n">vlsvextract_DP</span> <span class="o">&lt;</span><span class="n">file</span> <span class="n">name</span> <span class="n">mask</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">options</span><span class="o">&gt;</span>

 <span class="n">To</span> <span class="n">get</span> <span class="n">a</span> <span class="nb">list</span> <span class="n">of</span> <span class="n">options</span> <span class="n">use</span> <span class="o">--</span><span class="n">help</span>

 <span class="n">Options</span><span class="p">:</span>
<span class="o">--</span><span class="n">help</span>                <span class="n">display</span> <span class="n">help</span>
<span class="o">--</span><span class="n">debug</span>               <span class="n">write</span> <span class="n">debugging</span> <span class="n">info</span> <span class="n">to</span> <span class="n">stderr</span>
<span class="o">--</span><span class="n">cellid</span> <span class="n">arg</span>          <span class="n">Set</span> <span class="n">cell</span> <span class="nb">id</span>
<span class="o">--</span><span class="n">cellidlist</span> <span class="n">arg</span>      <span class="n">Set</span> <span class="nb">list</span> <span class="n">of</span> <span class="n">cell</span> <span class="n">ids</span>
<span class="o">--</span><span class="n">rotate</span>              <span class="n">Rotate</span> <span class="n">velocities</span> <span class="n">so</span> <span class="n">that</span> <span class="n">they</span> <span class="n">face</span> <span class="n">z</span><span class="o">-</span><span class="n">axis</span>
<span class="o">--</span><span class="n">plasmaFrame</span>         <span class="n">Shift</span> <span class="n">the</span> <span class="n">distribution</span> <span class="n">so</span> <span class="n">that</span> <span class="n">the</span> <span class="n">bulk</span> <span class="n">velocity</span> <span class="ow">is</span> <span class="mi">0</span>
<span class="o">--</span><span class="n">coordinates</span> <span class="n">arg</span>     <span class="n">Set</span> <span class="n">spatial</span> <span class="n">coordinates</span> <span class="n">x</span> <span class="n">y</span> <span class="n">z</span>
<span class="o">--</span><span class="n">unit</span> <span class="n">arg</span>            <span class="n">Sets</span> <span class="n">the</span> <span class="n">units</span><span class="o">.</span> <span class="n">Options</span><span class="p">:</span> <span class="n">re</span><span class="p">,</span> <span class="n">km</span><span class="p">,</span> <span class="n">m</span> <span class="p">(</span><span class="n">OPTIONAL</span><span class="p">)</span>
<span class="o">--</span><span class="n">point1</span> <span class="n">arg</span>          <span class="n">Set</span> <span class="n">the</span> <span class="n">starting</span> <span class="n">point</span> <span class="n">x</span> <span class="n">y</span> <span class="n">z</span> <span class="n">of</span> <span class="n">a</span> <span class="n">line</span>
<span class="o">--</span><span class="n">point2</span> <span class="n">arg</span>          <span class="n">Set</span> <span class="n">the</span> <span class="n">ending</span> <span class="n">point</span> <span class="n">x</span> <span class="n">y</span> <span class="n">z</span> <span class="n">of</span> <span class="n">a</span> <span class="n">line</span>
<span class="o">--</span><span class="n">pointamount</span> <span class="n">arg</span>     <span class="n">Number</span> <span class="n">of</span> <span class="n">points</span> <span class="n">along</span> <span class="n">a</span> <span class="n">line</span> <span class="p">(</span><span class="n">OPTIONAL</span><span class="p">)</span>
<span class="o">--</span><span class="n">outputdirectory</span> <span class="n">arg</span> <span class="n">The</span> <span class="n">directory</span> <span class="n">where</span> <span class="n">the</span> <span class="n">file</span> <span class="ow">is</span> <span class="n">saved</span> <span class="p">(</span><span class="n">default</span> <span class="n">current</span> <span class="n">folder</span><span class="p">)</span> <span class="p">(</span><span class="n">OPTIONAL</span><span class="p">)</span>
</pre></div>
</div>
<p>For example, let’s pick a VDF from the foreshock of the Mercury 5D example run; see VisIt lecture one method on how we can find the cellID, here we have a cellID pre-picked.</p>
<p><code class="docutils literal notranslate"><span class="pre">./vlsvextract_DP</span> <span class="pre">/scratch/project_465000693/example_runs/Mercury5D/bulk/bulk.0000122.vlsv</span> <span class="pre">--cellid</span> <span class="pre">332776</span></code></p>
<p>This can be used to extract VDFs over lines and multiple files as well.</p>
</section>
<section id="vlsvdiff">
<h4>vlsvdiff<a class="headerlink" href="#vlsvdiff" title="Link to this heading"></a></h4>
<p><code class="docutils literal notranslate"><span class="pre">vlsvdiff</span></code> we use for e.g. continuous integration testing. There is an included testpackage, from which one can generate reference data and compare the effects of one’s code edits locally.</p>
<p>Other use is to extract differences between different files - for example, time differences.</p>
</section>
</section>
<section id="other-output-files">
<h3>Other output files<a class="headerlink" href="#other-output-files" title="Link to this heading"></a></h3>
<p>If the PHIPROF profiler suite is in use, you will also see e.g. <code class="docutils literal notranslate"><span class="pre">phiprof_0.txt</span></code> in the run directory,
providing rough ASCII tables of run-time timers, useful for rudimentary profiling of the Vlasiator
code, solvers, and I/O.</p>
</section>
<section id="interesting-questions-you-might-get">
<h3>Interesting questions you might get<a class="headerlink" href="#interesting-questions-you-might-get" title="Link to this heading"></a></h3>
<p>Q: Why are the output formats so convoluted?</p>
<p>A: They are optimized for run-time performance, so that each MPI task can simply pour its data into
one contiguous region on-disk via MPI writes.</p>
</section>
<section id="typical-pitfalls">
<h3>Typical pitfalls<a class="headerlink" href="#typical-pitfalls" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>Read Vlasov grid data and forget the order the cells based on CELLIDS</p></li>
<li><p>Read FSGrid data and accidentally order that also according to CELLIDS</p></li>
</ul>
</section>
</section>
<span id="document-visit"></span><section id="visit-and-vlsv-bootcamp">
<h2>VisIt and VLSV bootcamp<a class="headerlink" href="#visit-and-vlsv-bootcamp" title="Link to this heading"></a></h2>
<section id="why-we-teach-this-lesson">
<h3>Why we teach this lesson<a class="headerlink" href="#why-we-teach-this-lesson" title="Link to this heading"></a></h3>
<p><a class="reference external" href="https://visit-dav.github.io/visit-website/index.html">VisIt</a> is a scalable 3D-visualization software, suitable for supercomputing environments. It also has a home-brewed plugin to read and plot .vlsv files, which is why we prefer using it over similar alternatives such as ParaView. In this lesson, we look at configuring a VisIt client-server on LUMI, using VisIt and the .vlsv plugin for data exploration in 3D. VisIt and VLSV plugin are pre-installed on the LUMI workspace.</p>
</section>
<section id="intended-learning-outcomes">
<h3>Intended learning outcomes<a class="headerlink" href="#intended-learning-outcomes" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>Using a client-server with VisIt.</p></li>
<li><p>Basic exploration of .vlsv files</p>
<ul>
<li><p>SpatialGrid</p></li>
<li><p>FsGrid</p></li>
<li><p>Finding stored VDFs</p></li>
</ul>
</li>
<li><p>Plotting fieldlines and streamlines</p></li>
<li><p>Plotting contours</p></li>
<li><p>vlsvextract and plotting a VDF</p></li>
</ul>
<p>Maybe:</p>
<ul class="simple">
<li><p>Compiling and installing the .vlsv plugin</p></li>
</ul>
</section>
<section id="timing">
<h3>Timing<a class="headerlink" href="#timing" title="Link to this heading"></a></h3>
<p>Wednesday morning</p>
</section>
<section id="preparations-for-the-exercises">
<h3>Preparations for the exercises<a class="headerlink" href="#preparations-for-the-exercises" title="Link to this heading"></a></h3>
<p>If you haven’t yet done so, please:</p>
<ol class="arabic simple">
<li><p>Install <a class="reference external" href="https://visit-dav.github.io/visit-website/releases-as-tables/#series-33">VisIt 3.3</a> locally.</p></li>
<li><p>Download <a class="reference external" href="https://github.com/ENCCS/plasma-pepsc-workshop/raw/main/content/visit/host_lumi_pepsc.xml">host_lumi_pepsc.xml</a> and place it into your local <code class="docutils literal notranslate"><span class="pre">$HOME/.visit/hosts</span></code></p></li>
<li><p>Open VisIt, go to Options - Host profiles.. and change the Account to your username under the lumi-pepsc host.</p></li>
<li><p>From options, click <code class="docutils literal notranslate"><span class="pre">Save</span> <span class="pre">settings</span></code> so the username is saved to your config.</p></li>
</ol>
<p>Feel free to get to know the <a href="#id2"><span class="problematic" id="id3">`VisIt manuals&lt;https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/getting_started/index.html&gt;`_</span></a> as well!</p>
</section>
<section id="the-hands-on">
<h3>The hands-on<a class="headerlink" href="#the-hands-on" title="Link to this heading"></a></h3>
<ol class="arabic simple">
<li><p>Launch your local VisIt</p></li>
</ol>
<p>VisIt produces a plentiful amount of windows.</p>
<p>The main one (<code class="docutils literal notranslate"><span class="pre">gui</span></code>) is the tall one with plotting tools, from database list on the top, a time slider, and a plotting pipeline window, currently empty.</p>
<p>The other one you encounter by default is the <code class="docutils literal notranslate"><span class="pre">viewer</span></code> window. This will render your plots and let you navigate them - see the toolbar on the top for navigation, zooming, saving viewpoints, etc. You can have multiple windows as well, and there are handy layout buttons available.</p>
<section id="launching-to-client-server">
<h4>Launching to client-server<a class="headerlink" href="#launching-to-client-server" title="Link to this heading"></a></h4>
<ol class="arabic simple">
<li><p>Click “Open”</p></li>
<li><p>Choose <em>Host</em> <code class="docutils literal notranslate"><span class="pre">lumi-pepsc</span></code>, as given by the host configuration</p>
<ul class="simple">
<li><p>This opens a VisIt metadata server on the frontend.</p></li>
</ul>
</li>
<li><p>Navigate <em>Path</em> to <code class="docutils literal notranslate"><span class="pre">/pfs/lustrep2/scratch/project_465000693/example_runs/Mercury5D/bulk</span></code></p></li>
<li><p>With file grouping at on/smart, open the bulk files as a database</p></li>
<li><p>A <em>Compute engine launch</em> prompt appears. Launch one on <code class="docutils literal notranslate"><span class="pre">small</span></code>, adjust cpu counts if needed.</p>
<ul class="simple">
<li><p>Might take a bit to queue…</p></li>
<li><p>A larger number of cores helps esp. with loading data from the .vlsv files!</p></li>
</ul>
</li>
</ol>
</section>
<section id="while-we-wait-a-bit">
<h4>While we wait a bit…<a class="headerlink" href="#while-we-wait-a-bit" title="Link to this heading"></a></h4>
<p>VisIt works by handling 3D data in pipelines, with two noteworthy concepts. Firstly, we can consider the mesh: the spatial structure that tells us where we have data, and how these data points are connected. In other words, this contains the geometry and topology of the dataset. We can extract new meshes from the existing ones via geometric operations, such as slicing. These operations can also be chained, and at the end we choose what data we actually want show on the mesh at the end of the pipeline. We’ll see plenty of examples soon!</p>
</section>
<section id="first-plots">
<h4>First plots<a class="headerlink" href="#first-plots" title="Link to this heading"></a></h4>
<p>Mostly, we’ll be using pseudocolor plots. Let’s get a feel for the 2D dataset!</p>
<ol class="arabic simple">
<li><p>Set the time slider to the end</p></li>
<li><p>Add a Pseudocolor plot -&gt; <code class="docutils literal notranslate"><span class="pre">proton/vg_rho</span></code> to get the proton number density</p></li>
<li><p>Notice we have now the full domain plotted - including the 3rd dimension, which is included even in the 2D/5D runs.</p></li>
</ol>
<figure class="align-default">
<a class="reference internal image-reference" href="_images/2d_0_rho.png"><img alt="_images/2d_0_rho.png" src="_images/2d_0_rho.png" style="width: 600px;" /></a>
</figure>
<ol class="arabic simple">
<li><p>Select the plot and use the Operator button, navigate to Slicing -&gt; Slice</p>
<ul class="simple">
<li><p>Double-click on the new operator in the pipeline menu.</p></li>
<li><p>De-select “Project to 2D”</p></li>
<li><p>Set Normal axis as Z, with intercept at 0</p></li>
<li><p>Click Apply in the dialog</p></li>
<li><p>Click Draw in the main window!</p></li>
</ul>
</li>
</ol>
<figure class="align-default">
<a class="reference internal image-reference" href="_images/2d_1_slice.png"><img alt="_images/2d_1_slice.png" src="_images/2d_1_slice.png" style="width: 600px;" /></a>
</figure>
<p>Double-clicking on the operators or or the plots opens attribure windows for those objects. Feel free to e.g. adjust the Pseudocolor colormaps or variable ranges from Pseudocolor attributes!</p>
<p>Let’s identify the system boundaries next.</p>
</section>
<section id="vlasiator-boundaries">
<h4>Vlasiator boundaries<a class="headerlink" href="#vlasiator-boundaries" title="Link to this heading"></a></h4>
<ol class="arabic simple">
<li><p>Set the pseudocolor variable to <code class="docutils literal notranslate"><span class="pre">vg_boundarytype</span></code>.</p></li>
</ol>
<figure class="align-default">
<a class="reference internal image-reference" href="visit/2d_2_boundary.png"><img alt="visit/2d_2_boundary.png" src="visit/2d_2_boundary.png" style="width: 600px;" /></a>
</figure>
<p>Let’s compare that to the <a class="reference external" href="https://github.com/fmihpc/vlasiator/blob/676f26a5e74c4c2b40e6d5e3294c413da0157ac3/common.h#L450">sysboundarytype enum</a>:</p>
<div class="highlight-c++ notranslate"><div class="highlight"><pre><span></span><span class="k">namespace</span><span class="w"> </span><span class="nn">sysboundarytype</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="k">enum</span><span class="w"> </span><span class="p">{</span>
<span class="w">      </span><span class="n">DO_NOT_COMPUTE</span><span class="p">,</span><span class="w">   </span><span class="cm">/*!&lt; E.g. cells within the ionospheric outer radius should not be computed at all. */</span>
<span class="w">      </span><span class="n">NOT_SYSBOUNDARY</span><span class="p">,</span><span class="w">  </span><span class="cm">/*!&lt; Cells within the simulation domain are not boundary cells. */</span>
<span class="w">      </span><span class="n">IONOSPHERE</span><span class="p">,</span><span class="w">       </span><span class="cm">/*!&lt; Ionospheric current model. */</span>
<span class="w">      </span><span class="n">OUTFLOW</span><span class="p">,</span><span class="w">          </span><span class="cm">/*!&lt; No fixed conditions on the fields and distribution function. */</span>
<span class="w">      </span><span class="n">MAXWELLIAN</span><span class="p">,</span><span class="w">       </span><span class="cm">/*!&lt; Set Maxwellian boundary condition, i.e. set fields and distribution function. */</span>
<span class="w">      </span><span class="n">COPYSPHERE</span><span class="p">,</span><span class="w">       </span><span class="cm">/*!&lt; A sphere with copy-condition for perturbed B as the simple inner boundary */</span>
<span class="w">      </span><span class="n">OUTER_BOUNDARY_PADDING</span><span class="p">,</span><span class="w"> </span><span class="cm">/*!&lt; These cells only occur on FSGrid, where boundaries are not at the highest refinement level */</span>
<span class="w">      </span><span class="n">N_SYSBOUNDARY_CONDITIONS</span>
<span class="w">  </span><span class="p">};</span>
<span class="p">}</span>
</pre></div>
</div>
<p>We find here the <code class="docutils literal notranslate"><span class="pre">COPYSPHERE</span></code> (5) boundary and <code class="docutils literal notranslate"><span class="pre">DO_NOT_COMPUTE</span></code> (1) cells covering the planet, approximately, as the inner boundary. Then, we can focus on the actual simulation domain:</p>
<ol class="arabic simple">
<li><p>Add an operator to the plot: Selection -&gt; threshold</p></li>
<li><p>Open the threshold window, remove the “default” variable</p></li>
<li><p>Add <code class="docutils literal notranslate"><span class="pre">vg_boundarytype</span></code> as a threshold variable, set min and max to 1 (<code class="docutils literal notranslate"><span class="pre">NOT_SYSBOUNDARY</span></code>)</p></li>
<li><p>Click apply</p></li>
</ol>
<figure class="align-default">
<a class="reference internal image-reference" href="_images/2d_3_threshold.png"><img alt="_images/2d_3_threshold.png" src="_images/2d_3_threshold.png" style="width: 600px;" /></a>
</figure>
<p>Now you can change the variable to e.g. <code class="docutils literal notranslate"><span class="pre">proton/vg_rho</span></code>, without system boundaries confounding the plot.</p>
</section>
<section id="vector-plots">
<h4>Vector plots<a class="headerlink" href="#vector-plots" title="Link to this heading"></a></h4>
<p>Let’s look at the vector plot type. Add one of <code class="docutils literal notranslate"><span class="pre">vg_b_vol</span></code>, and click Draw. This probably looks very empty:</p>
<figure class="align-default">
<a class="reference internal image-reference" href="_images/2d_4_vector.png"><img alt="_images/2d_4_vector.png" src="_images/2d_4_vector.png" style="width: 600px;" /></a>
</figure>
<p>Let’s go to Vector plot attributes, Geometry tab, and unselect Scale by magnitude, Apply:</p>
<figure class="align-default">
<a class="reference internal image-reference" href="_images/2d_5_vector.png"><img alt="_images/2d_5_vector.png" src="_images/2d_5_vector.png" style="width: 600px;" /></a>
</figure>
</section>
<section id="picking">
<h4>Picking<a class="headerlink" href="#picking" title="Link to this heading"></a></h4>
<p>Let’s see how to find an interesting cell and its CellID with VisIt.</p>
<p>Let’s use the plot of <code class="docutils literal notranslate"><span class="pre">proton/vg_rho</span></code> as a reference value slice in the background. Add another pseudocolor plot of <code class="docutils literal notranslate"><span class="pre">vg_f_saved</span></code>, and add a Threshold operator to diplay only cells with vg_f_saved = 1. Draw, and we should have cells with VDFs stored visible on top of the background slice.</p>
<figure class="align-default">
<a class="reference internal image-reference" href="_images/2d_6_fsaved.png"><img alt="_images/2d_6_fsaved.png" src="_images/2d_6_fsaved.png" style="width: 600px;" /></a>
</figure>
<p>Zoom in to the foreshock, select the Zonal pick operator, and click on a cell that looks like it could have interesting dynamics:</p>
<figure class="align-default" id="id1">
<a class="reference internal image-reference" href="_images/visit_pick.png"><img alt="_images/visit_pick.png" src="_images/visit_pick.png" style="width: 200px;" /></a>
<figcaption>
<p><span class="caption-text">Pick operators in the VisIt viewer. Z for zonal, N for nodal. S for spreadsheet.</span><a class="headerlink" href="#id1" title="Link to this image"></a></p>
</figcaption>
</figure>
<p>The following Pick window should open, showing the picked coordinates and the plotted variable.</p>
<figure class="align-default">
<a class="reference internal image-reference" href="_images/visit-pick-window-default.png"><img alt="_images/visit-pick-window-default.png" src="_images/visit-pick-window-default.png" style="width: 400px;" /></a>
</figure>
<p>That is not yet very useful. Adding <code class="docutils literal notranslate"><span class="pre">CellID</span></code> to the query variables helps! We should get a large-ish number, like 332776.</p>
</section>
</section>
<section id="going-3d">
<h3>Going 3D<a class="headerlink" href="#going-3d" title="Link to this heading"></a></h3>
<p>Let’s add a dimension to our plotting, and inspect one of these VDFs. The file <code class="docutils literal notranslate"><span class="pre">/scratch/project_465000693/example_runs/Mercury5D/velgrid.332776.0000122.vlsv</span></code> contains an extracted VDF (with <code class="docutils literal notranslate"><span class="pre">vlsvextract</span></code>) from the foreshock area, open it, and add a new window!</p>
<figure class="align-default">
<a class="reference internal image-reference" href="visit/new_window.png"><img alt="visit/new_window.png" src="visit/new_window.png" style="width: 300px;" /></a>
</figure>
<p>Let’s start by plotting the full proton v-space mesh: Pseudocolor-&gt;proton</p>
<figure class="align-default">
<a class="reference internal image-reference" href="_images/vdf_0_pseudocolor.png"><img alt="_images/vdf_0_pseudocolor.png" src="_images/vdf_0_pseudocolor.png" style="width: 600px;" /></a>
</figure>
<p>This is now the outer edge of the VDF. We need to do something else if we want to have a look inside. Let’s add a Threeslice operator and Draw again.</p>
<figure class="align-default">
<a class="reference internal image-reference" href="_images/vdf_1_threeslice.png"><img alt="_images/vdf_1_threeslice.png" src="_images/vdf_1_threeslice.png" style="width: 600px;" /></a>
</figure>
<p>Quite a bit of structure there! But we still have the blocky v-space halo with values below the threshold. Let’s add an aptly-named thresholding operator:</p>
<figure class="align-default">
<a class="reference internal image-reference" href="_images/vdf_2_threshold.png"><img alt="_images/vdf_2_threshold.png" src="_images/vdf_2_threshold.png" style="width: 600px;" /></a>
</figure>
<p>Slices are good, but what if we want to have a more thorough view of the 3D structure of the VDF? Let’s remove the Threeslice operator from the plot, and draw. Now that we still have the threshold operator, we should see the outer edge of the VDF at the threshold value.</p>
<figure class="align-default">
<a class="reference internal image-reference" href="_images/vdf_3_just_threshold.png"><img alt="_images/vdf_3_just_threshold.png" src="_images/vdf_3_just_threshold.png" style="width: 600px;" /></a>
</figure>
<p>Try adjusting the threshold value e.g. to <code class="docutils literal notranslate"><span class="pre">1e-13</span></code>!</p>
<section id="a-proper-3d-run">
<h4>A proper 3D run<a class="headerlink" href="#a-proper-3d-run" title="Link to this heading"></a></h4>
<p>Next, let’s see what one of our old low-resolution 3D tests looked like. Open the database at <code class="docutils literal notranslate"><span class="pre">/scratch/project_465000693/example_data/EGE</span></code>.</p>
<p>Let’s start by getting a quick overview with a pseudocolor plot of <code class="docutils literal notranslate"><span class="pre">proton/vg_rho</span></code> once more, and add a threeslice operator.</p>
<figure class="align-default">
<a class="reference internal image-reference" href="_images/3d_0_threeslice.png"><img alt="_images/3d_0_threeslice.png" src="_images/3d_0_threeslice.png" style="width: 600px;" /></a>
</figure>
</section>
<section id="streamline-plots">
<h4>Streamline plots<a class="headerlink" href="#streamline-plots" title="Link to this heading"></a></h4>
<p>Let’s add some fieldlines! These are produced in VisIt through an IntegralCurve system, which can be a bit hard to get into. Add a new pseudocolor plot with the <em>variable</em>: operators-&gt;IntegralCurve-&gt;vg_b_vol:</p>
<figure class="align-default">
<a class="reference internal image-reference" href="_images/3d_1_integralcurves.png"><img alt="_images/3d_1_integralcurves.png" src="_images/3d_1_integralcurves.png" style="width: 600px;" /></a>
</figure>
<p>Double-clicking on the integralcurve operator let’s you adjust the seeding of the lines. Let’s do something like the following - a spherical region with some radius of ~5e7 meters:</p>
<figure class="align-default">
<a class="reference internal image-reference" href="_images/3d_2_integralcurves.png"><img alt="_images/3d_2_integralcurves.png" src="_images/3d_2_integralcurves.png" style="width: 600px;" /></a>
</figure>
</section>
</section>
<section id="other-practical-aspects">
<h3>Other practical aspects<a class="headerlink" href="#other-practical-aspects" title="Link to this heading"></a></h3>
<p>VisIt may crash from time to time. Save your session often!</p>
<p>There are plenty of levers and clever tricks to pull in VisIt, this is really just scratching the surface in a short time.</p>
</section>
<section id="typical-pitfalls">
<h3>Typical pitfalls<a class="headerlink" href="#typical-pitfalls" title="Link to this heading"></a></h3>
<p>Forgetting to click <code class="docutils literal notranslate"><span class="pre">Apply</span></code> or <code class="docutils literal notranslate"><span class="pre">Draw</span></code> buttons.</p>
<p>Not noticing the small button with arrows to display</p>
<p>Not saving your session often.</p>
</section>
</section>
<span id="document-analysator"></span><section id="introduction-to-analysator">
<h2>Introduction to Analysator<a class="headerlink" href="#introduction-to-analysator" title="Link to this heading"></a></h2>
<section id="why-we-teach-this-lesson">
<h3>Why we teach this lesson<a class="headerlink" href="#why-we-teach-this-lesson" title="Link to this heading"></a></h3>
<p>Here we introduce the use of Analysator python tools for plotting and accessing .vlsv data.</p>
<p>The Analysator tool package contains .vlsv file accessing routines, helper routines,
data post-processing libraries, and a handful of useful plotting scripts. The Analysator
files and routines are designed to act as a starting point for more involved python post-processing
of Vlasiator data, not as an end point, and as such, can and should be extended by the end-user.</p>
</section>
<section id="intended-learning-outcomes">
<h3>Intended learning outcomes<a class="headerlink" href="#intended-learning-outcomes" title="Link to this heading"></a></h3>
<p>You shall learn how to open a .vlsv file in Analysator, how to view file contents, and how to perform
plotting of Vlasiator spatial data in 2D and 3D. You will practice accessing Vlasov grid, FSgrid, and
ionospheric grid data. You will learn about some of the most useful Analysator plotting options.</p>
</section>
<section id="timing">
<h3>Timing<a class="headerlink" href="#timing" title="Link to this heading"></a></h3>
</section>
<section id="preparing-exercises">
<h3>Preparing exercises<a class="headerlink" href="#preparing-exercises" title="Link to this heading"></a></h3>
<p>Clone the Analysator git repository onto the computer where you intend to use it.
Vlasiator .vlsv output files can be quite large, but at the same time you should have access to a
graphical interface on the machine.</p>
<p>For cloning, you can use the SSH or the HTTP method.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span><span class="w"> </span><span class="nv">$HOME</span>
git<span class="w"> </span>clone<span class="w"> </span>https://github.com/fmihpc/analysator.git
git<span class="w"> </span>clone<span class="w"> </span>git@github.com:fmihpc/analysator.git
</pre></div>
</div>
<p>Next, you should set your <code class="docutils literal notranslate"><span class="pre">PYTHONPATH</span></code> variable to include your Analysator install directory. One way is to add this line to your startup script:
<code class="docutils literal notranslate"><span class="pre">export</span> <span class="pre">PYTHONPATH=$PYTHONPATH:$HOME/analysator</span></code></p>
</section>
<section id="analysator-on-lumi-with-jupyterhub">
<h3>Analysator on LUMI with jupyterhub<a class="headerlink" href="#analysator-on-lumi-with-jupyterhub" title="Link to this heading"></a></h3>
<p>For this introductory course, we will be using analysator in an interactive manner through the web interface at
<a class="reference external" href="https://www.lumi.csc.fi">https://www.lumi.csc.fi</a></p>
<p>Log in and select a Jupyter session (<em>not</em> Jupyter for courses). Verify that you have the correct project selected (<code class="docutils literal notranslate"><span class="pre">project_465000693</span></code>) and the interactive partition. You may want to select more than just 1 CPU in order to have enough memory to open large files. In settings, select “advanced” and type the following into the window “Script to start”:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>module<span class="w"> </span>use<span class="w"> </span>/appl/local/csc/modulefiles/
module<span class="w"> </span>load<span class="w"> </span>pytorch/2.1
<span class="nb">export</span><span class="w"> </span><span class="nv">PTNOLATEX</span><span class="o">=</span><span class="m">1</span>
<span class="nb">export</span><span class="w"> </span><span class="nv">PYTHONPATH</span><span class="o">=</span><span class="nv">$PYTHONPATH</span>:<span class="nv">$HOME</span>/analysator:
</pre></div>
</div>
<p>N.B. Remember to include the semicolon at the end of the last line! Next, launch the Jupyter session. Once your job has queued and launched, you can launch the Jupyter session and verify operation by importing Analysator. Type the command in the window and execute with shift+Enter.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span> <span class="kn">import</span> <span class="nn">pytools</span> <span class="k">as</span> <span class="nn">pt</span>
<span class="n">Using</span> <span class="n">LaTeX</span> <span class="n">formatting</span>
<span class="n">Using</span> <span class="n">backend</span> <span class="n">module</span><span class="p">:</span><span class="o">//</span><span class="n">matplotlib_inline</span><span class="o">.</span><span class="n">backend_inline</span>
<span class="n">Using</span> <span class="n">matplotlib</span> <span class="n">version</span> <span class="mf">3.8.1</span>

<span class="p">[</span><span class="mi">2</span><span class="p">]:</span>
</pre></div>
</div>
</section>
<section id="analysator-required-packages-on-other-systems">
<h3>Analysator required packages on other systems<a class="headerlink" href="#analysator-required-packages-on-other-systems" title="Link to this heading"></a></h3>
<p>Analysator should work mostly equally well on both Python 2.7 and Python 3.x. However, versions prior to
Python 3.8 no longer receive security support and are thus not recommended. Use of iPython, jupyter, or
a similar interface is recommended for ease of use.</p>
<p>Analysator only requires a small selection of python libraries, namely matplotlib, scipy&gt;1.7 and numpy. On modern
systems, these should be pre-installed. Some legacy sections of Analysator also use MayaVi2 but those
are no longer updated or supported.</p>
<p>To verify the availability of required libraries, it is suggested to try importing Analysator in python:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">In</span> <span class="p">[</span><span class="mi">1</span><span class="p">]:</span> <span class="kn">import</span> <span class="nn">pytools</span> <span class="k">as</span> <span class="nn">pt</span>
<span class="n">Using</span> <span class="n">LaTeX</span> <span class="n">formatting</span>
<span class="n">Using</span> <span class="n">backend</span> <span class="n">module</span><span class="p">:</span><span class="o">//</span><span class="n">matplotlib_inline</span><span class="o">.</span><span class="n">backend_inline</span>
<span class="n">Using</span> <span class="n">matplotlib</span> <span class="n">version</span> <span class="mf">3.8.1</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">2</span><span class="p">]:</span>
</pre></div>
</div>
</section>
<section id="other-practical-aspects">
<h3>Other practical aspects<a class="headerlink" href="#other-practical-aspects" title="Link to this heading"></a></h3>
<p>A TeX Live installation (or similar) is recommended for formatting of plotting text. If one is not available
on the target system, output can be forced to use TeX-like markup supported directly by matplotlib.
This is achieved by setting the system variable <code class="docutils literal notranslate"><span class="pre">export</span> <span class="pre">PTNOLATEX=1</span></code>. This will negatively impact output
of e.g. bolded text, but is required on e.g. the LUMI web interface.</p>
<p>On systems without an x-windowing system such as compute nodes on a cluster (or if using it is
prohibitively slow due to e.g. network weather), Analysator can be set to ignore X-windowing and
use a non-interactive frontend by setting the system variable <code class="docutils literal notranslate"><span class="pre">export</span> <span class="pre">PTNONINTERACTIVE=1</span></code>. In this
case, outputs are generated into .png files and should be transferred to another system for viewing. This is the suggested approach when using a batch job to generate several images/frames in order to e.g. build a movie.</p>
<p>If necessary, the matplotlib frontend can be declared manually with a system variable,
for example, <code class="docutils literal notranslate"><span class="pre">`export</span> <span class="pre">PTBACKEND=Qt5Agg`</span></code></p>
<p>The default directory for image file output for some Analysator plotting tools is <code class="docutils literal notranslate"><span class="pre">$HOME/Plots</span></code>.
This setting can be altered with the system variable <code class="docutils literal notranslate"><span class="pre">export</span> <span class="pre">PTOUTPUTDIR=/target/directory/</span></code>.</p>
</section>
<section id="analysator-function-options">
<h3>Analysator function options<a class="headerlink" href="#analysator-function-options" title="Link to this heading"></a></h3>
<p>The formalism of providing Analysator plotting functions with arguments is similar to matlab or IDL, utilizing keywords. Many keywords have a default value of e.g. None, which the code checks against.</p>
</section>
<section id="interactive-help">
<h3>Interactive help<a class="headerlink" href="#interactive-help" title="Link to this heading"></a></h3>
<p>Most Analysator functions and classes contain up-to-date help, which is accessable in the python interpreter:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pt</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">plot_colormap</span><span class="err">?</span>
</pre></div>
</div>
</section>
<section id="interactive-plots">
<h3>Interactive plots<a class="headerlink" href="#interactive-plots" title="Link to this heading"></a></h3>
<p>On some systems you can activate interactive backends in Jupyter notebooks by issuing the command <code class="docutils literal notranslate"><span class="pre">%matplotlib</span> <span class="pre">ipympl</span></code> or <code class="docutils literal notranslate"><span class="pre">%matplotlib</span> <span class="pre">notebook</span></code> before importing pytools. This is not supported on the LUMI web interface.</p>
</section>
<section id="reading-data">
<h3>Reading data<a class="headerlink" href="#reading-data" title="Link to this heading"></a></h3>
<p>Access to Vlasiator output .vlsv files is handled through the Vlsvreader class. There are a number of
useful plotting routines which do not require editing the data directly, but for any in-depth scripting,
direct access routines are likely necessary.</p>
<section id="vlsvreader">
<h4>VlsvReader<a class="headerlink" href="#vlsvreader" title="Link to this heading"></a></h4>
<p>Open a file for access by creating a VlsvReader object.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">f</span><span class="o">=</span><span class="n">pt</span><span class="o">.</span><span class="n">vlsvfile</span><span class="o">.</span><span class="n">VlsvReader</span><span class="p">(</span><span class="s2">&quot;/path/to/simulation/bulk.0001234.vlsv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="listing-available-variables">
<h4>Listing available variables<a class="headerlink" href="#listing-available-variables" title="Link to this heading"></a></h4>
<p>Within python, you can list available variables as a concise list, or as a list of all available data reducers and operators:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">f</span><span class="o">.</span><span class="n">list</span><span class="p">()</span>
<span class="n">f</span><span class="o">.</span><span class="n">list</span><span class="p">(</span><span class="n">datareducer</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">operator</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="reading-in-vlasov-grid-mpigrid-variables">
<h4>Reading in vlasov grid (MPIgrid) variables<a class="headerlink" href="#reading-in-vlasov-grid-mpigrid-variables" title="Link to this heading"></a></h4>
<p>In older Vlasiator versions (before 5.0, simulation identifier second letter A through F) most
variables are saved on the MPIgrid and there is no identifying naming convention. Since version
5.0, with simulation version identifier letters starting from G, vlasov grid variables are
prepended with <code class="docutils literal notranslate"><span class="pre">vg_</span></code>. Note that for per-population variables, this is placed after the population name.</p>
<p>Variables are read and returned as numpy arrays. MPIgrid (Vlasov grid) cell scalar variables are returned
as a simple 1-dimensional array. Vectors, tensors and so on have additional dimensions tacked on. Note that
the ordering of CellIDs (and thus, the corresponding order of proton number densities and all other MPIgrid
variables) will vary between files. The list of MPIgrid CellIDs and the corresponding proton number
densities can be found with</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cellids</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">read_variable</span><span class="p">(</span><span class="s1">&#39;cellid&#39;</span><span class="p">)</span>
<span class="n">rho</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">read_variable</span><span class="p">(</span><span class="s1">&#39;proton/vg_rho&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>In order to use the read data, it needs to be sorted and rearranged to correspond with the
spatial grid structure. If the grid is 2-D and AMR was not used, this is relatively straightforward.
Select the coordinate sizes to match the simulation domain.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">xsize</span><span class="p">,</span> <span class="n">ysize</span><span class="p">,</span> <span class="n">zsize</span><span class="p">]</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">get_spatial_mesh_size</span><span class="p">()</span>
<span class="n">rho_shaped</span> <span class="o">=</span> <span class="n">rho</span><span class="p">[</span><span class="n">cellids</span><span class="o">.</span><span class="n">argsort</span><span class="p">()]</span><span class="o">.</span><span class="n">reshape</span><span class="p">([</span><span class="n">ysize</span><span class="p">,</span><span class="n">xsize</span><span class="p">])</span>
</pre></div>
</div>
<p>For vector data, use</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">bvol</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">read_variable</span><span class="p">(</span><span class="s1">&#39;vg_b_vol&#39;</span><span class="p">)</span>
<span class="n">bvol_shaped</span> <span class="o">=</span> <span class="n">bvol</span><span class="p">[</span><span class="n">cellids</span><span class="o">.</span><span class="n">argsort</span><span class="p">()]</span><span class="o">.</span><span class="n">reshape</span><span class="p">([</span><span class="n">ysize</span><span class="p">,</span><span class="n">xsize</span><span class="p">,</span><span class="mi">3</span><span class="p">])</span>
</pre></div>
</div>
</section>
<section id="reading-in-vlasov-grid-mpigrid-amr-variables">
<h4>Reading in vlasov grid (MPIgrid) AMR variables<a class="headerlink" href="#reading-in-vlasov-grid-mpigrid-amr-variables" title="Link to this heading"></a></h4>
<p>Since the AMR mesh is not refined in blocks but rather as an octree-mesh, the cells
from which the refined mesh consists of does not directly translate to a 2D array.
Re-sampling the input data is a somewhat involved process, and the interested reader can
peruse the contents of e.g. the <code class="docutils literal notranslate"><span class="pre">pyPlots/plot_colormap3dslice.py</span></code> file for a working example.</p>
</section>
<section id="reading-in-field-solver-grid-fsgrid-variables">
<h4>Reading in field solver grid (FSgrid) variables<a class="headerlink" href="#reading-in-field-solver-grid-fsgrid-variables" title="Link to this heading"></a></h4>
<p>Since Vlasiator version 5.0, field solver grid (FSgrid) variables can be output and are
prepended with <code class="docutils literal notranslate"><span class="pre">fg_</span></code>. FSgrid variables are returned as a numpy array, pre-sorted by the
reading routine, with dimensions matching the spatial dimensions and, if applicable, vector size.
For example, reading volumetric B-fields might yield an array of shape <code class="docutils literal notranslate"><span class="pre">(1024,</span> <span class="pre">736,</span> <span class="pre">736,</span> <span class="pre">3)</span></code>.
There is a separate routine for reading FSgrid variables, but the standard <code class="docutils literal notranslate"><span class="pre">read_variable()</span></code>
routine will redirect to the FSgrid routine if an FSgrid variable is requested.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">fg_b</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">read_fsgrid_variable</span><span class="p">(</span><span class="s1">&#39;fg_b&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Please note that FSgrid variables do not support reading via CellID. Transforming CellIDs to coordinates
and to FSgrid file indices is possible via functions provided by <code class="docutils literal notranslate"><span class="pre">pt.vlsvfile.VlsvReader</span></code> but are outside
the scope of this introductory tutorial.</p>
</section>
<section id="reading-variables-with-metadata">
<h4>Reading variables with metadata<a class="headerlink" href="#reading-variables-with-metadata" title="Link to this heading"></a></h4>
<p>Since Vlasiator 5.0, metadata is included for stored variables. The function <code class="docutils literal notranslate"><span class="pre">read_variable_info</span></code> returns
an object with the following fields: <code class="docutils literal notranslate"><span class="pre">data</span></code> (as per the <code class="docutils literal notranslate"><span class="pre">read_variable</span></code> or <code class="docutils literal notranslate"><span class="pre">read_fsgrid_variable</span></code>
call), <code class="docutils literal notranslate"><span class="pre">name</span></code>, <code class="docutils literal notranslate"><span class="pre">units</span></code>, <code class="docutils literal notranslate"><span class="pre">latex</span></code> (LaTeX-formatted name), <code class="docutils literal notranslate"><span class="pre">latexunits</span></code> (LaTeX-formatted unit)</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">vg_b_vol_with_info</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">read_variable_info</span><span class="p">(</span><span class="s1">&#39;vg_b_vol&#39;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="reading-spatial-cut-throughs">
<h4>Reading spatial cut-throughs<a class="headerlink" href="#reading-spatial-cut-throughs" title="Link to this heading"></a></h4>
<p>Reading a spatial profile through the simulation can be achieved with the <code class="docutils literal notranslate"><span class="pre">cut_through()</span></code> method.
This supports only Vlasov grid data, not FSgrid data. AMR support is not yet included. Select the
starting and final positions and read the line profile with</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cut</span><span class="o">=</span><span class="n">pt</span><span class="o">.</span><span class="n">calculations</span><span class="o">.</span><span class="n">cut_through</span><span class="p">(</span><span class="n">f</span><span class="p">,</span><span class="n">pos1</span><span class="p">,</span><span class="n">pos2</span><span class="p">);</span>
</pre></div>
</div>
<p>here <code class="docutils literal notranslate"><span class="pre">f</span></code> is the .vlsv file used for reading, <code class="docutils literal notranslate"><span class="pre">pos1</span></code> and <code class="docutils literal notranslate"><span class="pre">pos2</span></code> are XYZ coordinates (in metres) and the
returned structure contains the relevant cellIDs (<code class="docutils literal notranslate"><span class="pre">cut[0]</span></code>) and position along the cut (<code class="docutils literal notranslate"><span class="pre">cut[1]</span></code>, in metres).
You can read the actual cut data with</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">variable</span><span class="o">=</span><span class="n">get_data</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">read_variable</span><span class="p">(</span><span class="s2">&quot;vg_variablename&quot;</span><span class="p">,</span><span class="n">cut</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">data</span><span class="p">))</span>
</pre></div>
</div>
<p>Plot the data with</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">cut</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">data</span><span class="o">/</span><span class="n">Re</span><span class="p">,</span> <span class="n">variable</span><span class="p">)</span>
</pre></div>
</div>
<p>Instead of reading all cells along a cut, there exists an alternative function which proceeds primarily along
the cut in the dominant cartesian direction and returns one cellID per row/column.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cut</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">calculations</span><span class="o">.</span><span class="n">cut_through_step</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">pos1</span><span class="p">,</span> <span class="n">pos2</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="writing-data-with-vlsvwriter">
<h3>Writing Data with VlsvWriter<a class="headerlink" href="#writing-data-with-vlsvwriter" title="Link to this heading"></a></h3>
<p>From time to time, one may wish to perform more involved operations on the grid, and re-use them later. <code class="docutils literal notranslate"><span class="pre">VlsvWriter</span></code> can be used to save derived data on SpatialGrid. It operates by copying the grid metadata and data layout from an existing file at initialization, and can thereafter be used to store the results of more involved processing.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">f</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">vlsvfile</span><span class="o">.</span><span class="n">VlsvReader</span><span class="p">(</span><span class="n">input_file</span><span class="p">)</span>
<span class="c1"># Initialize, copy only the SpatialGrid mesh</span>
<span class="n">writer</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">vlsvfile</span><span class="o">.</span><span class="n">VlsvWriter</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">output_file</span><span class="p">,</span> <span class="n">copy_meshes</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;SpatialGrid&#39;</span><span class="p">])</span>
<span class="c1"># Copy some list of variables as a baseline. varlist accepts datareducer variables as well.</span>
<span class="n">writer</span><span class="o">.</span><span class="n">copy_variables</span><span class="p">(</span><span class="n">f</span><span class="p">,</span><span class="n">varlist</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CellID&quot;</span><span class="p">,</span><span class="s2">&quot;proton/vg_rho&quot;</span><span class="p">,</span><span class="s2">&quot;proton/vg_v&quot;</span><span class="p">,</span><span class="s2">&quot;vg_b_vol&quot;</span><span class="p">,</span><span class="s2">&quot;vg_e_vol&quot;</span><span class="p">,</span><span class="s2">&quot;vg_beta&quot;</span><span class="p">,</span><span class="s2">&quot;vg_beta_star&quot;</span><span class="p">])</span>

<span class="c1"># Do some heavy lifting that you don&#39;t want to repeat each time:</span>
<span class="n">orthogonality</span> <span class="o">=</span> <span class="n">lengthy_calculation_for_orthogonality</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<span class="c1"># Take care that this variable is compatible with the SpatialGrid variables,</span>
<span class="c1"># and that is has the same memory layout as CellIDs!</span>

<span class="c1"># Wrap the result with metadata</span>
<span class="n">varinfo</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">calculations</span><span class="o">.</span><span class="n">VariableInfo</span><span class="p">(</span><span class="n">orthogonality</span><span class="p">,</span>
                                       <span class="n">name</span><span class="o">=</span><span class="s2">&quot;vg_LMN_orthogonality&quot;</span><span class="p">,</span>
                                       <span class="n">units</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
                                       <span class="n">latex</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;$|\hat</span><span class="si">{L}</span><span class="s2">_\mathrm</span><span class="si">{MGA}</span><span class="s2">\times\hat</span><span class="si">{L}</span><span class="s2">_\mathrm</span><span class="si">{MDD}</span><span class="s2">|$&quot;</span><span class="p">,</span><span class="n">latexunits</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

<span class="c1"># Write the result to SpatialGrid with the output_file writer</span>
<span class="n">writer</span><span class="o">.</span><span class="n">write_variable_info</span><span class="p">(</span><span class="n">varinfo</span><span class="p">,</span> <span class="s1">&#39;SpatialGrid&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,)</span>
</pre></div>
</div>
</section>
<section id="interesting-questions-you-might-get">
<h3>Interesting questions you might get<a class="headerlink" href="#interesting-questions-you-might-get" title="Link to this heading"></a></h3>
<p>Q: Why are the output formats so convoluted?</p>
<p>A: They are optimized for run-time performance, so that each MPI task can simply pour its data into
one contiguous region on-disk via MPI writes.</p>
<p>A2: Evolution over time leads to interesting design choices.</p>
</section>
<section id="typical-pitfalls">
<h3>Typical pitfalls<a class="headerlink" href="#typical-pitfalls" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>Read Vlasov grid data and forget the order the cells based on CELLIDS</p></li>
<li><p>Read FSGrid data and accidentally order that also according to CELLIDS</p></li>
</ul>
</section>
</section>
<span id="document-analysator_exercises"></span><section id="analysator-exercises">
<h2>Analysator exercises<a class="headerlink" href="#analysator-exercises" title="Link to this heading"></a></h2>
<section id="why-we-teach-this-lesson">
<h3>Why we teach this lesson<a class="headerlink" href="#why-we-teach-this-lesson" title="Link to this heading"></a></h3>
<p>This page has a number of suggested exercises to help the
participants to learn of some of the possibilities of the Analysator
visualization toolkit. One should first familiarize oneself with the
Analysator introduction page and keep those lessons in mind. Remember
that most plotting routines accept <strong>keywords</strong> with certain pre-determined default values.</p>
</section>
<section id="outputs">
<h3>Outputs<a class="headerlink" href="#outputs" title="Link to this heading"></a></h3>
<p>Running Analysator within a Jupyter interface will by default show the results on-screen, but will also store the image as a .png file in the default Analysator output directory. You can deactivate .png output by providing a plotting routine with the <code class="docutils literal notranslate"><span class="pre">draw=True</span></code> keyword.</p>
</section>
<section id="vlsvreader">
<h3>VlsvReader<a class="headerlink" href="#vlsvreader" title="Link to this heading"></a></h3>
<p>It may be helpful to open a file for access by creating a VlsvReader object. Here are three sample files: one from a 5D run, one from a 6D run, and one from a coarse 6D test with an ionospheric inner boundary active.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">f2d</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">vlsvfile</span><span class="o">.</span><span class="n">VlsvReader</span><span class="p">(</span><span class="s2">&quot;/scratch/project_465000693/example_data/AGD/bulk.0000904.vlsv&quot;</span><span class="p">)</span>
<span class="n">f3d</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">vlsvfile</span><span class="o">.</span><span class="n">VlsvReader</span><span class="p">(</span><span class="s2">&quot;/scratch/project_465000693/example_data/EGE/bulk.0001500.vlsv&quot;</span><span class="p">)</span>
<span class="n">fiono</span> <span class="o">=</span> <span class="n">pt</span><span class="o">.</span><span class="n">vlsvfile</span><span class="o">.</span><span class="n">VlsvReader</span><span class="p">(</span><span class="s2">&quot;/scratch/project_465000693/example_data/IGtest/bulk.0000229.vlsv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="listing-available-variables">
<h3>Listing available variables<a class="headerlink" href="#listing-available-variables" title="Link to this heading"></a></h3>
<p>Within python, you can list available variables for a given file as a list, or as an extended
list of all available data reducers and operators. The return value of this function will be grouped as PARAMETERs (one per file), VARIABLEs (one per simulation cell)</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">f</span><span class="o">.</span><span class="n">list</span><span class="p">()</span>
<span class="n">f</span><span class="o">.</span><span class="n">list</span><span class="p">(</span><span class="n">operator</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">datareducer</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>It is important to note that the list of datareducers is <em>not</em> limited to those
supported by the available data, but rather lists all supported by the plotting library.</p>
</section>
<section id="plotting-2d-5d-simulation-data">
<h3>Plotting 2D/5D simulation data<a class="headerlink" href="#plotting-2d-5d-simulation-data" title="Link to this heading"></a></h3>
<p>The easiest manner to plot 2D data from a 5D simulation run is to use the pre-existing
plotting routines. These will sort, scale, and crop the data as requested and provide
suitable axes and color bars. You can access a given file via several approaches:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pt</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">plot_colormap</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;/scratch/project_465000693/example_data/AGD/bulk.0000904.vlsv&quot;</span><span class="p">)</span>
<span class="n">pt</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">plot_colormap</span><span class="p">(</span><span class="n">vlsvobj</span><span class="o">=</span><span class="n">f2d</span><span class="p">)</span>
<span class="n">pt</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">plot_colormap</span><span class="p">(</span><span class="n">filedir</span><span class="o">=</span><span class="s2">&quot;/scratch/project_465000693/example_data/AGD/&quot;</span><span class="p">,</span> <span class="n">step</span><span class="o">=</span><span class="mi">904</span><span class="p">)</span>
</pre></div>
</div>
<p>The first thing to trial is plotting different variables from the data. First, use <code class="docutils literal notranslate"><span class="pre">f2d.list()</span></code> to review which outputs were active during the simulation. PARAMETERs cannot be plotted as they are singular values, VARIABLEs can. For any given VLSV file you should see at least overall vg-variables and per-population vg-variables (e.g. proton values). Often, you can find also fieldsolver grid fg-variables and, only in ionospheric 6D runs, ionospheric grid ig-variables.</p>
<section id="altering-basic-plot-properties">
<h4>Altering basic plot properties<a class="headerlink" href="#altering-basic-plot-properties" title="Link to this heading"></a></h4>
<p>Selecting a new output variable (replacing the default proton number density) is performed as, for example:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pt</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">plot_colormap</span><span class="p">(</span><span class="n">vlsvobj</span><span class="o">=</span><span class="n">f2d</span><span class="p">,</span> <span class="n">var</span><span class="o">=</span><span class="s1">&#39;vg_blocks&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Select a handful of output variables from <code class="docutils literal notranslate"><span class="pre">f2d.list()</span></code> and try plotting them. Once you’re familiar with how vg-variables work, you can also try proton-specific values and then fg-variables.</p>
<p>For scalar variables such as number densities, the default setting is to use a logarithmic colour scale. When selecting a vector variable such as <code class="docutils literal notranslate"><span class="pre">proton/vg_v</span></code>, the default option is to reduce the value to the scalar magnitude of the vector. To view a singular component of a vector, you can select the component via the <code class="docutils literal notranslate"><span class="pre">operator</span></code> or <code class="docutils literal notranslate"><span class="pre">op</span></code> keyword:,  <code class="docutils literal notranslate"><span class="pre">op='x'</span></code>, <code class="docutils literal notranslate"><span class="pre">op='y'</span></code>, or  <code class="docutils literal notranslate"><span class="pre">op='z'</span></code>. This will automatically swap the colour scale to be linear, facilitating both positive and negative values. If, on the other hand, you want to plot the magnitude of a component, you can fold the output into an absolute value with the keyword <code class="docutils literal notranslate"><span class="pre">absolute=True</span></code>. As the default color map for a linear plot is a diverging map with white in the centre, it is possible to enforce white mapping to zero by forcing the range to be symmetric with <code class="docutils literal notranslate"><span class="pre">`symmetric=True</span></code>.</p>
<p>Next, try zooming around. You can adjust the plot domain by cropping it, selecting a plotting box either in metres or in units of Earth radius. You can select the crop box as either <code class="docutils literal notranslate"><span class="pre">boxm=[xmin,xmax,ymin,ymax]</span></code> in metres, or <code class="docutils literal notranslate"><span class="pre">boxre=[xmin,xmax,ymin,ymax]</span></code> in units of Earth radius. The default unit used for plot axes is Earth radii, but you can change them to metres with <code class="docutils literal notranslate"><span class="pre">axisunit=0</span></code>, kilometres with <code class="docutils literal notranslate"><span class="pre">axisunit=3</span></code>, or e.g. megametres with <code class="docutils literal notranslate"><span class="pre">axisunit=6</span></code>.</p>
</section>
<section id="altering-the-colour-scale">
<h4>Altering the colour scale<a class="headerlink" href="#altering-the-colour-scale" title="Link to this heading"></a></h4>
<p>Plotting is all about pretty pictures, and pretty pictures are all about colormaps. <a href="#id1"><span class="problematic" id="id2">`https://matplotlib.org/stable/users/explain/colors/colormaps.html`_</span></a> contains a list of available matplotlib colormaps, supported by analysator. In addition to the standard colormaps, we have created some extra ones such as <code class="docutils literal notranslate"><span class="pre">hot_desaturated</span></code> and <code class="docutils literal notranslate"><span class="pre">warhol</span></code>. Try plotting some of your variables with various colormaps e.g. with the keywords <code class="docutils literal notranslate"><span class="pre">colormap='viridis'</span></code> or <code class="docutils literal notranslate"><span class="pre">colormap='plasma'</span></code>.</p>
<p>Next, try plotting a variable which would make sense to plot on a linear scale, such as <code class="docutils literal notranslate"><span class="pre">vg_rank</span></code>. To activate a linear scale, give the keyword <code class="docutils literal notranslate"><span class="pre">lin=True</span></code>. To adjust the number of tick marks on the colourbar axis, you can provide them as e.g. <code class="docutils literal notranslate"><span class="pre">lin=5</span></code>.</p>
<p>By default, the minimum and maximum values of the colourbar axis are selected from the visible data. If you which to adjust the limits, to over- or undersaturate regions, you can set the minimum and maximum values of the scale with e.g. <code class="docutils literal notranslate"><span class="pre">vmin=1</span></code> and <code class="docutils literal notranslate"><span class="pre">vmax=100</span></code>. Now, as a next exercise, plot a magnetic field variable, such as <code class="docutils literal notranslate"><span class="pre">fg_b</span></code>, and turn off the masking of the inner boundary region with <code class="docutils literal notranslate"><span class="pre">nomask=True</span></code>. You will notice that the centre of the dipole field will dominate the plot. Next, adjust the minimum and maximum values along with the selected colour map to view the regions of interest of your plot. You may note that different regions of the magnetosphere (sheath, tail, foreshock) are best plotted with different colourbar ranges.</p>
<p>When plotting magnetospheric magnetic field values, you might note that having the field strength in Tesla is not always smart. By providing the keyword <code class="docutils literal notranslate"><span class="pre">vscale=None</span></code>, you allow the routine to auto-scale to a suggested value, e.g. nanotesla. Setting the value manually as <code class="docutils literal notranslate"><span class="pre">vscale=1e9</span></code> provides the same result. Other scaling factors can also be used, and they may or may not offer suitable unit names.</p>
<p>The tick marks of the colourbar default to using scientific notation, which may be deactivated with the keyword <code class="docutils literal notranslate"><span class="pre">usesci=False</span></code>.</p>
<p>Should you wish to evaluate a wide range of variability whist still allowing both positive and negative values, Analysator allows the use of the symmetric logarithmic colour scale. Try it by plotting the out-of-plane magnetic field component <code class="docutils literal notranslate"><span class="pre">var=fg_b,</span> <span class="pre">op='z',</span> <span class="pre">symlog=0</span></code> where giving symlog the value of zero allows it to self-determine the extent of the linear range at the centre between two mirrored logarithmic ranges.</p>
</section>
<section id="overlaying-data-on-top-of-the-plot">
<h4>Overlaying data on top of the plot<a class="headerlink" href="#overlaying-data-on-top-of-the-plot" title="Link to this heading"></a></h4>
<p>In addition to plotting a single value as a colormap, it is possible to overlay other information. A Vlasiator watermark is available by setting the <code class="docutils literal notranslate"><span class="pre">wmark</span></code> (colour) or <code class="docutils literal notranslate"><span class="pre">wmarkb</span></code> (black) keyword, which takes a location string such as <code class="docutils literal notranslate"><span class="pre">&quot;NW&quot;,&quot;NE&quot;,&quot;SW&quot;</span></code>, or <code class="docutils literal notranslate"><span class="pre">&quot;SW&quot;</span></code>. Setting <code class="docutils literal notranslate"><span class="pre">Earth=true</span></code> plots an Earth symbol inside the inner boundary with a realistic radius.</p>
<p>Setting the keword <code class="docutils literal notranslate"><span class="pre">fsaved=True</span></code> will add contours delienating all spatial cells which contain saved-to-disk velocity distribution functions.</p>
<p>Vector quantities can be overplotted as streamlines or vector maps. Give some a go!
.. code-block:: python</p>
<blockquote>
<div><dl class="field-list simple">
<dt class="field-odd">kword vectors<span class="colon">:</span></dt>
<dd class="field-odd"><p>Set to a vector variable to overplot (unit length vectors, color displays variable magnitude)</p>
</dd>
<dt class="field-even">kword vectordensity<span class="colon">:</span></dt>
<dd class="field-even"><p>Aim for how many vectors to show in plot window (default 100)</p>
</dd>
<dt class="field-odd">kword vectorcolormap<span class="colon">:</span></dt>
<dd class="field-odd"><p>Colormap to use for overplotted vectors (default: gray)</p>
</dd>
<dt class="field-even">kword vectorsize<span class="colon">:</span></dt>
<dd class="field-even"><p>Scaling of vector sizes</p>
</dd>
<dt class="field-odd">kword streamlines<span class="colon">:</span></dt>
<dd class="field-odd"><p>Set to a vector variable to overplot as streamlines</p>
</dd>
<dt class="field-even">kword streamlinedensity<span class="colon">:</span></dt>
<dd class="field-even"><p>Set streamline density (default 1)</p>
</dd>
<dt class="field-odd">kword streamlinecolor<span class="colon">:</span></dt>
<dd class="field-odd"><p>Set streamline color (default white)</p>
</dd>
<dt class="field-even">kword streamlinethick<span class="colon">:</span></dt>
<dd class="field-even"><p>Set streamline thickness</p>
</dd>
</dl>
</div></blockquote>
<p>If a pre-generated fluxfunction file is available (only for 5D simulations), magnetic field lines can be plotted
with greater accuracy than via streamlines. Flux functions are generated with the fluxfunction tool distributed as part of Vlasiator. A plotting example:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pt</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">plot_colormap</span><span class="p">(</span><span class="n">filedir</span><span class="o">=</span><span class="s1">&#39;/scratch/project_465000693/example_data/AGD/&#39;</span><span class="p">,</span> <span class="n">var</span><span class="o">=</span><span class="s1">&#39;proton/vg_v&#39;</span><span class="p">,</span> <span class="n">boxre</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">40</span><span class="p">,</span><span class="o">-</span><span class="mi">20</span><span class="p">,</span><span class="mi">20</span><span class="p">],</span> <span class="n">fluxdir</span><span class="o">=</span><span class="s1">&#39;/scratch/project_465000693/example_data/AGD/flux/&#39;</span><span class="p">,</span><span class="n">step</span><span class="o">=</span><span class="mi">904</span><span class="p">,</span><span class="n">lin</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">vscale</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span><span class="n">vmin</span><span class="o">=</span><span class="mi">400</span><span class="p">,</span> <span class="n">flux_levels</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="fine-tuning-of-plot-properties">
<h4>Fine-tuning of plot properties<a class="headerlink" href="#fine-tuning-of-plot-properties" title="Link to this heading"></a></h4>
<p>Several keywords exist for fine-tuning Analysator plot properties. As usual, the description of these can be found by calling the help functionality with <code class="docutils literal notranslate"><span class="pre">pt.plot.plot_colormap?</span></code>. Some examples are provided below</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="p">:</span><span class="n">kword</span> <span class="n">noborder</span><span class="p">:</span>    <span class="n">Plot</span> <span class="n">figure</span> <span class="n">edge</span><span class="o">-</span><span class="n">to</span><span class="o">-</span><span class="n">edge</span> <span class="n">without</span> <span class="n">borders</span> <span class="p">(</span><span class="n">default</span> <span class="n">off</span><span class="p">)</span>
<span class="p">:</span><span class="n">kword</span> <span class="n">noxlabels</span><span class="p">:</span>   <span class="n">Suppress</span> <span class="n">x</span><span class="o">-</span><span class="n">axis</span> <span class="n">labels</span> <span class="ow">and</span> <span class="n">title</span>
<span class="p">:</span><span class="n">kword</span> <span class="n">noylabels</span><span class="p">:</span>   <span class="n">Suppress</span> <span class="n">y</span><span class="o">-</span><span class="n">axis</span> <span class="n">labels</span> <span class="ow">and</span> <span class="n">title</span>
<span class="p">:</span><span class="n">kword</span> <span class="n">scale</span><span class="p">:</span>       <span class="n">Scale</span> <span class="n">text</span> <span class="n">size</span> <span class="p">(</span><span class="n">default</span><span class="o">=</span><span class="mf">1.0</span><span class="p">)</span>
<span class="p">:</span><span class="n">kword</span> <span class="n">thick</span><span class="p">:</span>       <span class="n">line</span> <span class="ow">and</span> <span class="n">axis</span> <span class="n">thickness</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mf">1.0</span>
<span class="p">:</span><span class="n">kword</span> <span class="n">nocb</span><span class="p">:</span>        <span class="n">Set</span> <span class="n">to</span> <span class="n">suppress</span> <span class="n">drawing</span> <span class="n">of</span> <span class="n">colourbar</span>
<span class="p">:</span><span class="n">kword</span> <span class="n">internalcb</span><span class="p">:</span>  <span class="n">Set</span> <span class="n">to</span> <span class="n">draw</span> <span class="n">colorbar</span> <span class="n">inside</span> <span class="n">plot</span> <span class="n">instead</span> <span class="n">of</span> <span class="n">outside</span><span class="o">.</span> <span class="n">If</span> <span class="nb">set</span> <span class="n">to</span> <span class="n">a</span> <span class="n">text</span>
                    <span class="n">string</span><span class="p">,</span> <span class="n">tries</span> <span class="n">to</span> <span class="n">use</span> <span class="n">that</span> <span class="k">as</span> <span class="n">the</span> <span class="n">location</span><span class="p">,</span> <span class="n">e</span><span class="o">.</span><span class="n">g</span><span class="o">.</span> <span class="s2">&quot;NW&quot;</span><span class="p">,</span><span class="s2">&quot;NE&quot;</span><span class="p">,</span><span class="s2">&quot;SW&quot;</span><span class="p">,</span><span class="s2">&quot;SW&quot;</span>
<span class="p">:</span><span class="n">kword</span> <span class="n">highres</span><span class="p">:</span>     <span class="n">Creates</span> <span class="n">the</span> <span class="n">image</span> <span class="ow">in</span> <span class="n">high</span> <span class="n">resolution</span><span class="p">,</span> <span class="n">scaled</span> <span class="n">up</span> <span class="n">by</span> <span class="n">this</span> <span class="n">value</span> <span class="p">(</span><span class="n">suitable</span> <span class="k">for</span> <span class="nb">print</span><span class="p">)</span><span class="o">.</span>
<span class="p">:</span><span class="n">kword</span> <span class="n">tickinterval</span><span class="p">:</span> <span class="n">Interval</span> <span class="n">at</span> <span class="n">which</span> <span class="n">to</span> <span class="n">have</span> <span class="n">ticks</span> <span class="n">on</span> <span class="n">axes</span> <span class="p">(</span><span class="ow">not</span> <span class="n">colorbar</span><span class="p">)</span>
<span class="p">:</span><span class="n">kword</span> <span class="n">title</span><span class="p">:</span>       <span class="n">string</span> <span class="n">to</span> <span class="n">use</span> <span class="k">as</span> <span class="n">plot</span> <span class="n">title</span> <span class="n">instead</span> <span class="n">of</span> <span class="n">time</span><span class="o">.</span>
                    <span class="n">Special</span> <span class="n">case</span><span class="p">:</span> <span class="n">Set</span> <span class="n">to</span> <span class="s2">&quot;msec&quot;</span> <span class="n">to</span> <span class="n">plot</span> <span class="n">time</span> <span class="k">with</span> <span class="n">millisecond</span> <span class="n">accuracy</span> <span class="ow">or</span> <span class="s2">&quot;musec&quot;</span>
                    <span class="k">for</span> <span class="n">microsecond</span> <span class="n">accuracy</span><span class="o">.</span> <span class="s2">&quot;sec&quot;</span> <span class="ow">is</span> <span class="n">integer</span> <span class="n">second</span> <span class="n">accuracy</span><span class="o">.</span>
<span class="p">:</span><span class="n">kword</span> <span class="n">cbtitle</span><span class="p">:</span>     <span class="n">string</span> <span class="n">to</span> <span class="n">use</span> <span class="k">as</span> <span class="n">colorbar</span> <span class="n">title</span> <span class="n">instead</span> <span class="n">of</span> <span class="nb">map</span> <span class="n">name</span>
<span class="p">:</span><span class="n">kword</span> <span class="n">nooverwrite</span><span class="p">:</span> <span class="n">Set</span> <span class="n">to</span> <span class="n">only</span> <span class="n">perform</span> <span class="n">actions</span> <span class="k">if</span> <span class="n">the</span> <span class="n">target</span> <span class="n">output</span> <span class="n">file</span> <span class="n">does</span> <span class="ow">not</span> <span class="n">yet</span> <span class="n">exist</span>
</pre></div>
</div>
</section>
</section>
<section id="plotting-3d-6d-simulation-data">
<h3>Plotting 3D/6D simulation data<a class="headerlink" href="#plotting-3d-6d-simulation-data" title="Link to this heading"></a></h3>
<p>For 3D/6D simulation data, interactive VisIt plotting is often the easiest and most powerful way to go. Analysator still provides several useful tools for plotting 3D data reduced into 3D images. These methods can be expensive, as the AMR MPIgrid data is resampled to the highest resolution. This allows handling of both vg-grid and fg-grid variables. The simplest method is to plot a Cartesian 2D slice out of the 3D domain:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pt</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">plot_colormap3dslice</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;/scratch/project_465000693/example_data/EGE/bulk.0001500.vlsv&quot;</span><span class="p">)</span>
<span class="n">pt</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">plot_colormap3dslice</span><span class="p">(</span><span class="n">vlsvobj</span><span class="o">=</span><span class="n">f3d</span><span class="p">)</span>
<span class="n">pt</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">plot_colormap3dslice</span><span class="p">(</span><span class="n">filedir</span><span class="o">=</span><span class="s2">&quot;/scratch/project_465000693/example_data/EGE/&quot;</span><span class="p">,</span> <span class="n">step</span><span class="o">=</span><span class="mi">1500</span><span class="p">)</span>
</pre></div>
</div>
<p>For these Cartesian slices, many of the options of regular 2D plots are accepted. In addition, you can select the direction normal to the plot slice with e.g. <code class="docutils literal notranslate"><span class="pre">normal='y'</span></code> (the default) or <code class="docutils literal notranslate"><span class="pre">normal='z'</span></code>. The coordinate along this normal direction used for the slice is set with the keyword <code class="docutils literal notranslate"><span class="pre">cutpoint</span></code> in metres or <code class="docutils literal notranslate"><span class="pre">cutpointre</span></code> in Earth radii.</p>
<p>Another, much heavier option is the <em>threeslice</em> which intersects three Cartesian planes through the simulation domain and plots them all. Generating one of these images can take several tens of seconds and require significant memory from the Jupyter server.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pt</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">plot_threeslice</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;/scratch/project_465000693/example_data/EGE/bulk.0001500.vlsv&quot;</span><span class="p">,</span><span class="n">draw</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
<p>Another again heavy option is to plot an isosurface of a 3D plot, showing one variable at a surface where another variable is constant. Generating one of these images may also take several tens of seconds and require significant memory from the Jupyter server. A plotting example:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pt</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">plot_isosurface</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s2">&quot;/scratch/project_465000693/example_data/EGE/bulk.0001500.vlsv&quot;</span><span class="p">,</span><span class="n">surf_var</span><span class="o">=</span><span class="s1">&#39;vg_b_vol&#39;</span><span class="p">,</span><span class="n">surf_op</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span><span class="n">surf_level</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="n">color_var</span><span class="o">=</span><span class="s1">&#39;proton/vg_v&#39;</span><span class="p">,</span><span class="n">color_op</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span><span class="n">draw</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="n">usesci</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vscale</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">boxre</span><span class="o">=</span><span class="p">[</span><span class="o">-</span><span class="mi">40</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="o">-</span><span class="mi">20</span><span class="p">,</span><span class="mi">20</span><span class="p">,</span><span class="o">-</span><span class="mi">20</span><span class="p">,</span><span class="mi">20</span><span class="p">],</span><span class="n">angle</span><span class="o">=</span><span class="p">[</span><span class="mi">20</span><span class="p">,</span><span class="mi">210</span><span class="p">])</span>
</pre></div>
</div>
<p>For isosurfaces, cropping the plotting region can help significantly with both image clarity and plot time. The <code class="docutils literal notranslate"><span class="pre">angle</span></code> keyword is used to define both the elevation angle and the azimuthal rotation around the z axis. The surface can be set to be opaque with <code class="docutils literal notranslate"><span class="pre">transparent=False</span></code>.</p>
</section>
<section id="plotting-ionospheric-simulation-data">
<h3>Plotting ionospheric simulation data<a class="headerlink" href="#plotting-ionospheric-simulation-data" title="Link to this heading"></a></h3>
<p>A separate routine exists for plotting ionospheric values flattened on a polar plot. For example:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pt</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">plot_ionosphere</span><span class="p">(</span><span class="n">vlsvobj</span><span class="o">=</span><span class="n">fiono</span><span class="p">,</span> <span class="n">var</span><span class="o">=</span><span class="s1">&#39;ig_fac&#39;</span><span class="p">,</span><span class="n">viewdir</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">symlog</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">draw</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
<section id="advanced-methods-axes-post-processing-overlaying">
<h4>Advanced methods: axes, post-processing, overlaying<a class="headerlink" href="#advanced-methods-axes-post-processing-overlaying" title="Link to this heading"></a></h4>
<p>Several advanced methods exist beyond the scope of this tutorial.</p>
<p>An Analysator plot can be placed inside another image using the <code class="docutils literal notranslate"><span class="pre">axes</span></code> keyword. Examples can be found in <code class="docutils literal notranslate"><span class="pre">examples/gridspec_plot.py</span></code> and <code class="docutils literal notranslate"><span class="pre">examples/multi_panel_plot.py</span></code>.</p>
<p>Variables read from the vlsv file can be passed to user-provided functions for post-processing. There are two types of functions supported:</p>
</section>
</section>
<section id="expression-functions">
<h3>Expression functions<a class="headerlink" href="#expression-functions" title="Link to this heading"></a></h3>
<p>An expression takes arrays of variables, computes a new value from them (a scalar or a vector), and then returns it to the main plotting routine. The colormap variable is replaced with the result of the expression. Examples of more involved expressions can be found in <code class="docutils literal notranslate"><span class="pre">examples/generate_panel.py</span></code>. A long list of simpler expressions can be found in <code class="docutils literal notranslate"><span class="pre">pyPlots/plot_helpers.py</span></code>.</p>
<p>Expressions can also be instructed to be given several timesteps of data in order to facilitate time derivatives of variables.</p>
</section>
<section id="external-functions">
<h3>External functions<a class="headerlink" href="#external-functions" title="Link to this heading"></a></h3>
<p>External functions act in many ways like expressions, but they are also given the axes of the plot along with coordinate information. An external function does not replace the main variable of the plotting routine, but can be used to overlay information, variables, or other information on top of the main plot.</p>
<section id="advanced-methods-2-energy-spectrograms-virtual-spacecraft">
<h4>Advanced methods 2: energy spectrograms, virtual spacecraft<a class="headerlink" href="#advanced-methods-2-energy-spectrograms-virtual-spacecraft" title="Link to this heading"></a></h4>
<p>More involved data analysis is also outside the scope of this tutorial, but the interested reader should look at topics within the Analysator wiki.</p>
<p>Plotting a virtual spacecraft time profile involves deciding on the position for the spacecraft and choosing a time extent for the measurement. There is no quick tool for this, as for each case the required variables are likely different, but a simple example script can be found as <code class="docutils literal notranslate"><span class="pre">examples/plot_VSC.py</span></code></p>
<p>Plotting time-energy spectrograms is a highly requested feature, and rudimentary scripts do exist within the Analysator repository, e.g. <code class="docutils literal notranslate"><span class="pre">scripts/plot_time_energy_spectrogram.py</span></code>. Future updates will most likely add a full-fledged plotting routine to the toolkit.</p>
</section>
</section>
<section id="interesting-questions-you-might-get">
<h3>Interesting questions you might get<a class="headerlink" href="#interesting-questions-you-might-get" title="Link to this heading"></a></h3>
<p>Q: How can I plot XXX or YYY?</p>
<p>A: look through the examples and the post-processing scripts in <code class="docutils literal notranslate"><span class="pre">pyPlots/plot_helpers.py</span></code>. If you are not successful, you can look through those for insipration, code it yourself, and make a pull request!</p>
<p>A2: You can of course always ask the developers directly - perhaps the functionality exists in a non-released feature branch.</p>
<p>Q: Why does my plotting routine crash?</p>
<p>A: Analysator has formed organically over many years, with new tools and features coming in along with new data types. Sometimes mistakes slip through. Also, it’s worth checking if you can find more up-to-date versions of scipy, matplotlib, or numpy - that may help.</p>
</section>
<section id="typical-pitfalls">
<h3>Typical pitfalls<a class="headerlink" href="#typical-pitfalls" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p>Forgetting to pull the latest version of Analysator</p></li>
<li><p>Directly editing the plotting scripts instead of using external or expression functionality, leading to conflicts when Analysator master gets updated</p></li>
</ul>
</section>
</section>
</div>
<div class="toctree-wrapper compound">
<span id="document-quick-reference"></span><section id="quick-reference">
<h2>Quick Reference<a class="headerlink" href="#quick-reference" title="Link to this heading"></a></h2>
<p>Reference cards sorted by topic.</p>
<section id="adaptive-mesh-refinement-cfg">
<h3>Adaptive Mesh Refinement cfg<a class="headerlink" href="#adaptive-mesh-refinement-cfg" title="Link to this heading"></a></h3>
<p>All parameters under <code class="docutils literal notranslate"><span class="pre">[AMR]</span></code></p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">max_spatial_level</span></code>: Maximum AMR level for both static and adaptive refinement</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">refine_radius</span></code>: Maximum distance from origin to refine cells, default unlimited. On static refinement this should be less than gridbuilder.x_min to avoid tail box refinement on the back wall.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">adapt_refinement</span></code>: Set true to use adaptive refinement</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">use_alpha1</span></code>: Use the maximum of dimensionless gradients as a refinement index, default true (is used if adapt_refinement is set)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">use_alpha2</span></code>: Use J/B_perp (measuring current sheets) as a refinement index, default true (is used if adapt_refinement is set)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">alpha1_refine_threshold</span></code>, <code class="docutils literal notranslate"><span class="pre">alpha_2_refine_threshold</span></code>: Minimum values of alpha to refine a cell, default 0.5</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">alpha1_coarsen_threshold</span></code>, <code class="docutils literal notranslate"><span class="pre">alpha_2_coarsen_threshold</span></code>: Maximum values of alpha to coarsen a cell, default half of refine threshold</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">refine_cadence</span></code>: Adapt refinement in the grid every N load balances, default 5 (so every 5th load balance)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">refine_after</span></code>: Minimum time to start refinement, allows you to initialize the grid on minimal refinement. Default 0, i.e. no minimum time</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">alpha1_dX_weight</span></code>: Multiplier for each constituent variable in the calculation of alpha1, default 1.</p></li>
</ul>
</section>
</section>
<span id="document-vlasiator-cfg-reference"></span><section id="vlasiator-config-reference">
<h2>Vlasiator config reference<a class="headerlink" href="#vlasiator-config-reference" title="Link to this heading"></a></h2>
<p>These can be viewed by running <code class="docutils literal notranslate"><span class="pre">./vlasiator</span> <span class="pre">--help</span></code>.</p>
<section id="command-line-arguments">
<h3>Command line arguments<a class="headerlink" href="#command-line-arguments" title="Link to this heading"></a></h3>
<div class="highlight-cfg notranslate"><div class="highlight"><pre><span></span><span class="na">Usage</span><span class="o">:</span><span class="w"> </span><span class="s">main [options (options given on the command line override options given everywhere else)], where options are::</span>
<span class="w"> </span><span class="na">--help                                                                       print this help message</span>
<span class="w"> </span><span class="na">--version                                                                    print version information</span>
<span class="w"> </span><span class="na">--global_config arg                                                          read options from the global configuration file arg (relative to the current</span>
<span class="w">                                                                              </span><span class="na">working directory). Options given in this file are overridden by options given</span>
<span class="w">                                                                              </span><span class="na">in the user&#39;s and run&#39;s configuration files and by options given in environment</span>
<span class="w">                                                                              </span><span class="na">variables (prefixed with MAIN_) and the command line</span>
<span class="w"> </span><span class="na">--user_config arg                                                            read options from the user&#39;s configuration file arg (relative to the current</span>
<span class="w">                                                                              </span><span class="na">working directory). Options given in this file override options given in the</span>
<span class="w">                                                                              </span><span class="na">global configuration file. Options given in this file are overridden by options</span>
<span class="w">                                                                              </span><span class="na">given in the run&#39;s configuration file and by options given in environment</span>
<span class="w">                                                                              </span><span class="na">variables (prefixed with MAIN_) and the command line</span>
<span class="w"> </span><span class="na">--run_config arg                                                             read options from the run&#39;s configuration file arg (relative to the current</span>
<span class="w">                                                                              </span><span class="na">working directory). Options given in this file override options given in the</span>
<span class="w">                                                                              </span><span class="na">user&#39;s and global configuration files. Options given in this override options</span>
<span class="w">                                                                              </span><span class="na">given in the user&#39;s and global configuration files. Options given in this file</span>
<span class="w">                                                                              </span><span class="na">are overridden by options given in environment variables (prefixed with MAIN_)</span>
<span class="w">                                                                              </span><span class="na">and the command line</span>
</pre></div>
</div>
</section>
<section id="configuration-file-options">
<h3>Configuration file options<a class="headerlink" href="#configuration-file-options" title="Link to this heading"></a></h3>
<div class="highlight-cfg notranslate"><div class="highlight"><pre><span></span><span class="na">propagate_field                                                            Propagate magnetic field during the simulation</span>
<span class="na">propagate_vlasov_acceleration                                              Propagate distribution functions during the simulation in velocity space. Iffalse, it is propagated with zero length timesteps.</span>
<span class="na">propagate_vlasov_translation                                               Propagate distribution functions during the simulation in ordinary space. Iffalse, it is propagated with zero length timesteps.</span>
<span class="na">dynamic_timestep                                                           If true,  timestep is set based on  CFL limits (default on)</span>
<span class="na">hallMinimumRho                                                             Minimum rho value used for the Hall and electron pressure gradient terms in theLorentz force and in the field solver. Default is very low and has no effect inpractice.</span>
<span class="na">project                                                                    Specify the name of the project to use. Supported to date (20150610)</span><span class="o">:</span><span class="w"> </span><span class="s">AlfvenDiffusion Dispersion Distributions Firehose Flowthrough Fluctuations Harris KHBLarmor Magnetosphere Multipeak Riemann1 Shock Shocktest Template test_fptestHall test_trans VelocityBox verificationLarmor</span>
<span class="na">ParticlePopulations                                                        Name of the simulated particle populations (string)</span>

<span class="k">[io]</span>
<span class="na">diagnostic_write_interval                                                  Write diagnostic output every arg time steps</span>
<span class="na">system_write_t_interval                                                    Save the simulation every arg simulated seconds. Negative values disable writes.[Define for all groups.]</span>
<span class="na">system_write_file_name                                                     Save the simulation to this file name series. [Define for all groups.]</span>
<span class="na">system_write_path                                                          Save this series in this location. Default is ./ [Define for all groups ornone.]</span>
<span class="na">system_write_distribution_stride                                           Every this many cells write out their velocity space. 0 is none. [Define for allgroups.]</span>
<span class="na">system_write_distribution_xline_stride                                     Every this many lines of cells along the x direction write out their velocityspace. 0 is none. [Define for all groups.]</span>
<span class="na">system_write_distribution_yline_stride                                     Every this many lines of cells along the y direction write out their velocityspace. 0 is none. [Define for all groups.]</span>
<span class="na">system_write_distribution_zline_stride                                     Every this many lines of cells along the z direction write out their velocityspace. 0 is none. [Define for all groups.]</span>
<span class="na">system_write_distribution_shell_radius                                     At cells intersecting spheres with those radii centred at the origin write outtheir velocity space. 0 is none.</span>
<span class="na">system_write_distribution_shell_stride                                     Every this many cells for those on selected shells write out their velocityspace. 0 is none.</span>
<span class="na">system_write_fsgrid_variables                                              If 0 don&#39;t write fsgrid DROs, if 1 do write them.</span>
<span class="na">system_write_mpiio_hint_key                                                MPI-IO hint key passed to the non-restart IO. Has to be matched byio.system_write_mpiio_hint_value.</span>
<span class="na">system_write_mpiio_hint_value                                              MPI-IO hint value passed to the non-restart IO. Has to be matched byio.system_write_mpiio_hint_key.</span>
<span class="na">restart_write_mpiio_hint_key                                               MPI-IO hint key passed to the restart IO. Has to be matched byio.restart_write_mpiio_hint_value.</span>
<span class="na">restart_write_mpiio_hint_value                                             MPI-IO hint value passed to the restart IO. Has to be matched byio.restart_write_mpiio_hint_key.</span>
<span class="na">restart_read_mpiio_hint_key                                                MPI-IO hint key passed to the restart IO. Has to be matched byio.restart_read_mpiio_hint_value.</span>
<span class="na">restart_read_mpiio_hint_value                                              MPI-IO hint value passed to the restart IO. Has to be matched byio.restart_read_mpiio_hint_key.</span>
<span class="na">write_initial_state                                                        Write initial state, not even the 0.5 dt propagation is done. Do not use forrestarting.</span>
<span class="na">write_full_bgb_data                                                        Write a dedicated file containing all BGB components and first derivatives, thenexit.</span>
<span class="na">restart_walltime_interval                                                  Save the complete simulation in given walltime intervals. Negative valuesdisable writes.</span>
<span class="na">number_of_restarts                                                         Exit the simulation after certain number of walltime-based restarts.</span>
<span class="na">vlsv_buffer_size                                                           Buffer size passed to VLSV writer (bytes, up to uint64_t), default 0 as this issensible on sisu</span>
<span class="na">write_restart_stripe_factor                                                Stripe factor for restart and initial grid writing. Default 0 to inherit.</span>
<span class="na">write_system_stripe_factor                                                 Stripe factor for bulk file writing. Default 0 to inherit.</span>
<span class="na">write_as_float                                                             If true, write in floats instead of doubles</span>
<span class="na">restart_write_path                                                         Path to the location where restart files should be written. Defaults to thelocal directory, also if the specified destination is not writeable.</span>
<span class="na">system_write_all_data_reducers                                             If 0 don&#39;t write all DROs, if 1 do write them.</span>
<span class="na">diagnostic_write_all_data_reducers                                         Write all available diagnostic reducers</span>

<span class="k">[restart]</span>
<span class="na">write_as_float                                                             If true, write restart fields in floats instead of doubles</span>
<span class="na">filename                                                                   Restart from this vlsv file. No restart if empty file.</span>
<span class="na">overrideReadFsGridDecompositionX                                           Manual FsGridDecomposition for field solver grid stored in a restart file.</span>
<span class="na">overrideReadFsGridDecompositionY                                           Manual FsGridDecomposition for field solver grid stored in a restart file.</span>
<span class="na">overrideReadFsGridDecompositionZ                                           Manual FsGridDecomposition for field solver grid stored in a restart file.</span>

<span class="k">[gridbuilder]</span>
<span class="na">geometry                                                                   Simulation geometry XY4D,XZ4D,XY5D,XZ5D,XYZ6D</span>
<span class="na">x_min                                                                      Minimum value of the x-coordinate.</span>
<span class="na">x_max                                                                      Minimum value of the x-coordinate.</span>
<span class="na">y_min                                                                      Minimum value of the y-coordinate.</span>
<span class="na">y_max                                                                      Minimum value of the y-coordinate.</span>
<span class="na">z_min                                                                      Minimum value of the z-coordinate.</span>
<span class="na">z_max                                                                      Minimum value of the z-coordinate.</span>
<span class="na">x_length                                                                   Number of cells in x-direction in initial grid.</span>
<span class="na">y_length                                                                   Number of cells in y-direction in initial grid.</span>
<span class="na">z_length                                                                   Number of cells in z-direction in initial grid.</span>
<span class="na">dt                                                                         Initial timestep in seconds.</span>
<span class="na">t_max                                                                      Maximum simulation time, in seconds. If timestep_max limit is hit first thistime will never be reached</span>
<span class="na">timestep_max                                                               Max. value for timesteps. If t_max limit is hit first, this step will never bereached</span>

<span class="k">[fieldsolver]</span>
<span class="na">maxWaveVelocity                                                            Maximum wave velocity allowed in the fastest velocity determination in m/s,default unlimited</span>
<span class="na">maxSubcycles                                                               Maximum allowed field solver subcycles</span>
<span class="na">resistivity                                                                Resistivity for the eta*J term in Ohm&#39;s law.</span>
<span class="na">diffusiveEterms                                                            Enable diffusive terms in the computation of E</span>
<span class="na">ohmHallTerm                                                                Enable/choose spatial order of the Hall term in Ohm&#39;s law. 0</span><span class="o">:</span><span class="w"> </span><span class="s">off, 1: 1stspatial order, 2: 2nd spatial order</span>
<span class="na">ohmGradPeTerm                                                              Enable/choose spatial order of the electron pressure gradient term in Ohm&#39;s law.0</span><span class="o">:</span><span class="w"> </span><span class="s">off, 1: 1st spatial order.</span>
<span class="na">electronTemperature                                                        Upstream electron temperature to be used for the electron pressure gradient term(K).</span>
<span class="na">electronDensity                                                            Upstream electron density to be used for the electron pressure gradient term(m^-3).</span>
<span class="na">electronPTindex                                                            Polytropic index for electron pressure gradient term. 0 is isobaric, 1 isisothermal, 1.667 is adiabatic electrons,</span>
<span class="na">maxCFL                                                                     The maximum CFL limit for field propagation. Used to set timestep ifdynamic_timestep is true.</span>
<span class="na">minCFL                                                                     The minimum CFL limit for field propagation. Used to set timestep ifdynamic_timestep is true.</span>
<span class="na">manualFsGridDecompositionX                                                 Manual FsGridDecomposition for field solver grid.</span>
<span class="na">manualFsGridDecompositionY                                                 Manual FsGridDecomposition for field solver grid.</span>
<span class="na">manualFsGridDecompositionZ                                                 Manual FsGridDecomposition for field solver grid.</span>

<span class="k">[vlasovsolver]</span>
<span class="na">maxSlAccelerationRotation                                                  Maximum rotation angle (degrees) allowed by the Semi-Lagrangian solver (Use &gt;25values with care)</span>
<span class="na">maxSlAccelerationSubcycles                                                 Maximum number of subcycles for acceleration</span>
<span class="na">maxCFL                                                                     The maximum CFL limit for vlasov propagation in ordinary space. Used to settimestep if dynamic_timestep is true.</span>
<span class="na">minCFL                                                                     The minimum CFL limit for vlasov propagation in ordinary space. Used to settimestep if dynamic_timestep is true.</span>
<span class="na">accelerateMaxwellianBoundaries                                             Propagate maxwellian boundary cell contents in velocity space. Default false.</span>

<span class="k">[loadBalance]</span>
<span class="na">algorithm                                                                  Load balancing algorithm to be used</span>
<span class="na">tolerance                                                                  Load imbalance tolerance</span>
<span class="na">rebalanceInterval                                                          Load rebalance interval (steps)</span>
<span class="na">optionKey                                                                  Zoltan option key. Has to be matched by loadBalance.optionValue.</span>
<span class="na">optionValue                                                                Zoltan option value. Has to be matched by loadBalance.optionKey.</span>

<span class="k">[variables]</span>
<span class="na">output                                                                     List of data reduction operators (DROs) to add to the grid file output.  Eachvariable to be added has to be on a new line output</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">XXX. Names are caseinsensitive.  Available (20230628): fg_b fg_b_background fg_b_perturbedfg_b_background_vol fg_derivs_b_background fg_e vg_rhom vg_rhoqpopulations_vg_rho fg_rhom fg_rhoq vg_v fg_v populations_vg_vpopulations_vg_moments_thermal populations_vg_moments_nonthermalpopulations_vg_effectivesparsitythreshold populations_vg_rho_loss_adjustpopulations_vg_energydensity populations_vg_precipitationdifferentialfluxpopulations_vg_heatflux populations_vg_nonmaxwellianity vg_maxdt_accelerationvg_maxdt_translation populations_vg_maxdt_accelerationpopulations_vg_maxdt_translation fg_maxdt_fieldsolver vg_rank fg_rankfg_amr_level vg_loadbalance_weight vg_boundarytype fg_boundarytypevg_boundarylayer fg_boundarylayer populations_vg_blocks vg_f_savedpopulations_vg_acceleration_subcycles vg_e_vol fg_e_vol fg_e_hall vg_e_gradpefg_b_vol vg_b_vol vg_b_background_vol vg_b_perturbed_vol vg_pressure fg_pressurepopulations_vg_ptensor vg_b_vol_derivatives fg_derivs ig_fac ig_latitude ig_chi0ig_cellarea ig_upmappedarea ig_sigmap ig_sigmah ig_sigmaparallel ig_rhonig_electrontemp ig_solverinternals ig_upmappednodecoords ig_upmappedbig_openclosed ig_potential ig_precipitation ig_deltaphi ig_inplanecurrent ig_big_e vg_drift vg_ionospherecoupling vg_connection vg_fluxrope fg_curvaturevg_amr_drho vg_amr_du vg_amr_dpsq vg_amr_dbsq vg_amr_db vg_amr_alphavg_amr_reflevel vg_amr_jperb vg_amr_translate_comm vg_gridcoordinatesfg_gridcoordinates</span>
<span class="na">diagnostic                                                                 List of data reduction operators (DROs) to add to the diagnostic runtime output.Each variable to be added has to be on a new line diagnostic</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">XXX. Names arecase insensitive. Available (20221221): populations_vg_blocks vg_rhompopulations_vg_rho_loss_adjust vg_loadbalance_weight vg_maxdt_accelerationvg_maxdt_translation fg_maxdt_fieldsolver populations_vg_maxdt_accelerationpopulations_vg_maxdt_translation populations_vg_maxdistributionfunctionpopulations_vg_mindistributionfunction</span>

<span class="k">[variables_deprecated]</span>
<span class="na">output                                                                     List of deprecated names for data reduction operators (DROs). Names are caseinsensitive. Available (20190521)</span><span class="o">:</span><span class="w"> </span><span class="s">B BackgroundB fg_BackgroundB PerturbedBfg_PerturbedB E Rhom Rhoq populations_Rho V populations_Vpopulations_moments_Backstream populations_moments_NonBackstreampopulations_moments_thermal populations_moments_nonthermal populations_minvaluepopulations_EffectiveSparsityThreshold populations_RhoLossAdjustpopulations_rho_loss_adjustpopulations_EnergyDensitypopulations_PrecipitationFlux populations_precipitationdifferentialfluxLBweightvg_lbweight vg_loadbalanceweight MaxVdt MaxRdt populations_MaxVdtpopulations_MaxRdt populations_maxdt_acceleration populations_maxdt_translationMaxFieldsdt fg_maxfieldsdtMPIrank FsGridRank FsGridBoundaryType BoundaryTypeFsGridBoundaryLayer BoundaryLayer populations_Blocks fSavedvg_fsavedpopulations_accSubcycles populations_acceleration_subcyclesVolE vg_VolEEvol E_vol fg_VolE fg_Evol HallE fg_HallE GradPeE e_gradpe VolB vg_VolB fg_VolBB_vol Bvol vg_Bvol fg_volB fg_Bvol BackgroundVolB PerturbedVolB Pressurevg_Pressure fg_Pressure populations_PTensor BVOLderivs b_vol_derivs</span>
<span class="na">diagnostic                                                                 List of deprecated data reduction operators (DROs) to add to the diagnosticruntime output. Names are case insensitive. Available (20201111)</span><span class="o">:</span><span class="w"> </span><span class="s">rhompopulations_rholossadjust populations_rho_loss_adjust populations_blockslbweight loadbalance_weight vg_lbweight vg_loadbalanceweight maxvdtmaxdt_acceleration maxrdt maxdt_translation populations_maxvdtpopulations_maxrdt populations_maxdt_acceleration populations_maxdt_translationpopulations_maxdistributionfunction populations_mindistributionfunctionmaxfieldsdt maxdt_fieldsolver fg_maxfieldsdt</span>

<span class="k">[bailout]</span>
<span class="na">write_restart                                                              If 1, write a restart file on bailout. Gets reset when sending a STOP (1) or aKILL (0).</span>
<span class="na">min_dt                                                                     Minimum time step below which bailout occurs (s).</span>
<span class="na">max_memory                                                                 Maximum amount of memory used per node (in GiB) over which bailout occurs.</span>
<span class="na">velocity_space_wall_block_margin                                           Distance from the velocity space limits in blocks, if the distribution functionreaches that distance from the wall we bail out to avoid hitting the wall.</span>

<span class="k">[VAMR]</span>
<span class="na">vel_refinement_criterion                                                   Name of the velocity refinement criterion</span>
<span class="na">max_velocity_level                                                         Maximum velocity mesh refinement level</span>
<span class="na">refine_limit                                                               If the refinement criterion function returns a larger value than this, block isrefined</span>
<span class="na">coarsen_limit                                                              If the refinement criterion function returns a smaller value than this, blockcan be coarsened</span>

<span class="k">[AMR]</span>
<span class="na">max_spatial_level                                                          Maximum absolute spatial mesh refinement level</span>
<span class="na">max_allowed_spatial_level                                                  Maximum currently allowed spatial mesh refinement level</span>
<span class="na">should_refine                                                              If false, do not refine Vlasov grid regardless of max spatial level</span>
<span class="na">adapt_refinement                                                           If true, re-refine vlasov grid every refine_cadence balance</span>
<span class="na">refine_on_restart                                                          If true, re-refine vlasov grid on restart. DEPRECATED, consider using the DOMRcommand</span>
<span class="na">force_refinement                                                           If true, refine/unrefine the vlasov grid to match the config on restart</span>
<span class="na">should_filter                                                              If true, filter vlasov grid with boxcar filter on restart</span>
<span class="na">use_alpha1                                                                 Use the maximum of dimensionless gradients alpha_1 as a refinement index</span>
<span class="na">alpha1_refine_threshold                                                    Determines the minimum value of alpha_1 to refine cells</span>
<span class="na">alpha1_coarsen_threshold                                                   Determines the maximum value of alpha_1 to unrefine cells, default half of therefine threshold</span>
<span class="na">use_alpha2                                                                 Use J/B_perp as a refinement index</span>
<span class="na">alpha2_refine_threshold                                                    Determines the minimum value of alpha_2 to refine cells</span>
<span class="na">alpha2_coarsen_threshold                                                   Determines the maximum value of alpha_2 to unrefine cells, default half of therefine threshold</span>
<span class="na">refine_cadence                                                             Refine every nth load balance</span>
<span class="na">refine_after                                                               Start refinement after this many simulation seconds</span>
<span class="na">refine_radius                                                              Maximum distance from Earth to refine</span>
<span class="na">alpha1_drho_weight                                                         Multiplier for delta rho in alpha calculation</span>
<span class="na">alpha1_du_weight                                                           Multiplier for delta U in alpha calculation</span>
<span class="na">alpha1_dpsq_weight                                                         Multiplier for delta p squared in alpha calculation</span>
<span class="na">alpha1_dbsq_weight                                                         Multiplier for delta B squared in alpha calculation</span>
<span class="na">alpha1_db_weight                                                           Multiplier for delta B in alpha calculation</span>
<span class="na">number_of_boxes                                                            How many boxes to be refined, that number of centers and sizes have to then bedefined as well.</span>
<span class="na">box_half_width_x                                                           Half width in x of the box that is refined</span>
<span class="na">box_half_width_y                                                           Half width in y of the box that is refined</span>
<span class="na">box_half_width_z                                                           Half width in z of the box that is refined</span>
<span class="na">box_center_x                                                               x coordinate of the center of the box that is refined</span>
<span class="na">box_center_y                                                               y coordinate of the center of the box that is refined</span>
<span class="na">box_center_z                                                               z coordinate of the center of the box that is refined</span>
<span class="na">box_max_level                                                              max refinement level of the box that is refined</span>
<span class="na">transShortPencils                                                          if true, use one-cell pencils</span>
<span class="na">filterpasses                                                               AMR filter passes for each individual refinement level</span>

<span class="k">[fieldtracing]</span>
<span class="na">fieldLineTracer                                                            Field line tracing method to use for coupling ionosphere and magnetosphere(options are</span><span class="o">:</span><span class="w"> </span><span class="s">Euler, BS)</span>
<span class="na">tracer_max_allowed_error                                                   Maximum allowed error for the adaptive field line tracers</span>
<span class="na">tracer_max_attempts                                                        Maximum allowed attempts for the adaptive field line tracers</span>
<span class="na">tracer_min_dx                                                              Minimum allowed field line tracer step length for the adaptive field linetracers (m)</span>
<span class="na">fullbox_and_fluxrope_max_absolute_distance_to_trace                        Maximum absolute distance in m to trace along the field line before ending.Defaults to the sum of the simulation box edge lengths LX+LY+LZ if set &lt;</span><span class="o">=</span><span class="w"> </span><span class="s">0.</span>
<span class="na">fullbox_max_incomplete_cells                                               Maximum fraction of cells left incomplete when stopping tracing loop for fullbox tracing. Defaults to zero to process all, will be slow at scale! Bothfluxrope_max_incomplete_cells and fullbox_max_incomplete_cells will be achieved.</span>
<span class="na">fluxrope_max_incomplete_cells                                              Maximum fraction of cells left incomplete when stopping loop for flux ropetracing. Defaults to zero to process all, will be slow at scale! Bothfluxrope_max_incomplete_cells and fullbox_max_incomplete_cells will be achieved.</span>
<span class="na">use_reconstruction_cache                                                   Use the cache to store reconstruction coefficients. (0</span><span class="o">:</span><span class="w"> </span><span class="s">don&#39;t, 1: use)</span>
<span class="na">fluxrope_max_curvature_radii_to_trace                                      Maximum number of seedpoint curvature radii to trace forward and backward fromeach DCCRG cell to find flux ropes</span>
<span class="na">fluxrope_max_curvature_radii_extent                                        Maximum extent in seedpoint curvature radii from the seed a field line isallowed to extend to be counted as a flux rope</span>

<span class="k">[&lt;population&gt;_properties]</span>
<span class="na">charge                                                                     Particle charge, in units of elementary charges (int)</span>
<span class="na">mass_units                                                                 Units in which particle mass is given, either &#39;PROTON&#39; or &#39;ELECTRON&#39; (string)</span>
<span class="na">mass                                                                       Particle mass in given units (float)</span>

<span class="k">[&lt;population&gt;_sparse]</span>
<span class="na">minValue                                                                   Minimum value of distribution function in any cell of a velocity block for theblock to be considered to have contents</span>
<span class="na">blockAddWidthV                                                             Number of layers of blocks that are kept in velocity space around the blockswith content</span>
<span class="na">conserve_mass                                                              If true, then mass is conserved by scaling the dist. func. in the remainingblocks</span>
<span class="na">dynamicAlgorithm                                                           Type of algorithm used for calculating the dynamic minValue; 0</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">none, 1 =linear algorithm based on rho, 2 = linear algorithm based on Blocks, (Examplelinear algorithm: y = kx+b, where dynamicMinValue1=k*dynamicBulkValue1 + b, anddynamicMinValue2 = k*dynamicBulkValue2 + b</span>
<span class="na">dynamicMinValue1                                                           The minimum value for the dynamic minValue</span>
<span class="na">dynamicMinValue2                                                           The maximum value (value 2) for the dynamic minValue</span>
<span class="na">dynamicBulkValue1                                                          Minimum value for the dynamic algorithm range, so for example ifdynamicAlgorithm</span><span class="o">=</span><span class="s">1 then for sparse.dynamicBulkValue1 = 1e3,sparse.dynamicBulkValue2=1e5, we apply the algorithm to cells for which1e3&lt;cell.rho&lt;1e5</span>
<span class="na">dynamicBulkValue2                                                          Maximum value for the dynamic algorithm range, so for example ifdynamicAlgorithm</span><span class="o">=</span><span class="s">1 then for sparse.dynamicBulkValue1 = 1e3,sparse.dynamicBulkValue2=1e5, we apply the algorithm to cells for which1e3&lt;cell.rho&lt;1e5</span>

<span class="k">[&lt;population&gt;_vspace]</span>
<span class="na">vx_min                                                                     Minimum value for velocity mesh vx-coordinates.</span>
<span class="na">vx_max                                                                     Maximum value for velocity mesh vx-coordinates.</span>
<span class="na">vy_min                                                                     Minimum value for velocity mesh vy-coordinates.</span>
<span class="na">vy_max                                                                     Maximum value for velocity mesh vx-coordinates.</span>
<span class="na">vz_min                                                                     Minimum value for velocity mesh vz-coordinates.</span>
<span class="na">vz_max                                                                     Maximum value for velocity mesh vx-coordinates.</span>
<span class="na">vx_length                                                                  Initial number of velocity blocks in vx-direction.</span>
<span class="na">vy_length                                                                  Initial number of velocity blocks in vy-direction.</span>
<span class="na">vz_length                                                                  Initial number of velocity blocks in vz-direction.</span>
<span class="na">max_refinement_level                                                       Maximum allowed mesh refinement level.</span>

<span class="k">[&lt;population&gt;_thermal]</span>
<span class="na">vx                                                                         Center coordinate for the maxwellian distribution. Used for calculating thesuprathermal moments.</span>
<span class="na">vy                                                                         Center coordinate for the maxwellian distribution. Used for calculating thesuprathermal moments.</span>
<span class="na">vz                                                                         Center coordinate for the maxwellian distribution. Used for calculating thesuprathermal moments.</span>
<span class="na">radius                                                                     Radius of the maxwellian distribution. Used for calculating the suprathermalmoments. If set to 0 (default), the thermal/suprathermal DROs are skipped.</span>

<span class="k">[&lt;population&gt;_precipitation]</span>
<span class="na">nChannels                                                                  Number of energy channels for precipitation differential flux evaluation</span>
<span class="na">emin                                                                       Lowest energy channel (in eV) for precipitation differential flux evaluation</span>
<span class="na">emax                                                                       Highest energy channel (in eV) for precipitation differential flux evaluation</span>
<span class="na">lossConeAngle                                                              Fixed loss cone opening angle (in deg) for precipitation differential fluxevaluation</span>

<span class="k">[&lt;population&gt;_energydensity]</span>
<span class="na">limit1                                                                     Lower limit of second bin for energy density, given in units of solar wind ramenergy.</span>
<span class="na">limit2                                                                     Lower limit of third bin for energy density, given in units of solar wind ramenergy.</span>
<span class="na">solarwindspeed                                                             Incoming solar wind velocity magnitude in m/s. Used for calculating energydensities.</span>
<span class="na">solarwindenergy                                                            Incoming solar wind ram energy in eV. Used for calculating energy densities.</span>

<span class="k">[boundaries]</span>
<span class="na">boundary                                                                   List of boundary condition (BC) types to be used. Each boundary condition to beused has to be on a new line boundary</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">YYY. Available options are: Outflow,Ionosphere, Copysphere, Maxwellian.</span>
<span class="na">periodic_x                                                                 Set the grid periodicity in x-direction. &#39;yes&#39;(default)/&#39;no&#39;.</span>
<span class="na">periodic_y                                                                 Set the grid periodicity in y-direction. &#39;yes&#39;(default)/&#39;no&#39;.</span>
<span class="na">periodic_z                                                                 Set the grid periodicity in z-direction. &#39;yes&#39;(default)/&#39;no&#39;.</span>

<span class="k">[ionosphere]</span>
<span class="na">centerX                                                                    X coordinate of ionosphere center (m)</span>
<span class="na">centerY                                                                    Y coordinate of ionosphere center (m)</span>
<span class="na">centerZ                                                                    Z coordinate of ionosphere center (m)</span>
<span class="na">radius                                                                     Radius of the inner simulation boundary (unit is assumed to be R_E if value &lt;1000, otherwise m).</span>
<span class="na">innerRadius                                                                Radius of the ionosphere model (m).</span>
<span class="na">geometry                                                                   Select the geometry of the ionosphere, 0</span><span class="o">:</span><span class="w"> </span><span class="s">inf-norm (diamond), 1: 1-norm(square), 2: 2-norm (circle, DEFAULT), 3: 2-norm cylinder aligned with y-axis,use with polar plane/line dipole.</span>
<span class="na">precedence                                                                 Precedence value of the ionosphere system boundary condition (integer), thehigher the stronger.</span>
<span class="na">reapplyUponRestart                                                         If 0 (default), keep going with the state existing in the restart file. If 1,calls again applyInitialState. Can be used to change boundary conditionbehaviour during a run.</span>
<span class="na">baseShape                                                                  Select the seed mesh geometry for the spherical ionosphere grid. Options are</span><span class="o">:</span><span class="s">sphericalFibonacci, tetrahedron, icosahedron.</span>
<span class="na">conductivityModel                                                          Select ionosphere conductivity tensor construction model. Options are</span><span class="o">:</span><span class="w"> </span><span class="s">0=GUMICSstyle (Vertical B, only SigmaH and SigmaP), 1=Ridley et al 2004 (1000 mholongitudinal conductivity), 2=Koskinen 2011 full conductivity tensor.</span>
<span class="na">ridleyParallelConductivity                                                 Constant parallel conductivity value. 1000 mho is given without justification byRidley et al 2004.</span>
<span class="na">fibonacciNodeNum                                                           Number of nodes in the spherical fibonacci mesh.</span>
<span class="na">refineMinLatitude                                                          Refine the grid polewards of the given latitude. Multiple of these lines can begiven for successive refinement, paired up with refineMaxLatitude lines.</span>
<span class="na">refineMaxLatitude                                                          Refine the grid equatorwards of the given latitude. Multiple of these lines canbe given for successive refinement, paired up with refineMinLatitude lines.</span>
<span class="na">atmosphericModelFile                                                       Filename to read the MSIS atmosphere data from (default</span><span class="o">:</span><span class="w"> </span><span class="s">NRLMSIS.dat)</span>
<span class="na">recombAlpha                                                                Ionospheric recombination parameter (m^3/s)</span>
<span class="na">ionizationModel                                                            Ionospheric electron production rate model. Options are</span><span class="o">:</span><span class="w"> </span><span class="s">Rees1963, Rees1989,SergienkoIvanov (default).</span>
<span class="na">innerBoundaryVDFmode                                                       Inner boundary VDF construction method. Options ar</span><span class="o">:</span><span class="w"> </span><span class="s">FixedMoments,AverageMoments, AverageAllMoments, CopyAndLosscone, ForceL2EXB.</span>
<span class="na">F10_7                                                                      Solar 10.7 cm radio flux (sfu</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">10^{-22} W/m^2)</span>
<span class="na">backgroundIonisation                                                       Background ionoisation due to cosmic rays (mho)</span>
<span class="na">solverMaxIterations                                                        Maximum number of iterations for the conjugate gradient solver</span>
<span class="na">solverRelativeL2ConvergenceThreshold                                       Convergence threshold for the relative L2 metric</span>
<span class="na">solverMaxFailureCount                                                      Maximum number of iterations allowed to diverge before restarting the ionospheresolver</span>
<span class="na">solverMaxErrorGrowthFactor                                                 Maximum allowed factor of growth with respect to the minimum error beforerestarting the ionosphere solver</span>
<span class="na">solverGaugeFixing                                                          Gauge fixing method of the ionosphere solver. Options are</span><span class="o">:</span><span class="w"> </span><span class="s">pole, integral,equator</span>
<span class="na">shieldingLatitude                                                          Latitude below which the potential is set to zero in the equator gauge fixingscheme (degree)</span>
<span class="na">solverPreconditioning                                                      Use preconditioning for the solver? (0/1)</span>
<span class="na">solverUseMinimumResidualVariant                                            Use minimum residual variant</span>
<span class="na">solverToggleMinimumResidualVariant                                         Toggle use of minimum residual variant at every solver restart</span>
<span class="na">earthAngularVelocity                                                       Angular velocity of inner boundary convection, in rad/s</span>
<span class="na">plasmapauseL                                                               L-shell at which the plasmapause resides (for corotation)</span>
<span class="na">downmapRadius                                                              Radius from which FACs are coupled down into the ionosphere. Units are assumedto be RE if value &lt; 1000, otherwise m. If -1</span><span class="o">:</span><span class="w"> </span><span class="s">use inner boundary cells.</span>
<span class="na">unmappedNodeRho                                                            Electron density of ionosphere nodes that do not connect to the magnetospheredomain.</span>
<span class="na">unmappedNodeTe                                                             Electron temperature of ionosphere nodes that do not connect to themagnetosphere domain.</span>
<span class="na">couplingTimescale                                                          Magnetosphere-&gt;Ionosphere coupling timescale (seconds, 0</span><span class="o">=</span><span class="s">immediate coupling</span>
<span class="na">couplingInterval                                                           Time interval at which the ionosphere is solved (seconds)</span>

<span class="k">[&lt;population&gt;_ionosphere]</span>
<span class="na">rho                                                                        Number density of the ionosphere (m^-3)</span>
<span class="na">T                                                                          Temperature of the ionosphere (K)</span>
<span class="na">VX0                                                                        Bulk velocity of ionospheric distribution function in X direction (m/s)</span>
<span class="na">VY0                                                                        Bulk velocity of ionospheric distribution function in X direction (m/s)</span>
<span class="na">VZ0                                                                        Bulk velocity of ionospheric distribution function in X direction (m/s)</span>

<span class="k">[copysphere]</span>
<span class="na">centerX                                                                    X coordinate of copysphere center (m)</span>
<span class="na">centerY                                                                    Y coordinate of copysphere center (m)</span>
<span class="na">centerZ                                                                    Z coordinate of copysphere center (m)</span>
<span class="na">radius                                                                     Radius of copysphere (m).</span>
<span class="na">geometry                                                                   Select the geometry of the copysphere, 0</span><span class="o">:</span><span class="w"> </span><span class="s">inf-norm (diamond), 1: 1-norm(square), 2: 2-norm (circle, DEFAULT), 3: 2-norm cylinder aligned with y-axis,use with polar plane/line dipole.</span>
<span class="na">precedence                                                                 Precedence value of the copysphere system boundary condition (integer), thehigher the stronger.</span>
<span class="na">reapplyUponRestart                                                         If 0 (default), keep going with the state existing in the restart file. If 1,calls again applyInitialState. Can be used to change boundary conditionbehaviour during a run.</span>
<span class="na">zeroPerB                                                                   If 0 (default), normal copysphere behaviour of magnetic field at inner boundary.If 1, keep magnetic field static at the inner boundary</span>

<span class="k">[&lt;population&gt;_copysphere]</span>
<span class="na">rho                                                                        Number density of the copysphere (m^-3)</span>
<span class="na">T                                                                          Temperature of the copysphere (K)</span>
<span class="na">VX0                                                                        Bulk velocity of copyspheric distribution function in X direction (m/s)</span>
<span class="na">VY0                                                                        Bulk velocity of copyspheric distribution function in X direction (m/s)</span>
<span class="na">VZ0                                                                        Bulk velocity of copyspheric distribution function in X direction (m/s)</span>
<span class="na">fluffiness                                                                 Inertia of boundary smoothing when copying neighbour&#39;s moments and velocitydistributions (0</span><span class="o">=</span><span class="s">completely constant boundaries, 1=neighbours are interpolatedimmediately).</span>

<span class="k">[outflow]</span>
<span class="na">faceNoFields                                                               List of faces on which no field outflow boundary conditions are to be applied([xyz][+-]).</span>
<span class="na">precedence                                                                 Precedence value of the outflow system boundary condition (integer), the higherthe stronger.</span>
<span class="na">reapplyUponRestart                                                         If 0 (default), keep going with the state existing in the restart file. If 1,calls again applyInitialState. Can be used to change boundary conditionbehaviour during a run.</span>

<span class="k">[&lt;population&gt;_outflow]</span>
<span class="na">reapplyFaceUponRestart                                                     List of faces on which outflow boundary conditions are to be reapplied uponrestart ([xyz][+-]).</span>
<span class="na">face                                                                       List of faces on which outflow boundary conditions are to be applied([xyz][+-]).</span>
<span class="na">vlasovScheme_face_x+                                                       Scheme to use on the face x+ (Copy, Limit, None)</span>
<span class="na">vlasovScheme_face_x-                                                       Scheme to use on the face x- (Copy, Limit, None)</span>
<span class="na">vlasovScheme_face_y+                                                       Scheme to use on the face y+ (Copy, Limit, None)</span>
<span class="na">vlasovScheme_face_y-                                                       Scheme to use on the face y- (Copy, Limit, None)</span>
<span class="na">vlasovScheme_face_z+                                                       Scheme to use on the face z+ (Copy, Limit, None)</span>
<span class="na">vlasovScheme_face_z-                                                       Scheme to use on the face z- (Copy, Limit, None)</span>
<span class="na">quench                                                                     Factor by which to quench the inflowing parts of the velocity distributionfunction.</span>

<span class="k">[maxwellian]</span>
<span class="na">face                                                                       List of faces on which set Maxwellian boundary conditions are to be applied([xyz][+-]).</span>
<span class="na">precedence                                                                 Precedence value of the set Maxwellian boundary condition (integer), the higherthe stronger.</span>
<span class="na">reapplyUponRestart                                                         If 0 (default), keep going with the state existing in the restart file. If 1,calls again applyInitialState. Can be used to change boundary conditionbehaviour during a run.</span>
<span class="na">t_interval                                                                 Time interval in seconds for applying the varying inflow condition.</span>

<span class="k">[&lt;population&gt;_maxwellian]</span>
<span class="na">file_x+                                                                    Input files for the set Maxwellian inflow parameters on face x+. Data format perline</span><span class="o">:</span><span class="w"> </span><span class="s">time (s) density (p/m^3) Temperature (K) Vx Vy Vz (m/s) Bx By Bz (T).</span>
<span class="na">file_x-                                                                    Input files for the set Maxwellian inflow parameters on face x-. Data format perline</span><span class="o">:</span><span class="w"> </span><span class="s">time (s) density (p/m^3) Temperature (K) Vx Vy Vz (m/s) Bx By Bz (T).</span>
<span class="na">file_y+                                                                    Input files for the set Maxwellian inflow parameters on face y+. Data format perline</span><span class="o">:</span><span class="w"> </span><span class="s">time (s) density (p/m^3) Temperature (K) Vx Vy Vz (m/s) Bx By Bz (T).</span>
<span class="na">file_y-                                                                    Input files for the set Maxwellian inflow parameters on face y-. Data format perline</span><span class="o">:</span><span class="w"> </span><span class="s">time (s) density (p/m^3) Temperature (K) Vx Vy Vz (m/s) Bx By Bz (T).</span>
<span class="na">file_z+                                                                    Input files for the set Maxwellian inflow parameters on face z+. Data format perline</span><span class="o">:</span><span class="w"> </span><span class="s">time (s) density (p/m^3) Temperature (K) Vx Vy Vz (m/s) Bx By Bz (T).</span>
<span class="na">file_z-                                                                    Input files for the set Maxwellian inflow parameters on face z-. Data format perline</span><span class="o">:</span><span class="w"> </span><span class="s">time (s) density (p/m^3) Temperature (K) Vx Vy Vz (m/s) Bx By Bz (T).</span>
<span class="na">dynamic                                                                    Boolean value, is the set Maxwellian inflow dynamic in time or not.</span>

<span class="k">[Alfven]</span>
<span class="na">B0                                                                         Guiding field value (T)</span>
<span class="na">Bx_guiding                                                                 Guiding field x component</span>
<span class="na">By_guiding                                                                 Guiding field y component</span>
<span class="na">Bz_guiding                                                                 Guiding field z component</span>
<span class="na">Wavelength                                                                 Wavelength (m)</span>
<span class="na">A_mag                                                                      Amplitude of the magnetic perturbation</span>

<span class="k">[&lt;population&gt;_Alfven]</span>
<span class="na">rho                                                                        Number density (m^-3)</span>
<span class="na">Temperature                                                                Temperature (K)</span>
<span class="na">A_vel                                                                      Amplitude of the velocity perturbation</span>

<span class="k">[Diffusion]</span>
<span class="na">B0                                                                         Background field value (T)</span>

<span class="k">[&lt;population&gt;_Diffusion]</span>
<span class="na">rho                                                                        Number density (m^-3)</span>
<span class="na">Temperature                                                                Temperature (K)</span>
<span class="na">Scale_x                                                                    Scale length in x (m)</span>
<span class="na">Scale_y                                                                    Scale length in y (m)</span>

<span class="k">[Dispersion]</span>
<span class="na">B0                                                                         Guide magnetic field strength (T)</span>
<span class="na">magXPertAbsAmp                                                             Absolute amplitude of the magnetic perturbation along x (T)</span>
<span class="na">magYPertAbsAmp                                                             Absolute amplitude of the magnetic perturbation along y (T)</span>
<span class="na">magZPertAbsAmp                                                             Absolute amplitude of the magnetic perturbation along z (T)</span>
<span class="na">maxwCutoff                                                                 Cutoff for the maxwellian distribution</span>
<span class="na">angleXY                                                                    Orientation of the guide magnetic field with respect to the x-axis in x-y plane(rad)</span>
<span class="na">angleXZ                                                                    Orientation of the guide magnetic field with respect to the x-axis in x-z plane(rad)</span>

<span class="k">[&lt;population&gt;_Dispersion]</span>
<span class="na">VX0                                                                        Bulk velocity (m/s)</span>
<span class="na">VY0                                                                        Bulk velocity (m/s)</span>
<span class="na">VZ0                                                                        Bulk velocity (m/s)</span>
<span class="na">rho                                                                        Number density (m^-3)</span>
<span class="na">Temperature                                                                Temperature (K)</span>
<span class="na">densityPertRelAmp                                                          Relative amplitude of the density perturbation</span>
<span class="na">velocityPertAbsAmp                                                         Absolute amplitude of the velocity perturbation</span>

<span class="k">[Distributions]</span>
<span class="na">rho1                                                                       Number density, first peak (m^-3)</span>
<span class="na">rho2                                                                       Number density, second peak (m^-3)</span>
<span class="na">Tx1                                                                        Temperature, first peak (K)</span>
<span class="na">Tx2                                                                        Temperature, second peak (K)</span>
<span class="na">Ty1                                                                        Temperature, first peak (K)</span>
<span class="na">Ty2                                                                        Temperature, second peak (K)</span>
<span class="na">Tz1                                                                        Temperature, first peak (K)</span>
<span class="na">Tz2                                                                        Temperature, second peak (K)</span>
<span class="na">Vx1                                                                        Bulk velocity x component, first peak (m/s)</span>
<span class="na">Vx2                                                                        Bulk velocity x component, second peak (m/s)</span>
<span class="na">Vy1                                                                        Bulk velocity y component, first peak (m/s)</span>
<span class="na">Vy2                                                                        Bulk velocity y component, second peak (m/s)</span>
<span class="na">Vz1                                                                        Bulk velocity z component, first peak (m/s)</span>
<span class="na">Vz2                                                                        Bulk velocity z component, second peak (m/s)</span>
<span class="na">Bx                                                                         Magnetic field x component (T)</span>
<span class="na">By                                                                         Magnetic field y component (T)</span>
<span class="na">Bz                                                                         Magnetic field z component (T)</span>
<span class="na">dBx                                                                        Magnetic field x component cosine perturbation amplitude (T)</span>
<span class="na">dBy                                                                        Magnetic field y component cosine perturbation amplitude (T)</span>
<span class="na">dBz                                                                        Magnetic field z component cosine perturbation amplitude (T)</span>
<span class="na">magXPertAbsAmp                                                             Absolute amplitude of the random magnetic perturbation along x (T)</span>
<span class="na">magYPertAbsAmp                                                             Absolute amplitude of the random magnetic perturbation along y (T)</span>
<span class="na">magZPertAbsAmp                                                             Absolute amplitude of the random magnetic perturbation along z (T)</span>
<span class="na">rho1PertAbsAmp                                                             Absolute amplitude of the density perturbation, first peak</span>
<span class="na">rho2PertAbsAmp                                                             Absolute amplitude of the density perturbation, second peak</span>
<span class="na">lambda                                                                     B cosine perturbation wavelength (m)</span>

<span class="k">[Firehose]</span>
<span class="na">Bx                                                                         Magnetic field x component (T)</span>
<span class="na">By                                                                         Magnetic field y component (T)</span>
<span class="na">Bz                                                                         Magnetic field z component (T)</span>
<span class="na">lambda                                                                     Initial perturbation wavelength (m)</span>
<span class="na">amp                                                                        Initial perturbation amplitude (m)</span>

<span class="k">[&lt;population&gt;_Firehose]</span>
<span class="na">rho1                                                                       Number density, first peak (m^-3)</span>
<span class="na">rho2                                                                       Number density, second peak (m^-3)</span>
<span class="na">Tx1                                                                        Temperature x, first peak (K)</span>
<span class="na">Tx2                                                                        Temperature x, second peak (K)</span>
<span class="na">Ty1                                                                        Temperature y, first peak (K)</span>
<span class="na">Ty2                                                                        Temperature y, second peak (K)</span>
<span class="na">Tz1                                                                        Temperature z, first peak (K)</span>
<span class="na">Tz2                                                                        Temperature z, second peak (K)</span>
<span class="na">Vx1                                                                        Bulk velocity x component, first peak (m/s)</span>
<span class="na">Vx2                                                                        Bulk velocity x component, second peak (m/s)</span>
<span class="na">Vy1                                                                        Bulk velocity y component, first peak (m/s)</span>
<span class="na">Vy2                                                                        Bulk velocity y component, second peak (m/s)</span>
<span class="na">Vz1                                                                        Bulk velocity z component, first peak (m/s)</span>
<span class="na">Vz2                                                                        Bulk velocity z component, second peak (m/s)</span>

<span class="k">[Flowthrough]</span>
<span class="na">emptyBox                                                                   Is the simulation domain empty initially?</span>
<span class="na">densityModel                                                               Plasma density model, &#39;Maxwellian&#39; or &#39;SheetMaxwellian&#39;</span>
<span class="na">densityWidth                                                               Width of signal around origin</span>
<span class="na">Bx                                                                         Magnetic field x component (T)</span>
<span class="na">By                                                                         Magnetic field y component (T)</span>
<span class="na">Bz                                                                         Magnetic field z component (T)</span>

<span class="k">[&lt;population&gt;_Flowthrough]</span>
<span class="na">rho                                                                        Number density (m^-3)</span>
<span class="na">rhoBase                                                                    Background number density (m^-3)</span>
<span class="na">T                                                                          Temperature (K)</span>
<span class="na">VX0                                                                        Initial bulk velocity in x-direction</span>
<span class="na">VY0                                                                        Initial bulk velocity in y-direction</span>
<span class="na">VZ0                                                                        Initial bulk velocity in z-direction</span>

<span class="k">[Fluctuations]</span>
<span class="na">BX0                                                                        Background field value (T)</span>
<span class="na">BY0                                                                        Background field value (T)</span>
<span class="na">BZ0                                                                        Background field value (T)</span>
<span class="na">magXPertAbsAmp                                                             Amplitude of the magnetic perturbation along x</span>
<span class="na">magYPertAbsAmp                                                             Amplitude of the magnetic perturbation along y</span>
<span class="na">magZPertAbsAmp                                                             Amplitude of the magnetic perturbation along z</span>

<span class="k">[&lt;population&gt;_Fluctuations]</span>
<span class="na">rho                                                                        Number density (m^-3)</span>
<span class="na">Temperature                                                                Temperature (K)</span>
<span class="na">densityPertRelAmp                                                          Amplitude factor of the density perturbation</span>
<span class="na">velocityPertAbsAmp                                                         Amplitude of the velocity perturbation</span>
<span class="na">maxwCutoff                                                                 Cutoff for the maxwellian distribution</span>

<span class="k">[Harris]</span>
<span class="na">Scale_size                                                                 Harris sheet scale size (m)</span>
<span class="na">BX0                                                                        Magnetic field at infinity (T)</span>
<span class="na">BY0                                                                        Magnetic field at infinity (T)</span>
<span class="na">BZ0                                                                        Magnetic field at infinity (T)</span>

<span class="k">[&lt;population&gt;_Harris]</span>
<span class="na">Temperature                                                                Temperature (K)</span>
<span class="na">rho                                                                        Number density at infinity (m^-3)</span>

<span class="k">[KHB]</span>
<span class="na">rho1                                                                       Number density, this-&gt;TOP state (m^-3)</span>
<span class="na">rho2                                                                       Number density, this-&gt;BOTTOM state (m^-3)</span>
<span class="na">T1                                                                         Temperature, this-&gt;TOP state (K)</span>
<span class="na">T2                                                                         Temperature, this-&gt;BOTTOM state (K)</span>
<span class="na">Vx1                                                                        Bulk velocity x component, this-&gt;TOP state (m/s)</span>
<span class="na">Vx2                                                                        Bulk velocity x component, this-&gt;BOTTOM state (m/s)</span>
<span class="na">Vy1                                                                        Bulk velocity y component, this-&gt;TOP state (m/s)</span>
<span class="na">Vy2                                                                        Bulk velocity y component, this-&gt;BOTTOM state (m/s)</span>
<span class="na">Vz1                                                                        Bulk velocity z component, this-&gt;TOP state (m/s)</span>
<span class="na">Vz2                                                                        Bulk velocity z component, this-&gt;BOTTOM state (m/s)</span>
<span class="na">Bx1                                                                        Magnetic field x component, this-&gt;TOP state (T)</span>
<span class="na">Bx2                                                                        Magnetic field x component, this-&gt;BOTTOM state (T)</span>
<span class="na">By1                                                                        Magnetic field y component, this-&gt;TOP state (T)</span>
<span class="na">By2                                                                        Magnetic field y component, this-&gt;BOTTOM state (T)</span>
<span class="na">Bz1                                                                        Magnetic field z component, this-&gt;TOP state (T)</span>
<span class="na">Bz2                                                                        Magnetic field z component, this-&gt;BOTTOM state (T)</span>
<span class="na">lambda                                                                     Initial perturbation wavelength (m)</span>
<span class="na">amp                                                                        Initial perturbation amplitude (m)</span>
<span class="na">offset                                                                     Boundaries offset from 0 (m)</span>
<span class="na">transitionWidth                                                            Width of tanh transition for all changing values</span>

<span class="k">[Larmor]</span>
<span class="na">BX0                                                                        Background field value (T)</span>
<span class="na">BY0                                                                        Background field value (T)</span>
<span class="na">BZ0                                                                        Background field value (T)</span>
<span class="na">VX0                                                                        Bulk velocity in x</span>
<span class="na">VY0                                                                        Bulk velocity in y</span>
<span class="na">VZ0                                                                        Bulk velocuty in z</span>
<span class="na">rho                                                                        Number density (m^-3)</span>
<span class="na">Temperature                                                                Temperature (K)</span>
<span class="na">maxwCutoff                                                                 Cutoff for the maxwellian distribution</span>
<span class="na">Scale_x                                                                    Scale length in x (m)</span>
<span class="na">Scale_y                                                                    Scale length in y (m)</span>

<span class="k">[Magnetosphere]</span>
<span class="na">constBgBX                                                                  Constant flat Bx component in the whole simulation box. Default is none.</span>
<span class="na">constBgBY                                                                  Constant flat By component in the whole simulation box. Default is none.</span>
<span class="na">constBgBZ                                                                  Constant flat Bz component in the whole simulation box. Default is none.</span>
<span class="na">noDipoleInSW                                                               If set to 1, the dipole magnetic field is not set in the solar wind inflowcells. Default 0.</span>
<span class="na">dipoleScalingFactor                                                        Scales the field strength of the magnetic dipole compared to Earths.</span>
<span class="na">dipoleType                                                                 0</span><span class="o">:</span><span class="w"> </span><span class="s">Normal 3D dipole, 1: line-dipole for 2D polar simulations, 2: line-dipolewith mirror, 3: 3D dipole with mirror</span>
<span class="na">dipoleMirrorLocationX                                                      x-coordinate of dipole Mirror</span>
<span class="na">refine_L4radius                                                            Radius of L3-refined sphere or cap</span>
<span class="na">refine_L4nosexmin                                                          Low x-value of nose L3-refined box</span>
<span class="na">refine_L3radius                                                            Radius of L3-refined sphere or cap</span>
<span class="na">refine_L3nosexmin                                                          Low x-value of nose L3-refined box</span>
<span class="na">refine_L3tailheight                                                        Height in +-z of tail L3-refined box</span>
<span class="na">refine_L3tailwidth                                                         Width in +-y of tail L3-refined box</span>
<span class="na">refine_L3tailxmin                                                          Low x-value of tail L3-refined box</span>
<span class="na">refine_L3tailxmax                                                          High x-value of tail L3-refined box</span>
<span class="na">refine_L2radius                                                            Radius of L2-refined sphere</span>
<span class="na">refine_L2tailthick                                                         Thickness of L2-refined tail region</span>
<span class="na">refine_L1radius                                                            Radius of L1-refined sphere</span>
<span class="na">refine_L1tailthick                                                         Thickness of L1-refined tail region</span>
<span class="na">dipoleTiltPhi                                                              Magnitude of dipole tilt, in degrees</span>
<span class="na">dipoleTiltTheta                                                            Direction of dipole tilt from Sun-Earth-line, in degrees</span>
<span class="na">dipoleXFull                                                                X-coordinate up to which dipole is at full strength, in metres</span>
<span class="na">dipoleXZero                                                                X-coordinate after which dipole is at zero strength, in metres</span>
<span class="na">dipoleInflowBX                                                             Inflow magnetic field Bx component to which the vector potential dipoleconverges. Default is none.</span>
<span class="na">dipoleInflowBY                                                             Inflow magnetic field By component to which the vector potential dipoleconverges. Default is none.</span>
<span class="na">dipoleInflowBZ                                                             Inflow magnetic field Bz component to which the vector potential dipoleconverges. Default is none.</span>
<span class="na">zeroOutDerivativesX                                                        Zero Out Perpendicular components</span>
<span class="na">zeroOutDerivativesY                                                        Zero Out Perpendicular components</span>
<span class="na">zeroOutDerivativesZ                                                        Zero Out Perpendicular components</span>

<span class="k">[&lt;population&gt;_Magnetosphere]</span>
<span class="na">rho                                                                        Tail region number density (m^-3)</span>
<span class="na">T                                                                          Temperature (K)</span>
<span class="na">VX0                                                                        Initial bulk velocity in x-direction</span>
<span class="na">VY0                                                                        Initial bulk velocity in y-direction</span>
<span class="na">VZ0                                                                        Initial bulk velocity in z-direction</span>
<span class="na">taperInnerRadius                                                           Inner radius of the zone with a density tapering from the ionospheric value tothe background (m)</span>
<span class="na">taperOuterRadius                                                           Outer radius of the zone with a density tapering from the ionospheric value tothe background (m)</span>

<span class="k">[MultiPeak]</span>
<span class="na">Bx                                                                         Magnetic field x component (T)</span>
<span class="na">By                                                                         Magnetic field y component (T)</span>
<span class="na">Bz                                                                         Magnetic field z component (T)</span>
<span class="na">dBx                                                                        Magnetic field x component cosine perturbation amplitude (T)</span>
<span class="na">dBy                                                                        Magnetic field y component cosine perturbation amplitude (T)</span>
<span class="na">dBz                                                                        Magnetic field z component cosine perturbation amplitude (T)</span>
<span class="na">magXPertAbsAmp                                                             Absolute amplitude of the random magnetic perturbation along x (T)</span>
<span class="na">magYPertAbsAmp                                                             Absolute amplitude of the random magnetic perturbation along y (T)</span>
<span class="na">magZPertAbsAmp                                                             Absolute amplitude of the random magnetic perturbation along z (T)</span>
<span class="na">lambda                                                                     B cosine perturbation wavelength (m)</span>
<span class="na">densityModel                                                               Which spatial density model is used?</span>

<span class="k">[&lt;population&gt;_MultiPeak]</span>
<span class="na">n                                                                          Number of peaks to create</span>
<span class="na">rho                                                                        Number density (m^-3)</span>
<span class="na">Tx                                                                         Temperature (K)</span>
<span class="na">Ty                                                                         Temperature</span>
<span class="na">Tz                                                                         Temperature</span>
<span class="na">Vx                                                                         Bulk velocity x component (m/s)</span>
<span class="na">Vy                                                                         Bulk velocity y component (m/s)</span>
<span class="na">Vz                                                                         Bulk velocity z component (m/s)</span>
<span class="na">rhoPertAbsAmp                                                              Absolute amplitude of the density perturbation</span>

<span class="k">[VelocityBox]</span>
<span class="na">rho                                                                        Number density in full 6 dimensions (m^-6 s^3)</span>
<span class="na">Vx1                                                                        Box min x (m/s)</span>
<span class="na">Vx2                                                                        Box max x (m/s)</span>
<span class="na">Vy1                                                                        Box min y (m/s)</span>
<span class="na">Vy2                                                                        Box max y (m/s)</span>
<span class="na">Vz1                                                                        Box min z (m/s)</span>
<span class="na">Vz2                                                                        Box max z (m/s)</span>
<span class="na">Bx                                                                         Magnetic field x component (T)</span>
<span class="na">By                                                                         Magnetic field y component (T)</span>
<span class="na">Bz                                                                         Magnetic field z component (T)</span>

<span class="k">[Riemann]</span>
<span class="na">rho1                                                                       Number density, left state (m^-3)</span>
<span class="na">rho2                                                                       Number density, right state (m^-3)</span>
<span class="na">T1                                                                         Temperature, left state (K)</span>
<span class="na">T2                                                                         Temperature, right state (K)</span>
<span class="na">Vx1                                                                        Bulk velocity x component, left state (m/s)</span>
<span class="na">Vx2                                                                        Bulk velocity x component, right state (m/s)</span>
<span class="na">Vy1                                                                        Bulk velocity y component, left state (m/s)</span>
<span class="na">Vy2                                                                        Bulk velocity y component, right state (m/s)</span>
<span class="na">Vz1                                                                        Bulk velocity z component, left state (m/s)</span>
<span class="na">Vz2                                                                        Bulk velocity z component, right state (m/s)</span>
<span class="na">Bx1                                                                        Magnetic field x component, left state (T)</span>
<span class="na">Bx2                                                                        Magnetic field x component, right state (T)</span>
<span class="na">By1                                                                        Magnetic field y component, left state (T)</span>
<span class="na">By2                                                                        Magnetic field y component, right state (T)</span>
<span class="na">Bz1                                                                        Magnetic field z component, left state (T)</span>
<span class="na">Bz2                                                                        Magnetic field z component, right state (T)</span>

<span class="k">[Shock]</span>
<span class="na">BX0                                                                        Background field value (T)</span>
<span class="na">BY0                                                                        Background field value (T)</span>
<span class="na">BZ0                                                                        Background field value (T)</span>
<span class="na">EX0                                                                        Background electric field</span>
<span class="na">VX0                                                                        Bulk velocity in x</span>
<span class="na">VY0                                                                        Bulk velocity in y</span>
<span class="na">VZ0                                                                        Bulk velocuty in z</span>
<span class="na">rho                                                                        Number density (m^-3)</span>
<span class="na">Temperature                                                                Temperature (K)</span>
<span class="na">magPertAmp                                                                 Amplitude of the magnetic perturbation</span>
<span class="na">densityPertAmp                                                             Amplitude factor of the density perturbation</span>
<span class="na">velocityPertAmp                                                            Amplitude of the velocity perturbation</span>
<span class="na">maxwCutoff                                                                 Cutoff for the maxwellian distribution</span>
<span class="na">Scale_x                                                                    Scale length in x (m)</span>
<span class="na">Scale_y                                                                    Scale length in y (m)</span>
<span class="na">Sharp_Y                                                                    Sharpness of tannh</span>

<span class="k">[IPShock]</span>
<span class="na">BX0u                                                                       Upstream mag. field value (T)</span>
<span class="na">BY0u                                                                       Upstream mag. field value (T)</span>
<span class="na">BZ0u                                                                       Upstream mag. field value (T)</span>
<span class="na">BX0d                                                                       Downstream mag. field value (T)</span>
<span class="na">BY0d                                                                       Downstream mag. field value (T)</span>
<span class="na">BZ0d                                                                       Downstream mag. field value (T)</span>
<span class="na">Width                                                                      Shock Width (m)</span>
<span class="na">AMR_L1width                                                                L1 AMR region width (m)</span>
<span class="na">AMR_L2width                                                                L2 AMR region width (m)</span>
<span class="na">AMR_L3width                                                                L3 AMR region width (m)</span>
<span class="na">AMR_L4width                                                                L4 AMR region width (m)</span>

<span class="k">[&lt;population&gt;_IPShock]</span>
<span class="na">VX0u                                                                       Upstream Bulk velocity in x</span>
<span class="na">VY0u                                                                       Upstream Bulk velocity in y</span>
<span class="na">VZ0u                                                                       Upstream Bulk velocuty in z</span>
<span class="na">rhou                                                                       Upstream Number density (m^-3)</span>
<span class="na">Temperatureu                                                               Upstream Temperature (K)</span>
<span class="na">VX0d                                                                       Downstream Bulk velocity in x</span>
<span class="na">VY0d                                                                       Downstream Bulk velocity in y</span>
<span class="na">VZ0d                                                                       Downstream Bulk velocuty in z</span>
<span class="na">rhod                                                                       Downstream Number density (m^-3)</span>
<span class="na">Temperatured                                                               Downstream Temperature (K)</span>
<span class="na">maxwCutoff                                                                 Cutoff for the maxwellian distribution</span>

<span class="k">[Template]</span>
<span class="na">param                                                                      This is my project&#39;s parameter. Default is 0.0</span>

<span class="k">[test_fp]</span>
<span class="na">V0                                                                         Velocity magnitude (m/s)</span>
<span class="na">B0                                                                         Magnetic field value in the non-zero patch (T)</span>
<span class="na">rho                                                                        Number density (m^-3)</span>
<span class="na">Temperature                                                                Temperature (K)</span>
<span class="na">angle                                                                      Orientation of the propagation expressed in pi/4</span>
<span class="na">Bdirection                                                                 Direction of the magnetic field (0</span><span class="o">:</span><span class="s">x, 1:y, 2:z, 3:all)</span>
<span class="na">shear                                                                      Add a shear (if false, V</span><span class="o">=</span><span class="s">0.5 everywhere).</span>

<span class="k">[TestHall]</span>
<span class="na">BX0                                                                        Magnetic field x (T)</span>
<span class="na">BY0                                                                        Magnetic field y (T)</span>
<span class="na">BZ0                                                                        Magnetic field z (T)</span>
<span class="na">VX0                                                                        velocity x (m/s)</span>
<span class="na">VY0                                                                        velocity y (m/s)</span>
<span class="na">VZ0                                                                        velocity z (m/s)</span>
<span class="na">Temperature                                                                Temperature (K)</span>
<span class="na">rho                                                                        Number density (m^-3)</span>

<span class="k">[test_trans]</span>
<span class="na">cellPosition                                                               Position of the centre of the cells initiated (same used in velocity and space).</span>
<span class="na">peakValue                                                                  Value of the distribution function</span>

<span class="k">[VerificationLarmor]</span>
<span class="na">BX0                                                                        Background field value (T)</span>
<span class="na">BY0                                                                        Background field value (T)</span>
<span class="na">BZ0                                                                        Background field value (T)</span>
<span class="na">VX0                                                                        Bulk velocity in x</span>
<span class="na">VY0                                                                        Bulk velocity in y</span>
<span class="na">VZ0                                                                        Bulk velocity in z</span>
<span class="na">X0                                                                         Initial Position</span>
<span class="na">Y0                                                                         Initial Position</span>
<span class="na">Z0                                                                         Initial Position</span>
<span class="na">rho                                                                        Number density (m^-3)</span>

<span class="k">[Shocktest]</span>
<span class="na">rho1                                                                       Number density, left state (m^-3)</span>
<span class="na">rho2                                                                       Number density, right state (m^-3)</span>
<span class="na">T1                                                                         Temperature, left state (K)</span>
<span class="na">T2                                                                         Temperature, right state (K)</span>
<span class="na">Vx1                                                                        Bulk velocity x component, left state (m/s)</span>
<span class="na">Vx2                                                                        Bulk velocity x component, right state (m/s)</span>
<span class="na">Vy1                                                                        Bulk velocity y component, left state (m/s)</span>
<span class="na">Vy2                                                                        Bulk velocity y component, right state (m/s)</span>
<span class="na">Vz1                                                                        Bulk velocity z component, left state (m/s)</span>
<span class="na">Vz2                                                                        Bulk velocity z component, right state (m/s)</span>
<span class="na">Bx1                                                                        Magnetic field x component, left state (T)</span>
<span class="na">Bx2                                                                        Magnetic field x component, right state (T)</span>
<span class="na">By1                                                                        Magnetic field y component, left state (T)</span>
<span class="na">By2                                                                        Magnetic field y component, right state (T)</span>
<span class="na">Bz1                                                                        Magnetic field z component, left state (T)</span>
<span class="na">Bz2                                                                        Magnetic field z component, right state (T)</span>

<span class="k">[Project_common]</span>
<span class="na">seed                                                                       Seed for the RNG</span>
</pre></div>
</div>
</section>
</section>
<span id="document-analysator_supported"></span><section id="analysator-supported-data-reducers-and-vlasiator-variables">
<h2>Analysator supported data reducers and vlasiator variables<a class="headerlink" href="#analysator-supported-data-reducers-and-vlasiator-variables" title="Link to this heading"></a></h2>
<p>This documents is intended as a helpful but incomplete reference for vlasiator
outputs and post-processing data reducers supported by Analysator</p>
<section id="reading-variables-from-vlsv-files">
<h3>Reading variables from .vlsv files<a class="headerlink" href="#reading-variables-from-vlsv-files" title="Link to this heading"></a></h3>
<p>Analysator supports reading multiple different types of variables:</p>
<ul class="simple">
<li><p>MPIgrid variables</p></li>
<li><p>FSgrid variables</p></li>
<li><p>variables directly available in the data files</p></li>
<li><p>variables generated via receipes from the data file variables (named, somewhat incorrectly, datareducers)</p></li>
</ul>
</section>
<section id="vlasiator-variable-naming-scheme">
<h3>Vlasiator variable naming scheme<a class="headerlink" href="#vlasiator-variable-naming-scheme" title="Link to this heading"></a></h3>
<p>In Vlasiator versions before V5, variables were always saved on the DCCRG/MPI/Vlasov grid,
and as such, only a single naming scheme was required. With the introduction of Vlasiator V5,
variables could also be saved on the FSgrid. When the ionosphere grid was be implemented,
that can also be saved to on its own grid. To differentiate between the different grids, a
naming scheme was introduced where a prefix defines the grid type:</p>
<ul class="simple">
<li><p>No prefix: old pre-V5 data</p></li>
<li><p>prefix <code class="docutils literal notranslate"><span class="pre">vg_</span></code>: Vlasov/MPI/DCCRG grid</p></li>
<li><p>prefix <code class="docutils literal notranslate"><span class="pre">fg_</span></code>: fieldsolver/FSgrid</p></li>
<li><p>prefix <code class="docutils literal notranslate"><span class="pre">ig_</span></code>: ionosphere grid</p></li>
</ul>
</section>
<section id="multipop-naming-version-4-onwards">
<h3>Multipop naming (version 4 onwards)<a class="headerlink" href="#multipop-naming-version-4-onwards" title="Link to this heading"></a></h3>
<p>Some variables are directly connected to a given particle species, with the variable name having a
particle species prefix. For example: <code class="docutils literal notranslate"><span class="pre">proton/rho</span></code> or <code class="docutils literal notranslate"><span class="pre">proton/vg_rho</span></code> for V4 and V5 outputs
respectively. Note that multipop outputs are always on the Vlasov grid, but still have the <code class="docutils literal notranslate"><span class="pre">vg_</span></code>
prefix after the particle species name. Some variable such as <code class="docutils literal notranslate"><span class="pre">rho</span></code> and <code class="docutils literal notranslate"><span class="pre">blocks</span></code> which are
directly connected to particle species are availabe without the population identifier in datafiles
from version 3 and earlier, assuming the species to be protons.</p>
</section>
<section id="vlasiator-variable-correspondences">
<h3>Vlasiator variable correspondences<a class="headerlink" href="#vlasiator-variable-correspondences" title="Link to this heading"></a></h3>
<p>The below tables list variable names used in different versions of Vlasiator proper (updated: 3.6.2021).
A blank entry in a corresponding column indicates this variable is not available.
Some debugging variables such as field derivatives have been omitted from the table.</p>
<table class="docutils align-default" id="id1">
<caption><span class="caption-text">FSgrid to pre-V5</span><a class="headerlink" href="#id1" title="Link to this table"></a></caption>
<colgroup>
<col style="width: 30.0%" />
<col style="width: 70.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>FSgrid variables</p></th>
<th class="head"><p>Corresponding Vlasov grid variable in older versions</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>fg_b</p></td>
<td><p>B</p></td>
</tr>
<tr class="row-odd"><td><p>fg_b_background</p></td>
<td><p>background_B</p></td>
</tr>
<tr class="row-even"><td><p>fg_b_perturbed</p></td>
<td><p>perturbed_B</p></td>
</tr>
<tr class="row-odd"><td><p>fg_b_vol</p></td>
<td><p>B_vol</p></td>
</tr>
<tr class="row-even"><td><p>fg_e</p></td>
<td><p>E</p></td>
</tr>
<tr class="row-odd"><td><p>fg_e_hall_??</p></td>
<td><p>EXHALL_???_???</p></td>
</tr>
<tr class="row-even"><td><p>fg_e_vol</p></td>
<td><p>E_vol</p></td>
</tr>
<tr class="row-odd"><td><p>fg_rhom</p></td>
<td><p>Rhom</p></td>
</tr>
<tr class="row-even"><td><p>fg_rhoq</p></td>
<td><p>Rhoq</p></td>
</tr>
<tr class="row-odd"><td><p>fg_v</p></td>
<td><p>V</p></td>
</tr>
<tr class="row-even"><td><p>fg_pressure</p></td>
<td><p>Pressure</p></td>
</tr>
<tr class="row-odd"><td><p>fg_maxdt_fieldsolver</p></td>
<td><p>max_fields_dt</p></td>
</tr>
<tr class="row-even"><td><p>fg_rank</p></td>
<td><p>FSgrid_rank</p></td>
</tr>
<tr class="row-odd"><td><p>fg_amr_level</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>fg_boundarytype</p></td>
<td><p>FSgrid_boundaryType</p></td>
</tr>
<tr class="row-odd"><td><p>fg_boundarylayer</p></td>
<td><p>Boundary_layerx</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default" id="id2">
<caption><span class="caption-text">V5 Vlasov grid to pre-V5</span><a class="headerlink" href="#id2" title="Link to this table"></a></caption>
<colgroup>
<col style="width: 30.0%" />
<col style="width: 70.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>V5 Vlasov grid variables</p></th>
<th class="head"><p>Older Vlasov grid variables</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>vg_rhom</p></td>
<td><p>Rhom</p></td>
</tr>
<tr class="row-odd"><td><p>vg_rhoq</p></td>
<td><p>Rhoq</p></td>
</tr>
<tr class="row-even"><td><p>vg_v</p></td>
<td><p>V</p></td>
</tr>
<tr class="row-odd"><td><p>vg_pressure</p></td>
<td><p>Pressure</p></td>
</tr>
<tr class="row-even"><td></td>
<td><p>Blocks</p></td>
</tr>
<tr class="row-odd"><td><p>vg_b_vol</p></td>
<td><p>B_vol</p></td>
</tr>
<tr class="row-even"><td><p>vg_b_background_vol</p></td>
<td><p>BGB_vol</p></td>
</tr>
<tr class="row-odd"><td><p>vg_b_perturbed_vol</p></td>
<td><p>PERB_vol</p></td>
</tr>
<tr class="row-even"><td></td>
<td><p>E</p></td>
</tr>
<tr class="row-odd"><td><p>vg_e_vol</p></td>
<td><p>E_vol</p></td>
</tr>
<tr class="row-even"><td><p>vg_e_gradpe</p></td>
<td><p>EGRADPE</p></td>
</tr>
<tr class="row-odd"><td><p>vg_f_saved</p></td>
<td><p>fSaved</p></td>
</tr>
<tr class="row-even"><td><p>vg_maxdt_acceleration</p></td>
<td><p>max_v_dt</p></td>
</tr>
<tr class="row-odd"><td><p>vg_maxdt_translation</p></td>
<td><p>max_r_dt</p></td>
</tr>
<tr class="row-even"><td><p>vg_rank</p></td>
<td><p>MPI_rank</p></td>
</tr>
<tr class="row-odd"><td><p>vg_loadbalance_weight</p></td>
<td><p>LB_weight</p></td>
</tr>
<tr class="row-even"><td><p>vg_boundarytype</p></td>
<td><p>Boundary_type</p></td>
</tr>
<tr class="row-odd"><td><p>vg_boundarylayer</p></td>
<td><p>Boundary_layer</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default" id="id3">
<caption><span class="caption-text">Multipop correspondence</span><a class="headerlink" href="#id3" title="Link to this table"></a></caption>
<colgroup>
<col style="width: 30.0%" />
<col style="width: 60.0%" />
<col style="width: 10.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Per-population variables</p></th>
<th class="head"><p>Older per-population variables</p></th>
<th class="head"><p>Older pre-multipop variables</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>populations/vg_rho</p></td>
<td><p>populations/rho</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>populations/vg_v</p></td>
<td><p>populations/V</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>populations/vg_rho_thermal</p></td>
<td><p>populations/RhoNonBackstream</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>populations/vg_rho_nonthermal</p></td>
<td><p>populations/RhoBackstream</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>populations/vg_v_thermal</p></td>
<td><p>populations/VNonBackstream</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>populations/vg_v_nonthermal</p></td>
<td><p>populations/VBackstream</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>populations/vg_effectivesparsitythreshold</p></td>
<td><p>populations/EffectiveSparsityThreshold</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>populations/vg_rho_loss_adjust</p></td>
<td><p>populations/rho_loss_adjust</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>populations/vg_energydensity</p></td>
<td><p>populations/EnergyDensity</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>populations/vg_precipitationdifferentialflux</p></td>
<td><p>populations/PrecipitationDiffFlux</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>populations/vg_maxdt_acceleration</p></td>
<td><p>populations/MaxVdt</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>populations/vg_maxdt_translation</p></td>
<td><p>populations/MaxRdt</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>populations/vg_blocks</p></td>
<td><p>populations/Blocks</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>populations/vg_acceleration_subcycles</p></td>
<td><p>populations/acc_subcycles</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>populations/vg_ptensor_diagonal</p></td>
<td><p>populations/PTensorDiagonal</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>populations/vg_ptensor_offdiagonal</p></td>
<td><p>populations/PTensorOffDiagonal</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>populations/vg_ptensor_nonthermal_diagonal</p></td>
<td><p>populations/PTensorBackstreamDiagonal</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>populations/vg_ptensor_nonthermal_offdiagonal</p></td>
<td><p>populations/PTensorBackstreamOffDiagonal</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>populations/vg_ptensor_thermal_diagonal</p></td>
<td><p>populations/PTensorNonBackstreamDiagonal</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>populations/vg_ptensor_thermal_offdiagonal</p></td>
<td><p>populations/PTensorNonBackstreamOffDiagonal</p></td>
<td></td>
</tr>
</tbody>
</table>
</section>
<section id="analysator-datareducers">
<h3>Analysator datareducers<a class="headerlink" href="#analysator-datareducers" title="Link to this heading"></a></h3>
<p>Post-processed derived variables (called somewhat mistakenly datareducers) are available in analysator.
One limitation is that in current versions of analysator (as of March 2022) all datareducers require
Vlasov grid variables as they work with CellID lists. Also, datareducers do not support spatial or
temporal derivatives. However, both use of FSgrid variables and spatial and temporal derivatives in
plotting routines is possible via expressions and external functions as called by plot_colormap
and other such plotting routines (see also plot_helpers.py).</p>
<p>The up-to-date list of datareducers can always be found in pyVlsv/reduction.py</p>
</section>
<section id="list-of-datareducers-for-vlasiator-versions-1-4">
<h3>List of datareducers for Vlasiator versions 1…4<a class="headerlink" href="#list-of-datareducers-for-vlasiator-versions-1-4" title="Link to this heading"></a></h3>
<p>Note: If several populations exist for a v4 multipop run, temperature reducers are incorrect.
If the value is available directly instead of as a datareducer, that name is listed instead in
parathentheses. If it is available through multipop datareducers, it is shown with “mpop”.</p>
<table class="docutils align-default" id="id4">
<caption><span class="caption-text">Generic datareducers</span><a class="headerlink" href="#id4" title="Link to this table"></a></caption>
<colgroup>
<col style="width: 30.0%" />
<col style="width: 60.0%" />
<col style="width: 10.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>v5 datareducer</p></th>
<th class="head"><p>v1…4 datareducer</p></th>
<th class="head"><p>explanation</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>(vg_v)</p></td>
<td><p>v</p></td>
<td><p>velocity</p></td>
</tr>
<tr class="row-odd"><td><p>vg_vms</p></td>
<td><p>vms</p></td>
<td><p>magnetosonic speed</p></td>
</tr>
<tr class="row-even"><td><p>vg_vs</p></td>
<td><p>vs</p></td>
<td><p>sound speed</p></td>
</tr>
<tr class="row-odd"><td><p>vg_va</p></td>
<td><p>va</p></td>
<td><p>Alfvén speed</p></td>
</tr>
<tr class="row-even"><td><p>vg_ma</p></td>
<td><p>ma</p></td>
<td><p>Alfvén Mach number</p></td>
</tr>
<tr class="row-odd"><td><p>vg_mms</p></td>
<td><p>mms</p></td>
<td><p>magnetosonic Mach number</p></td>
</tr>
<tr class="row-even"><td><p>vg_di</p></td>
<td><p>di</p></td>
<td><p>Ion inertial length</p></td>
</tr>
<tr class="row-odd"><td><p>vg_v_parallel</p></td>
<td><p>vparallel</p></td>
<td><p>V parallel to B</p></td>
</tr>
<tr class="row-even"><td><p>vg_v_perpendicular</p></td>
<td><p>vperpendicular</p></td>
<td><p>V perpendicular to B</p></td>
</tr>
<tr class="row-odd"><td><p>mpop</p></td>
<td><p>vparallelbackstream</p></td>
<td><p>non-thermal proton V parallel to B</p></td>
</tr>
<tr class="row-even"><td><p>mpop</p></td>
<td><p>vperpendicularbackstream</p></td>
<td><p>non-thermal proton V perpendicular to B</p></td>
</tr>
<tr class="row-odd"><td><p>mpop</p></td>
<td><p>vparallelnonbackstream</p></td>
<td><p>thermal proton V parallel to B</p></td>
</tr>
<tr class="row-even"><td><p>mpop</p></td>
<td><p>vperpendicularnonbackstream</p></td>
<td><p>thermal proton V perpendicular to B</p></td>
</tr>
<tr class="row-odd"><td><p>vg_e_parallel</p></td>
<td><p>eparallel</p></td>
<td><p>E component parallel to B</p></td>
</tr>
<tr class="row-even"><td><p>vg_e_perpendicular</p></td>
<td><p>eperpendicular</p></td>
<td><p>E component perpendicular to B</p></td>
</tr>
<tr class="row-odd"><td><p>vg_pdyn</p></td>
<td><p>pdyn</p></td>
<td><p>dynamic pressure</p></td>
</tr>
<tr class="row-even"><td><p>vg_pdynx</p></td>
<td><p>pdynx</p></td>
<td><p>dynamic pressure from only x-directional velocity</p></td>
</tr>
<tr class="row-odd"><td><p>vg_poynting</p></td>
<td><p>poynting</p></td>
<td><p>Poynting vector</p></td>
</tr>
<tr class="row-even"><td><p>n/a</p></td>
<td><p>firstadiabatic</p></td>
<td><p>mean first adiabatic invariant</p></td>
</tr>
<tr class="row-odd"><td><p>vg_egradpe_parallel</p></td>
<td><p>n/a</p></td>
<td><p>Electric field electron pressure gradient term parallel to B</p></td>
</tr>
<tr class="row-even"><td><p>vg_egradpe_perpendicular</p></td>
<td><p>n/a</p></td>
<td><p>Electric field electron pressure gradient term perpendicular to B</p></td>
</tr>
<tr class="row-odd"><td><p>vg_restart_v</p></td>
<td><p>restart_v</p></td>
<td><p>estimate V from a restart file</p></td>
</tr>
<tr class="row-even"><td><p>vg_restart_rho</p></td>
<td><p>restart_rho</p></td>
<td><p>estimate particle number density from a restart file</p></td>
</tr>
<tr class="row-odd"><td><p>vg_restart_rhom</p></td>
<td><p>restart_rhom</p></td>
<td><p>estimate mass density from a restart file</p></td>
</tr>
<tr class="row-even"><td><p>vg_restart_rhoq</p></td>
<td><p>restart_rhoq</p></td>
<td><p>estimate charge density from a restart file</p></td>
</tr>
</tbody>
</table>
</section>
<section id="list-of-per-population-datareducers-for-vlasiator-versions-4-and-5">
<h3>List of per-population datareducers for Vlasiator versions 4 and 5.<a class="headerlink" href="#list-of-per-population-datareducers-for-vlasiator-versions-4-and-5" title="Link to this heading"></a></h3>
<p>For these multipop datareducers, replace <code class="docutils literal notranslate"><span class="pre">pop</span></code> with required population name. If <code class="docutils literal notranslate"><span class="pre">pop/</span></code> is
omitted, a sum over per-population values is provided.</p>
<table class="docutils align-default" id="id5">
<caption><span class="caption-text">Multipop datareduces</span><a class="headerlink" href="#id5" title="Link to this table"></a></caption>
<colgroup>
<col style="width: 40.0%" />
<col style="width: 30.0%" />
<col style="width: 30.0%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>v5 datareducer</p></th>
<th class="head"><p>v4 datareducer</p></th>
<th class="head"><p>explanation</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>pop/vg_rhom</p></td>
<td><p>pop/rhom</p></td>
<td><p>mass density</p></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_rhoq</p></td>
<td><p>pop/rhoq</p></td>
<td><p>charge density</p></td>
</tr>
<tr class="row-even"><td><p>pop/vg_pdyn</p></td>
<td><p>pop/pdyn</p></td>
<td><p>dynamic pressure</p></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_pdynx</p></td>
<td><p>pop/pdynx</p></td>
<td><p>dynamic pressure from only x-directional velocity</p></td>
</tr>
<tr class="row-even"><td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_v_parallel</p></td>
<td><p>pop/vparallel</p></td>
<td><p>V parallel to B</p></td>
</tr>
<tr class="row-even"><td><p>pop/vg_v_perpendicular</p></td>
<td><p>pop/vperpendicular</p></td>
<td><p>V perpendicular to B</p></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_v_parallel_nonthermal</p></td>
<td><p>pop/vparallelbackstream</p></td>
<td><p>non-thermal V parallel to B</p></td>
</tr>
<tr class="row-even"><td><p>pop/vg_v_perpendicular_nonthermal</p></td>
<td><p>pop/vperpendicularbackstream</p></td>
<td><p>non-thermal V perpendicular to B</p></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_v_parallel_thermal</p></td>
<td><p>pop/vparallelnonbackstream</p></td>
<td><p>thermal V parallel to B</p></td>
</tr>
<tr class="row-even"><td><p>pop/vg_v_perpendicular_thermal</p></td>
<td><p>pop/vperpendicularnonbackstream</p></td>
<td><p>thermal V perpendicular to B</p></td>
</tr>
<tr class="row-odd"><td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>pop/vg_thermalvelocity</p></td>
<td><p>pop/thermalvelocity</p></td>
<td><p>estimate of thermal velocity</p></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_larmor</p></td>
<td><p>pop/larmor</p></td>
<td><p>estimate of representative Larmor radius</p></td>
</tr>
<tr class="row-even"><td><p>pop/vg_firstadiabatic</p></td>
<td><p>pop/firstadiabatic</p></td>
<td><p>estimate of first adiabatic invariant (old!)</p></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_gyroperiod</p></td>
<td><p>pop/gyroperiod</p></td>
<td><p>gyroperiod</p></td>
</tr>
<tr class="row-even"><td><p>pop/vg_plasmaperiod</p></td>
<td><p>pop/plasmaperiod</p></td>
<td><p>plasma period</p></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_precipitationintegralenergyflux</p></td>
<td><p>n/a</p></td>
<td><p>Integral energy flux of precipitating particles estimate (several energy channels)</p></td>
</tr>
<tr class="row-even"><td><p>pop/vg_precipitationmeanenergy</p></td>
<td><p>n/a</p></td>
<td><p>Mean energy of channel for precipitating particle fluxes</p></td>
</tr>
<tr class="row-odd"><td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>pop/vg_pressure</p></td>
<td><p>pop/pressure</p></td>
<td><p>pressure</p></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_ptensor</p></td>
<td><p>pop/ptensor</p></td>
<td><p>pressure tensor</p></td>
</tr>
<tr class="row-even"><td><p>pop/vg_ptensor_rotated</p></td>
<td><p>pop/ptensorrotated</p></td>
<td><p>pressure tensor rotated to align with B</p></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_p_parallel</p></td>
<td><p>pop/pparallel</p></td>
<td><p>pressure parallel to B</p></td>
</tr>
<tr class="row-even"><td><p>pop/vg_p_perpendicular</p></td>
<td><p>pop/pperpendicular</p></td>
<td><p>pressure perpendicular to B</p></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_p_anisotropy</p></td>
<td><p>pop/pperpoverpar</p></td>
<td><p>ratio of perpendicular pressure to parallel pressure</p></td>
</tr>
<tr class="row-even"><td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_p_nonthermal</p></td>
<td><p>pop/pbackstream</p></td>
<td><p>pressure due to non-thermal particles</p></td>
</tr>
<tr class="row-even"><td><p>pop/vg_ptensor_nonthermal</p></td>
<td><p>pop/ptensorbackstream</p></td>
<td><p>pressure tensor due to non-thermal particles</p></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_ptensor_rotated_nonthermal</p></td>
<td><p>pop/ptensorrotatedbackstream</p></td>
<td><p>pressure tensor due to non-thermal particles rotated to align with B</p></td>
</tr>
<tr class="row-even"><td><p>pop/vg_p_parallel_nonthermal</p></td>
<td><p>pop/pparallelbackstream</p></td>
<td><p>pressure component due to non-thermal particles parallel to B</p></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_p_perpendicular_nonthermal</p></td>
<td><p>pop/pperpendicularbackstream</p></td>
<td><p>pressure component due to non-thermal particles perpendicular to B</p></td>
</tr>
<tr class="row-even"><td><p>pop/vg_p_anisotropy_nonthermal</p></td>
<td><p>pop/pperpoverparbackstream</p></td>
<td><p>ratio of perpendicular pressure to parallel pressure for non-thermal particles</p></td>
</tr>
<tr class="row-odd"><td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>pop/vg_p_thermal</p></td>
<td><p>pop/pnonbackstream</p></td>
<td><p>pressure due to thermal particles</p></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_ptensor_thermal</p></td>
<td><p>pop/ptensornonbackstream</p></td>
<td><p>pressure tensor due to thermal particles</p></td>
</tr>
<tr class="row-even"><td><p>pop/vg_ptensor_rotated_thermal</p></td>
<td><p>pop/ptensorrotatednonbackstream</p></td>
<td><p>pressure tensor due to thermal particles rotated to align with B</p></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_p_parallel_thermal</p></td>
<td><p>pop/pparallelnonbackstream</p></td>
<td><p>pressure component due to thermal particles parallel to B</p></td>
</tr>
<tr class="row-even"><td><p>pop/vg_p_perpendicular_thermal</p></td>
<td><p>pop/pperpendicularnonbackstream</p></td>
<td><p>pressure component due to thermal particles perpendicular to B</p></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_p_anisotropy_thermal</p></td>
<td><p>pop/pperpoverparnonbackstream</p></td>
<td><p>ratio of perpendicular pressure to parallel pressure for thermal particles</p></td>
</tr>
<tr class="row-even"><td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_temperature</p></td>
<td><p>pop/temperature</p></td>
<td><p>temperature</p></td>
</tr>
<tr class="row-even"><td><p>pop/vg_ttensor</p></td>
<td><p>pop/ttensor</p></td>
<td><p>temperature tensor</p></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_ttensor_rotated</p></td>
<td><p>pop/ttensorrotated</p></td>
<td><p>temperature tensor rotated to align with B</p></td>
</tr>
<tr class="row-even"><td><p>pop/vg_t_parallel</p></td>
<td><p>pop/tparallel</p></td>
<td><p>temperature parallel to B</p></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_t_perpendicular</p></td>
<td><p>pop/tperpendicular</p></td>
<td><p>temperature perpendicular to B</p></td>
</tr>
<tr class="row-even"><td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_t_nonthermal</p></td>
<td><p>pop/tbackstream</p></td>
<td><p>temperature of non-thermal particles</p></td>
</tr>
<tr class="row-even"><td><p>pop/vg_ttensor_nonthermal</p></td>
<td><p>pop/ttensorbackstream</p></td>
<td><p>temperature tensor of non-thermal particles</p></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_ttensor_rotated_nonthermal</p></td>
<td><p>pop/ttensorrotatedbackstream</p></td>
<td><p>temperature tensor of non-thermal particles rotated to align with B</p></td>
</tr>
<tr class="row-even"><td><p>pop/vg_t_parallel_nonthermal</p></td>
<td><p>pop/tparallelbackstream</p></td>
<td><p>temperature component of non-thermal particles parallel to B</p></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_t_perpendicular_nonthermal</p></td>
<td><p>pop/tperpendicularbackstream</p></td>
<td><p>temperature component of non-thermal particles perpendicular to B</p></td>
</tr>
<tr class="row-even"><td><p>pop/vg_t_thermal</p></td>
<td><p>pop/tnonbackstream</p></td>
<td><p>temperature of thermal particles</p></td>
</tr>
<tr class="row-odd"><td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>pop/vg_ttensor_thermal</p></td>
<td><p>pop/ttensornonbackstream</p></td>
<td><p>temperature tensor of thermal particles</p></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_ttensor_rotated_thermal</p></td>
<td><p>pop/ttensorrotatednonbackstream</p></td>
<td><p>temperature temperature tensor of thermal particles rotated to align with B</p></td>
</tr>
<tr class="row-even"><td><p>pop/vg_t_parallel_thermal</p></td>
<td><p>pop/tparallelnonbackstream</p></td>
<td><p>temperature component of thermal particles parallel to B</p></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_t_perpendicular_thermal</p></td>
<td><p>pop/tperpendicularnonbackstream</p></td>
<td><p>temperature component of thermal particles perpendicular to B</p></td>
</tr>
<tr class="row-even"><td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_t_anisotropy</p></td>
<td><p>pop/tperpoverpar</p></td>
<td><p>ratio of perpendicular temperature to parallel temperature</p></td>
</tr>
<tr class="row-even"><td><p>pop/vg_t_anisotropy_nonthermal</p></td>
<td><p>pop/tperpoverparbackstream</p></td>
<td><p>ratio of perpendicular temperature to parallel temperature for non-thermal particles</p></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_t_anisotropy_thermal</p></td>
<td><p>pop/tperpoverparnonbackstream</p></td>
<td><p>ratio of perpendicular temperature to parallel temperature for thermal particles</p></td>
</tr>
<tr class="row-even"><td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_agyrotropy</p></td>
<td><p>pop/agyrotropy</p></td>
<td><p>agyrotropy (according to Swisdak 2016)</p></td>
</tr>
<tr class="row-even"><td><p>pop/vg_agyrotropy_nonthermal</p></td>
<td><p>pop/agyrotropybackstream</p></td>
<td><p>agyrotropy for non-thermal particles</p></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_agyrotropy_thermal</p></td>
<td><p>pop/agyrotropynonbackstream</p></td>
<td><p>agyrotropy for thermal particles</p></td>
</tr>
<tr class="row-even"><td></td>
<td></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_beta</p></td>
<td><p>pop/beta</p></td>
<td><p>plasma beta</p></td>
</tr>
<tr class="row-even"><td><p>pop/vg_beta_parallel</p></td>
<td><p>pop/betaparallel</p></td>
<td><p>plasma beta component parallel to B</p></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_beta_perpendicular</p></td>
<td><p>pop/betaperpendicular</p></td>
<td><p>plasma beta component perpendicular to B</p></td>
</tr>
<tr class="row-even"><td><p>pop/vg_beta_anisotropy</p></td>
<td><p>pop/betaperpoverpar</p></td>
<td><p>ratio of perpendicular beta to parallel beta</p></td>
</tr>
<tr class="row-odd"><td><p>pop/vg_beta_anisotropy_nonthermal</p></td>
<td><p>pop/betaperpoverparbackstream</p></td>
<td><p>ratio of perpendicular beta to parallel beta for non-thermal particles</p></td>
</tr>
<tr class="row-even"><td><p>pop/vg_beta_anisotropy_thermal</p></td>
<td><p>pop/betaperpoverparnonbackstream</p></td>
<td><p>ratio of perpendicular beta to parallel beta for thermal particles</p></td>
</tr>
<tr class="row-odd"><td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</section>
</section>
<span id="document-new_project"></span><section id="new-project-from-template">
<h2>New project from template<a class="headerlink" href="#new-project-from-template" title="Link to this heading"></a></h2>
<section id="timing">
<h3>Timing<a class="headerlink" href="#timing" title="Link to this heading"></a></h3>
<p>Thursday, maybe.</p>
</section>
<section id="setting-up-a-new-project">
<h3>Setting up a new project<a class="headerlink" href="#setting-up-a-new-project" title="Link to this heading"></a></h3>
<section id="set-up-a-project-folder">
<h4>Set up a project folder<a class="headerlink" href="#set-up-a-project-folder" title="Link to this heading"></a></h4>
<p>First you need to create a subdirectory in the <code class="docutils literal notranslate"><span class="pre">projects/</span></code> directory. We will for this tutorial create the project <code class="docutils literal notranslate"><span class="pre">MyPetProject</span></code>, by convention we use capitalized words. Do <code class="docutils literal notranslate"><span class="pre">cd</span> <span class="pre">projects</span></code> and <code class="docutils literal notranslate"><span class="pre">mkdir</span> <span class="pre">MyPetProject</span></code>.</p>
</section>
<section id="set-up-the-project-code-files">
<h4>Set up the project code files<a class="headerlink" href="#set-up-the-project-code-files" title="Link to this heading"></a></h4>
<p>We need three files in <code class="docutils literal notranslate"><span class="pre">projects/MyPetProject</span></code>, the code file <code class="docutils literal notranslate"><span class="pre">MyPetProject.cpp</span></code>, its header file <code class="docutils literal notranslate"><span class="pre">MyPetProject.h</span></code> and a template/example configuration file used at runtime (waaaay later, not yet), <code class="docutils literal notranslate"><span class="pre">MyPetProject.cfg</span></code>. The first two should be named like that for ease later in the coding process, the configuration file can have any name as that name is being passed as an option <code class="docutils literal notranslate"><span class="pre">--run_config</span> <span class="pre">MyPetProject.cfg</span></code> to the code at execution.</p>
<p>You can start with three empty files or copy them from a suitably similar existing project.</p>
</section>
<section id="code-the-project">
<h4>Code the project<a class="headerlink" href="#code-the-project" title="Link to this heading"></a></h4>
<section id="base-class">
<h5>Base class<a class="headerlink" href="#base-class" title="Link to this heading"></a></h5>
<p>Projects are classes in the <code class="docutils literal notranslate"><span class="pre">projects</span></code> namespace in Vlasiator. Individual projects are derived from the general <code class="docutils literal notranslate"><span class="pre">Project</span></code> class or from some special classes derived from that. The derived special classes are used to optimize the initialization of the velocity space by defining a different <code class="docutils literal notranslate"><span class="pre">findBlocksToInitialize()</span></code> function. This function gives a list of the velocity space blocks which will need to be initialized.</p>
<ul class="simple">
<li><p>In <code class="docutils literal notranslate"><span class="pre">Project</span></code>, all blocks are returned. If you know that a significant fraction will be below the threshold and thus won’t need to be initialized, this is not the good one to choose.</p></li>
<li><p>In the derived class <code class="docutils literal notranslate"><span class="pre">TriAxisSearch</span></code> the project needs to define the function <code class="docutils literal notranslate"><span class="pre">getV0()</span></code> (see below). This is the origin of the Maxwellian population you want, the code then finds out how far around that point it needs to go to catch all velocity cells above the threshold. It is a major gain in efficiency and memory when you have a small Maxwellian in your velocity box, with respect to the default way of looping through the whole velocity space and discarding cells below threshold afterwards.</p></li>
</ul>
<p>If none of the above suits your needs then you are welcome to code a new derived class to provide an efficient algorithm to select the blocks to initialize.</p>
</section>
<section id="basic-structure">
<h5>Basic structure<a class="headerlink" href="#basic-structure" title="Link to this heading"></a></h5>
<p>We present here the backbone of the project files.</p>
<section id="mypetproject-h">
<h6>MyPetProject.h<a class="headerlink" href="#mypetproject-h" title="Link to this heading"></a></h6>
<p>The basic backbone of the header file is</p>
<div class="highlight-c++ notranslate"><div class="highlight"><pre><span></span><span class="cp">#ifndef MYPETPROJECT_H</span>
<span class="cp">#define MYPETPROJECT_H</span>
<span class="cp">#include</span><span class="w"> </span><span class="cpf">&lt;stdlib.h&gt;</span>
<span class="cp">#include</span><span class="w"> </span><span class="cpf">&quot;../../definitions.h&quot;</span>
<span class="cp">#include</span><span class="w"> </span><span class="cpf">&quot;../project.h&quot;</span>
<span class="k">namespace</span><span class="w"> </span><span class="nn">projects</span><span class="w"> </span><span class="p">{</span>
<span class="w">   </span><span class="k">class</span><span class="w"> </span><span class="nc">MyPetProject</span><span class="o">:</span><span class="w"> </span><span class="k">public</span><span class="w"> </span><span class="n">Project</span><span class="w"> </span><span class="p">{</span>
<span class="w">      </span><span class="k">public</span><span class="o">:</span>
<span class="w">         </span><span class="n">MyPetProject</span><span class="p">();</span>
<span class="w">         </span><span class="k">virtual</span><span class="w"> </span><span class="o">~</span><span class="n">MyPetProject</span><span class="p">();</span>
<span class="w">         </span><span class="k">virtual</span><span class="w"> </span><span class="kt">bool</span><span class="w"> </span><span class="nf">initialize</span><span class="p">(</span><span class="kt">void</span><span class="p">);</span>
<span class="w">         </span><span class="k">static</span><span class="w"> </span><span class="kt">void</span><span class="w"> </span><span class="nf">addParameters</span><span class="p">(</span><span class="kt">void</span><span class="p">);</span>
<span class="w">         </span><span class="k">virtual</span><span class="w"> </span><span class="kt">void</span><span class="w"> </span><span class="nf">getParameters</span><span class="p">(</span><span class="kt">void</span><span class="p">);</span>
<span class="w">         </span><span class="k">virtual</span><span class="w"> </span><span class="kt">void</span><span class="w"> </span><span class="nf">calcCellParameters</span><span class="p">(</span><span class="n">Real</span><span class="o">*</span><span class="w"> </span><span class="n">cellParams</span><span class="p">,</span><span class="n">creal</span><span class="o">&amp;</span><span class="w"> </span><span class="n">t</span><span class="p">);</span>
<span class="w">         </span><span class="k">virtual</span><span class="w"> </span><span class="n">Real</span><span class="w"> </span><span class="nf">calcPhaseSpaceDensity</span><span class="p">(</span>
<span class="w">            </span><span class="n">creal</span><span class="o">&amp;</span><span class="w"> </span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">creal</span><span class="o">&amp;</span><span class="w"> </span><span class="n">y</span><span class="p">,</span><span class="w"> </span><span class="n">creal</span><span class="o">&amp;</span><span class="w"> </span><span class="n">z</span><span class="p">,</span>
<span class="w">            </span><span class="n">creal</span><span class="o">&amp;</span><span class="w"> </span><span class="n">dx</span><span class="p">,</span><span class="w"> </span><span class="n">creal</span><span class="o">&amp;</span><span class="w"> </span><span class="n">dy</span><span class="p">,</span><span class="w"> </span><span class="n">creal</span><span class="o">&amp;</span><span class="w"> </span><span class="n">dz</span><span class="p">,</span>
<span class="w">            </span><span class="n">creal</span><span class="o">&amp;</span><span class="w"> </span><span class="n">vx</span><span class="p">,</span><span class="w"> </span><span class="n">creal</span><span class="o">&amp;</span><span class="w"> </span><span class="n">vy</span><span class="p">,</span><span class="w"> </span><span class="n">creal</span><span class="o">&amp;</span><span class="w"> </span><span class="n">vz</span><span class="p">,</span>
<span class="w">            </span><span class="n">creal</span><span class="o">&amp;</span><span class="w"> </span><span class="n">dvx</span><span class="p">,</span><span class="w"> </span><span class="n">creal</span><span class="o">&amp;</span><span class="w"> </span><span class="n">dvy</span><span class="p">,</span><span class="w"> </span><span class="n">creal</span><span class="o">&amp;</span><span class="w"> </span><span class="n">dvz</span>
<span class="w">         </span><span class="p">);</span>
<span class="w">      </span><span class="k">protected</span><span class="o">:</span>
<span class="w">         </span><span class="c1">// (whatever you need)</span>
<span class="w">   </span><span class="p">};</span><span class="w"> </span><span class="c1">// class MyPetProject</span>
<span class="p">}</span><span class="w"> </span><span class="c1">// namespace</span>
<span class="cp">#endif</span>
</pre></div>
</div>
</section>
<section id="mypetproject-cpp">
<h6>MyPetProject.cpp<a class="headerlink" href="#mypetproject-cpp" title="Link to this heading"></a></h6>
<p>The basic backbone of the code file is (functions are explained below, not included here)</p>
<div class="highlight-c++ notranslate"><div class="highlight"><pre><span></span><span class="cp">#include</span><span class="w"> </span><span class="cpf">&lt;cstdlib&gt;</span>
<span class="cp">#include</span><span class="w"> </span><span class="cpf">&lt;iostream&gt;</span>
<span class="cp">#include</span><span class="w"> </span><span class="cpf">&lt;cmath&gt;</span>

<span class="cp">#include</span><span class="w"> </span><span class="cpf">&quot;../../common.h&quot;</span>
<span class="cp">#include</span><span class="w"> </span><span class="cpf">&quot;../../readparameters.h&quot;</span>
<span class="cp">#include</span><span class="w"> </span><span class="cpf">&quot;../../backgroundfield/backgroundfield.h&quot;</span>

<span class="cp">#include</span><span class="w"> </span><span class="cpf">&quot;MyPetProject.h&quot;</span>
<span class="k">namespace</span><span class="w"> </span><span class="nn">projects</span><span class="w"> </span><span class="p">{</span>
<span class="w">   </span><span class="n">MyPetProject</span><span class="o">::</span><span class="n">MyPetProject</span><span class="p">()</span><span class="o">:</span><span class="w"> </span><span class="n">Project</span><span class="p">()</span><span class="w"> </span><span class="p">{</span><span class="w"> </span><span class="p">}</span>
<span class="w">   </span><span class="n">MyPetProject</span><span class="o">::~</span><span class="n">MyPetProject</span><span class="p">()</span><span class="w"> </span><span class="p">{</span><span class="w"> </span><span class="p">}</span>
<span class="w">   </span><span class="c1">// we do not use the constructor/destructor at the moment</span>
<span class="w">   </span><span class="c1">// your code, see functions below</span>
<span class="p">}</span><span class="w"> </span><span class="c1">// namespace projects</span>
</pre></div>
</div>
</section>
</section>
<section id="functions">
<h5>Functions<a class="headerlink" href="#functions" title="Link to this heading"></a></h5>
<p>In the following we explain the functions you need in your project, i.e. the functions the code expects you to have in order to work. Not having them usually leads to an error message from the base project class informing you that you should use the function from the derived class (your project) and not the base class function. The constructors/destructors are not used at the moment.</p>
<section id="mypetproject-addparameters">
<h6>MyPetProject::addParameters()<a class="headerlink" href="#mypetproject-addparameters" title="Link to this heading"></a></h6>
<p>This is a static function because all <code class="docutils literal notranslate"><span class="pre">addParameters()</span></code> get called by the code to provide complete help. We use the Boost program options, thus to add a parameter use <code class="docutils literal notranslate"><span class="pre">ReadParameters::add(&quot;MyPetProject.param&quot;,</span> <span class="pre">&quot;This</span> <span class="pre">is</span> <span class="pre">the</span> <span class="pre">parameter</span> <span class="pre">for</span> <span class="pre">the</span> <span class="pre">MyPetProject</span> <span class="pre">project.&quot;,</span> <span class="pre">0.0);</span></code>. The first is the option name, it can then be used as such in a command line, like <code class="docutils literal notranslate"><span class="pre">--MyPetProject.param</span> <span class="pre">1.0</span></code> or in the configuration file as an entry of the form</p>
<p>but details on the configuration file come later. The last field is the default value in case the option is not set by the user. Make it a sensible value to avoid head-scratching and bug-hunting when a user wonders why your project does not work.</p>
</section>
<section id="mypetproject-getparameters">
<h6>MyPetProject::getParameters()<a class="headerlink" href="#mypetproject-getparameters" title="Link to this heading"></a></h6>
<p>This function is used to read in the parameters, it is used only if you actually use this project. The basic syntax is <code class="docutils literal notranslate"><span class="pre">ReadParameters::get(&quot;MyPetProject.param&quot;,</span> <span class="pre">this-&gt;param);</span></code>, where we typically save the parameter’s value into a member of the <code class="docutils literal notranslate"><span class="pre">MyPetProject</span></code> class. Of course it can be local to the function instead if it is not needed anywhere else, or you save it differently, as you wish.</p>
</section>
<section id="mypetproject-initialize">
<h6>MyPetProject::initialize()<a class="headerlink" href="#mypetproject-initialize" title="Link to this heading"></a></h6>
<p>This function can be used to set up things before any major computation is done, it is called early in the simulation initialization process. If nothing is needed, just <code class="docutils literal notranslate"><span class="pre">return</span> <span class="pre">true;</span></code>.</p>
</section>
<section id="mypetproject-setcellbackgroundfield">
<h6>MyPetProject::setCellBackgroundField()<a class="headerlink" href="#mypetproject-setcellbackgroundfield" title="Link to this heading"></a></h6>
<p>Using the capabilities offered by the background field classes, you can set what you need here (constant or dipole at the moment). If you wish to do it by hand, make sure you also set all relevant derivative terms and not only the fields. Note that the background field is assumed to be curl-free, if it is not the calculations involving the current density in the Vlasov and field solvers are wrong.</p>
</section>
<section id="mypetproject-calccellparameters">
<h6>MyPetProject::calcCellParameters()<a class="headerlink" href="#mypetproject-calccellparameters" title="Link to this heading"></a></h6>
<p>This function is used to set the cell’s (perturbed) magnetic field components. The electric field is computed self-consistently by the field solver.</p>
</section>
<section id="mypetproject-calcphasespacedensity">
<h6>MyPetProject::calcPhaseSpaceDensity()<a class="headerlink" href="#mypetproject-calcphasespacedensity" title="Link to this heading"></a></h6>
<p>This function is used to calculate the phase space density in each of the simulation cells in six dimensions. Typical examples are Maxwellians. Often one can code some form of averaging loops in that function and call one further function which has the actual distribution function calculation.</p>
</section>
<section id="triaxissearch-mypetproject-getv0">
<h6>(TriAxisSearch) MyPetProject::getV0()<a class="headerlink" href="#triaxissearch-mypetproject-getv0" title="Link to this heading"></a></h6>
<p>If you use the <code class="docutils literal notranslate"><span class="pre">TriAxisSearch</span></code> base class you have to provide this function. It must give the centre coordinate of the Maxwellian you want so that the <code class="docutils literal notranslate"><span class="pre">findBlocksToInitialize()</span></code> function can find out within what radius around this centre velocity blocks should be initialized.</p>
</section>
</section>
</section>
<section id="write-a-sensible-default-configuration-file">
<h4>Write a sensible default configuration file<a class="headerlink" href="#write-a-sensible-default-configuration-file" title="Link to this heading"></a></h4>
<p>Once you coded your project and you know what parameters you will need, write a default configuration file to document workable and sensible options for your project. This file will be saved along in the repository for reference. Try to keep it up-to-date during the life of the project so that it still reflects a sensible state and not what you had in your crazy mind when you just made that file to comply with this paragraph. Otherwise you will incur the wrath of the next user trying to quickly run your project for a test and the pain of figuring out a new set of sensible parameters after failing to back up the configuration files you were actually using.</p>
<p>The only compulsory parameter in this file is the line</p>
<div class="highlight-cfg notranslate"><div class="highlight"><pre><span></span><span class="na">project</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">MyPetProject</span>
</pre></div>
</div>
<p>otherwise Vlasiator will not run your project, no matter what.</p>
<p>A useful tool to check a cfg file is <code class="docutils literal notranslate"><span class="pre">tools/check_vlasiator_cfg.sh</span></code>. It takes the <code class="docutils literal notranslate"><span class="pre">vlasiator</span></code> executable as a first argument and the cfg to check as second argument and returns a list of unused available options as well as a list of invalid options.</p>
</section>
<section id="integrate-the-project-to-vlasiator">
<h4>Integrate the project to Vlasiator<a class="headerlink" href="#integrate-the-project-to-vlasiator" title="Link to this heading"></a></h4>
<p>We decided that all projects should be compiled when compiling Vlasiator. This avoids hassle with the Makefile and it also helps to keep projects supported when coding new things related to the project class infrastructure (… or shuffled into the <code class="docutils literal notranslate"><span class="pre">unsupported</span></code> folder…). But this comes at the cost of adding some bits here and there. The file <code class="docutils literal notranslate"><span class="pre">projects/project.cpp</span></code> must be edited.</p>
<ul>
<li><p>Add <code class="docutils literal notranslate"><span class="pre">#include</span> <span class="pre">&quot;MyPetProject/MyPetProject.h&quot;</span></code> in the top section. Please use alphabetic ordering.</p></li>
<li><p>Edit <code class="docutils literal notranslate"><span class="pre">Project::addParameters()</span></code> so that it also calls <code class="docutils literal notranslate"><span class="pre">projects::MyPetProject::addParameters();</span></code>. You see, that’s why it is static, we told you. Please use alphabetic ordering.</p></li>
<li><p>If your project is so cool it created a new parameter that might be useful to other projects, add it to the <code class="docutils literal notranslate"><span class="pre">Project_common</span></code> category in this function and in <code class="docutils literal notranslate"><span class="pre">Project::getParameters()</span></code>. If now you see that you need a parameter that actually was already available through the <code class="docutils literal notranslate"><span class="pre">Project_common</span></code> parameters, edit your code accordingly, no need to import twice the same stuff.</p></li>
<li><p>In the function <code class="docutils literal notranslate"><span class="pre">createProject()</span></code>, add</p>
<div class="highlight-c++ notranslate"><div class="highlight"><pre><span></span><span class="k">if</span><span class="p">(</span><span class="n">Parameters</span><span class="o">::</span><span class="n">projectName</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s">&quot;MyPetProject&quot;</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">   </span><span class="k">return</span><span class="w"> </span><span class="k">new</span><span class="w"> </span><span class="n">projects</span><span class="o">::</span><span class="n">MyPetProject</span><span class="p">;</span>
<span class="p">}</span>
</pre></div>
</div>
</li>
</ul>
<p>and yes, you guessed it, please use alphabetic ordering.</p>
<section id="set-up-the-project-compilation">
<h5>Set up the project compilation<a class="headerlink" href="#set-up-the-project-compilation" title="Link to this heading"></a></h5>
<p>You’re almost there. Now the code is ready, it needs to be compiled. For that, obviously, the <code class="docutils literal notranslate"><span class="pre">Makefile</span></code> needs to be edited.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">DEPS_PROJECTS</span></code> lists all project files, so add <code class="docutils literal notranslate"><span class="pre">projects/MyPetProject/MyPetProject.h</span></code> and <code class="docutils literal notranslate"><span class="pre">projects/MyPetProject/MyPetProject.cpp</span></code>. Please use alphabetic ordering.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">OBJS</span></code> should now include <code class="docutils literal notranslate"><span class="pre">MyPetProject.o</span></code>. Please…</p></li>
<li><p>Down in the actual making commands, add the relevant lines. Guess what order?</p></li>
</ul>
<p>As a savvy <code class="docutils literal notranslate"><span class="pre">Makefile</span></code> guru you will remember that before <code class="docutils literal notranslate"><span class="pre">${CMP}</span></code> it is a tab character, not spaces.</p>
</section>
</section>
</section>
<section id="make-debug">
<h3>(“make”, “debug”)+<a class="headerlink" href="#make-debug" title="Link to this heading"></a></h3>
<p>Now starts the actual work. In the base folder (where the <code class="docutils literal notranslate"><span class="pre">Makefile</span></code> is), use <code class="docutils literal notranslate"><span class="pre">make</span></code> to compile. If possible, use multiple processes to accelerate compilation by using <code class="docutils literal notranslate"><span class="pre">make</span> <span class="pre">-j</span> <span class="pre">N</span></code> where N is the number of concurrent compiling processes you want to use. It should be close to the number of physical cores available but not too much higher. In the unlikely event that the compilation should stop because it did not understand your code, debug, and iterate the above…</p>
</section>
</section>
</div>
<section id="who-is-the-course-for">
<span id="learner-personas"></span><h2>Who is the course for?<a class="headerlink" href="#who-is-the-course-for" title="Link to this heading"></a></h2>
<p>This hybrid workshop will bring together code developers, researchers, and research software
engineers working on plasma science, and create an amazing opportunity for sharing innovative
ideas and best practice for potential users to use the Vlasiator package and using its
capabilities and assorted tools for data analysis.</p>
</section>
<section id="about-the-course">
<h2>About the course<a class="headerlink" href="#about-the-course" title="Link to this heading"></a></h2>
<p>This workshop includes:</p>
<ul class="simple">
<li><p>running simulations at scale</p></li>
<li><p>benchmarking for deploying Vlasiator on supercomputing environments</p></li>
<li><p>designing a simulation setup, with notes on applicability and resources required</p></li>
<li><p>Accessing the .vlsv data via Analysator</p></li>
<li><p>Accessing the .vlsv data via the VisIt plugin</p></li>
</ul>
<p>After attending this workshop, you will:</p>
<ul class="simple">
<li><p>Understand core features of the Vlasiator package</p></li>
<li><p>Be efficient using the Vlasiator package to perform plasma simulations</p></li>
<li><p>Be productive in data analysis and visualization of simulation results</p></li>
<li><p>Be able to create your own project</p></li>
</ul>
</section>
<section id="see-also">
<h2>See also<a class="headerlink" href="#see-also" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>ENCCS: <a class="reference external" href="https://enccs.se/">https://enccs.se/</a></p></li>
<li><p>Plasma-PEPSC CoE: <a class="reference external" href="https://plasma-pepsc.eu/">https://plasma-pepsc.eu/</a></p></li>
<li><p>Follow ENCCS on <a class="reference external" href="https://www.linkedin.com/company/enccs">LinkedIn</a>, or <a class="reference external" href="https://twitter.com/EuroCC_Sweden">Twitter</a></p></li>
</ul>
</section>
<section id="credits">
<h2>Credits<a class="headerlink" href="#credits" title="Link to this heading"></a></h2>
<p>The lesson file structure and browsing layout is inspired by and derived from
<a class="reference external" href="https://github.com/coderefinery/sphinx-lesson">work</a> by <a class="reference external" href="https://coderefinery.org/">CodeRefinery</a> licensed under the <a class="reference external" href="http://opensource.org/licenses/mit-license.html">MIT license</a>. We have copied and adapted
most of their license text.</p>
<section id="instructional-material">
<h3>Instructional Material<a class="headerlink" href="#instructional-material" title="Link to this heading"></a></h3>
<p>This instructional material is made available under the
<a class="reference external" href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license (CC-BY-4.0)</a>.
The following is a human-readable summary of (and not a substitute for) the
<a class="reference external" href="https://creativecommons.org/licenses/by/4.0/legalcode">full legal text of the CC-BY-4.0 license</a>.
You are free to:</p>
<ul class="simple">
<li><p><strong>share</strong> - copy and redistribute the material in any medium or format</p></li>
<li><p><strong>adapt</strong> - remix, transform, and build upon the material for any purpose,
even commercially.</p></li>
</ul>
<p>The licensor cannot revoke these freedoms as long as you follow these license terms:</p>
<ul class="simple">
<li><p><strong>Attribution</strong> - You must give appropriate credit (mentioning that your work
is derived from work that is Copyright (c) ENCCS and individual contributors and, where practical, linking
to <a class="reference external" href="https://enccs.github.io/sphinx-lesson-template">https://enccs.github.io/sphinx-lesson-template</a>), provide a <a class="reference external" href="https://creativecommons.org/licenses/by/4.0/">link to the license</a>, and indicate if changes were
made. You may do so in any reasonable manner, but not in any way that suggests
the licensor endorses you or your use.</p></li>
<li><p><strong>No additional restrictions</strong> - You may not apply legal terms or
technological measures that legally restrict others from doing anything the
license permits.</p></li>
</ul>
<p>With the understanding that:</p>
<ul class="simple">
<li><p>You do not have to comply with the license for elements of the material in
the public domain or where your use is permitted by an applicable exception
or limitation.</p></li>
<li><p>No warranties are given. The license may not give you all of the permissions
necessary for your intended use. For example, other rights such as
publicity, privacy, or moral rights may limit how you use the material.</p></li>
</ul>
</section>
<section id="software">
<h3>Software<a class="headerlink" href="#software" title="Link to this heading"></a></h3>
<p>Except where otherwise noted, the example programs and other software provided
with this repository are made available under the <a class="reference external" href="http://opensource.org/">OSI</a>-approved
<a class="reference external" href="https://opensource.org/licenses/mit-license.html">MIT license</a>.</p>
</section>
</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, The contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>